---
title: "PF Liver Stage Time Course Analysis"
author: "Jenny L Smith"
date: "`r Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(tidyr)
library(tibble)
library(magrittr)
library(dplyr)
library(stringr)
library(glue)

library(ggplot2)
library(gridExtra)
library(ggpubr)
library(patchwork)
library(ComplexHeatmap)
library(circlize)

library(limma)
library(edgeR)
library(SummarizedExperiment)
library(patchwork)
library(maSigPro)
library(svglite)
suppressPackageStartupMessages(library(mclust))
```

```{r}
#https://www.tidyverse.org/blog/2021/02/svglite-2-0-0/
# system_fonts = list(sans = "Open Sans")
# systemfonts::system_fonts()
```


# Define Functions 

```{r}
colors_vector <- c(RColorBrewer::brewer.pal(9, "Set1"), 
            "blue1", "darkslategray3", "burlywood3", "#984EA3",
            "seagreen1", "yellow2", "orchid", "darkblue", 
            "lightsalmon2","slateblue1","lightskyblue4",
            "azure2", "chartreuse1",  "lemonchiffon2",
            "deeppink", "darkslategray1", "green4", "navajowhite", 
            "brown4", "darkgoldenrod2", "deepskyblue1", "lightpink") %>% 
  # c(., ggpubr::get_palette("jco", 5)) %>% 
  c("mediumpurple4","magenta", "peru") %>% 
  c(ggsci::pal_igv()(20))

# length(colors_vector)
```


```{r}
#from https://github.com/mikelove/DESeq2/blob/master/R/plots.R
#Want to return the whole scores matrix so can examine 3d pca plots. 
plotPCA.DESeq.mod <- function(object, intgroup="condition", ntop=500, returnData=FALSE, PC3=FALSE)
{
  library(matrixStats)
  # calculate the variance for each gene
  rv <- rowVars(assay(object))
  
  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))
  
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )
  
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }
  
  # assembly the data for the plot - first 10 PCs
  # d <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], group=group, intgroup.df, name=colnames(object))
  n <- min(10, ncol(as.data.frame(pca$x)))
  d <- data.frame(as.data.frame(pca$x)[,1:n], group=group, intgroup.df, name=colnames(object))
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:10]
    rot <- pca$rotation[,1:10] #for first 10 PCs
    dat <- list("scores"=d,"rotation"=rot)
    return(dat)
  }
}

#Updated on 6/9/17 to use variance stabilized transformed data as input (not center scaled log2, like in princomp)
PCA <- function(expnData,phenovector,
                title="",round=TRUE,colorCodes=NULL,
                ntop=500, GOI=NULL){
  
  suppressPackageStartupMessages(library(DESeq2))
  library(ggplot2)
  #expnData is the raw counts (not normalized) has patient IDs as colnames and genes as rownames. 
  
  samples <- intersect(names(phenovector), colnames(expnData))
  countData <- expnData[,samples]
  phenovector <- phenovector[samples]
  colData <- as.data.frame(phenovector)
  
  if(round){
    countData <- round(countData, digits = 0)
  }
  
  #Create as DESeq data set object (dds)
  dds <- DESeqDataSetFromMatrix(countData = countData,
                                  colData = colData,
                                  design = ~ 1)
  
  #perform variance stabilized transformation 
  dds <- dds[ rowSums(counts(dds)) > 10, ]
  varianceStab <- vst(dds, blind = TRUE)
  
  #if given a list of genes of interest
  if (! is.null(GOI)){
    GOI <- intersect(GOI, rownames(assay(varianceStab)))
  }else{
    GOI <- 1:nrow(varianceStab)
  }
  
  #PCA data frame with the wieghts/loadings and eigen vectors
  pca.dat <- plotPCA.DESeq.mod(varianceStab[GOI,], 
                               intgroup = "phenovector", 
                               ntop = ntop,
                               returnData=TRUE)
  
  plots <- lapply(c(2:3), function(pc){
      percentVar <- attr(pca.dat$scores, which="percentVar")
      y_var <- paste0("PC",pc)
      
      p <- ggplot(data=pca.dat$scores, 
             aes_string(x="PC1", y=y_var, color="phenovector")) + 
        geom_point(size=3, alpha=0.75) +
        xlab(paste0("PC1: ",round(percentVar[1] * 100),"% variance")) +
        ylab(paste0(y_var,": ",round(percentVar[pc] * 100),"% variance")) +
        labs(title=title) +
        theme_classic() +
        theme(legend.position = "top")
      if(!is.null(colorCodes)){
        p <- p + 
          scale_color_manual(values=colorCodes)
      }
      return(p)
  })
  
  #Final Results object
  res <- list(dds, varianceStab,pca.dat, plots)
  names(res) <- c("dds", "vst","pca_data","pca_plots")
  
  if(is.character(GOI)){
    res[["GOI"]] <- GOI
  }
  
  return(res)
}
```

```{r}
#changed on 2/14/18, see bottom of Heatmaps_Function.r for the original one used.
dge_dendrograms <- function(expnData, pheno, method,
                            genelist=NULL,add.count=1, percent=0.05,
                            ntop=500,filterTopGenes=FALSE, createDGE=TRUE,log=FALSE){
  #df with count data, patient IDs are column names and rows are genes.
  #pheno is a character vector with patient IDs as names, and the status for each in each group (eg pos,neg)
  #genelist is a character vector with the genes of interest
  #percent is the % of samples in the input expn matrix that must express a gene at 1 CPM. Filter to remove low count genes.
  #set log=TRUE if providing a log2 expression dataframe.
  #filterTopGenes shuold be a logical. If TRUE filter top 1000 most varied genes.
  suppressPackageStartupMessages(require(edgeR))
  suppressPackageStartupMessages(library(dendextend))
  suppressPackageStartupMessages(library(matrixStats))

  expnData <- expnData[, intersect(names(pheno), colnames(expnData))] #ensure correct order, drop rows with nas just in case
  if(createDGE){
    dge <- DGEList(counts = expnData)
    #keep the rows (genes) with at least 1 CPM in a minimum of 2 samples or X% of samples. 
    keep.dge <- rowSums(cpm(dge) >= 1) >= max(2,(percent*ncol(expnData))) 
    # subset for those genes with cmp >= 1 per gene in samples
    dge <- dge[keep.dge,] #
    dge <- calcNormFactors(dge)
    TMMCPM <- cpm(dge, normalized.lib.sizes = TRUE,
                  log = TRUE, prior.count = add.count)

  }else{
    TMMCPM <- as.matrix(expnData)
    if(!log){
      TMMCPM <- log2(expnData+add.count) #log2 transform counts
    }
  }

  if(is.null(genelist) & filterTopGenes){
      # calculate the variance for each gene
      rv <- rowVars(as.matrix(TMMCPM))
      # select the ntop genes by variance
      select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
      #select the top 1000 most varied genes
      TMMCPM <- TMMCPM[select,]
  }else if(!is.null(genelist)){
    TMMCPM <- TMMCPM[which(rownames(TMMCPM) %in% genelist), ] #subset the matrix to genes of interest
  }


  d1 <- dist(t(TMMCPM), method = "euclidean", diag = FALSE,
             upper = FALSE) #sample distances WITHOUT SCALING
  d2 <- dist(TMMCPM, method = "euclidean", diag = FALSE,
             upper = TRUE) #gene distances WITHOUT SCaling
  samp.c1 <- hclust(d1, method = method, members = NULL) #sample clustering
  gene.c2 <- hclust(d2, method = method, members = NULL) #gene clustering
  list <- list(TMMCPM,samp.c1,gene.c2)
  names(list) <- c("TMMCPM","samp.c1", "gene.c2")

  return(list)
}
```

```{r}
#Pheatmap
quickPheatmap <- function(expn,geneList, sample_data,
                          gene_index=NULL,
                          annots_col_colors=NULL){
  library(pheatmap)

  len <- 299
  col <- colorRampPalette(c("deepskyblue4", "deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=len)

  e <- expn[geneList, intersect(colnames(expn), rownames(sample_data))]
  sample_data <- sample_data[intersect(colnames(expn), rownames(sample_data)), ]
  anno.col <- sample_data
  #if scaling, this doesnt work. would need to scale the expression matrix myself
  # Breaks <- c(seq(min(e), 0, length.out=ceiling(len/2) + 1),
  #             seq(max(e)/len, max(e), length.out=floor(len/2)))
 
  if(is.null(annots_col_colors)){
    annots_col_colors <- lapply(anno.col, function(column){
        grps <- unique(column)
        n <- length(grps)
        pal <- RColorBrewer::brewer.pal(n,name="Set1")
        names(pal) <- grps
        return(pal)
      })
  }


  p <- pheatmap::pheatmap(mat= e ,
                     color=col,
                     border_color="black",
                     scale = "row", 
                     cluster_rows=FALSE,
                     cluster_cols=TRUE,
                     clustering_method="ward.D2",
                     clustering_distance_cols="correlation",
                     annotation_names_col=TRUE,
                     annotation_col = anno.col,
                     annotation_colors = annots_col_colors,
                     gaps_row=gene_index,
                     show_rownames=TRUE,
                     fontsize_row=5,
                     fontsize_col = 10,
                     # kmeans_k=9,
                     show_colnames=TRUE)


  return(p)

}
```

```{r}
create_annots_hmap <- function(expn, pheno_df, cols,
                               goi=NULL,cc=NULL, 
                               colorbar.height=5){
  #expn is the normalized expression values with genes as rownames
  #gene list a character vector
  #goi are genes of interest to highlight on the Heatmap. Character vector of gene symbols
  #cc are color codes in WHICH FORMAT?
  #cols is a character vector of column names

  #Example on how to use the reuslts:
  #hmap <- ComplexHmap(XXXX, hmap_anno_obj=res$annoColumn, XXX)
  # draw(hmap + res$geneLabels, heatmap_legend_side="right", annotation_legend_side="right")

    library(dplyr)
    suppressPackageStartupMessages(library(ComplexHeatmap))

    #subset the expression matix and the phenotype dataframe
    anno <- pheno_df %>%
      dplyr::filter(rownames(.) %in% colnames(expn)) %>% 
      rownames_to_column("id") %>% 
      dplyr::select(id, all_of(cols)) %>%
      mutate(id = factor(id, levels = colnames(expn))) %>%
      arrange(id) %>%  #ensure same order as the expn matrix
      column_to_rownames("id")
  
    #if no color codes provided, create one for each column
    if(is.null(cc)){
      pals <- c("npg", "aaas", "lancet", "jco", 
                    "ucscgb", "uchicago", "simpsons",
                    "rickandmorty")
      #if there are more columns than color palettes, just repeat the set of available color palettes
      if(length(ncol(anno)) > length(pals)){
        pals <- rep(pals,  ceiling(ncol(anno)/length(pals)))
      }
      #define unique colors for each column
      cc <- purrr::map(1:ncol(anno), function(i){
          grps <- unique(anno[[i]])
          ggpubr::get_palette(pals[i], k=length(grps)) %>% 
            magrittr::set_names(grps)
      })
      names(cc) <- colnames(anno)
    }
    
    #legend graphical parameters
    n <- max(sapply(cc, length))
    nrow <- ifelse(n <= 6, n, 6)
    ncol <- ceiling(n/nrow)
    params <- list(show_legend=TRUE,
                  labels_gp= gpar(fontsize=12),
                  title_gp= gpar(fontsize=16),
                  nrow = nrow,
                  ncol=ncol,
                  by_row=TRUE)
    
    #Create the complex heatmap annotation column
    annoCol <- suppressWarnings(HeatmapAnnotation(df = dplyr::select(anno, all_of(cols)),
                                   name="Main Groups",
                                   col=cc,
                                   which="column",
                                   gap=unit(1,"mm"),
                                   border = T,
                                   show_annotation_name = TRUE,
                                   annotation_name_gp=gpar(fontsize=12),
                                   annotation_name_offset=unit(1,"mm"),
                                   annotation_name_side="left",
                                   annotation_height=unit(colorbar.height, "cm"),
                                   annotation_legend_param = params,
                                   simple_anno_size_adjust=TRUE))
    res <- list("annoColumn"=annoCol)
    if(!is.null(goi)){
      regex <- paste0("^",goi,"$", collapse = "|")
      goi.idx <- grep(regex, rownames(expn))
      labels <- rownames(expn)[goi.idx] #needs to be numeric indexes for complexheatmap
      #create the row (gene) labels object
      labs <- ComplexHeatmap::rowAnnotation(link = anno_mark(at=goi.idx,
                                             labels=labels,
                                             which="row",
                                             link_width=unit(1, "mm")),
                            width= unit(1, "mm") + max_text_width(labels),
                            gp=gpar(fontsize=4))

      res[["geneLabels"]] <-  labs
    }
    return(res)
}


complex_hmap <- function(mat,
                        hmap_anno_obj,
                        hmap_anno_obj_genes=NULL,
                        name="z-scores",
                        scale=TRUE,
                        space.type="sRGB",
                        color_palette=NULL,
                        split=NULL, 
                        cluster.method="ward.D2",
                        show_sample_ids=FALSE,
                        dge_dendrograms.res=NULL,
                        samp_dend_order=NULL){
  #mat is the normalized, log2 (usually) transformed counts
  #name is the title
  #scale is whether to scale by row
  #color palette is a colorRamp2() object.
  #threshold is whether to make all z-scores in a certain range.
  #hmap_anno_obj is from HeatmapAnnotation() function
  #space.type is for the color/shades on the heatmap. See ?Heatmap for all the choices.
  #dge_dendrograms.res is the list object output from the dge_dendrograms() function.
  #samp_dend_order is the numeric vector or character vector of column names from the matrix (mat) or the dge_dengrograms.res$TMMCPM matix, in the desired order.

  suppressPackageStartupMessages(library(ComplexHeatmap))
  suppressPackageStartupMessages(require(circlize))
  library(RColorBrewer)
  suppressPackageStartupMessages(library(dendextend))
  ht_opt$message = FALSE

   if(scale){
      mat <- t(scale(t(mat))) ##rowwise scaling
   }
  
  print(range(mat))
  min <- round(min(mat))
  max <- round(max(mat))
  mid <- round(max/2, digits = 2)
    
  if(is.null(color_palette)){
    pal <- colorRamp2(c(min,-mid, 0, mid, max),
                      c("deepskyblue3", "deepskyblue","white", "red", "red3"),
                      space=space.type)
  }else{
    pal <- color_palette
  }
  # col <- colorRampPalette(c("cyan1", "cyan2", "cyan3", "cyan4","azure4","magenta4", "magenta3", "magenta2", "magenta1"))(n=299)
  #colorRamp2(c(-2, 0, 4), c("deepskyblue","white", "red"), space="RGB") #use for breaks.
  # colorPal <- colorRampPalette(c("deepskyblue4", "deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=299)

  #legend graphical parameters
  params <-  list(color_bar="continuous",
       legend_direction="horizontal",
       title_position="topcenter",
       legend_width=unit(5,"cm"),
       legend_height=unit(5,"cm"),
       title_gp=gpar(fontsize=10,
                     fontface="bold"),
      at = c(min,-mid, 0,mid, max),
      just = c("left", "top"))
  
    if(!is.null(dge_dendrograms.res)){
        if(is.null(samp_dend_order)){
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                           order=c(ncol(mat):1))
        }else{
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                                  order=samp_dend_order)
      }
    }else{
      clust=TRUE
    }

    #create the heatmap plot
    hmap <- Heatmap(mat,
                    name=name,
                    col=pal,

                    heatmap_legend_param=params,
                    row_title="Genes",
                    row_title_side="left",
                    row_title_gp=gpar(fontsize=15,
                                      fontface="bold"),
                    show_row_names=FALSE,
                    show_column_names=show_sample_ids,
                    row_names_gp=gpar(fontsize=3),

                    column_title="Samples",
                    column_title_side="bottom",
                    column_title_gp=gpar(fontsize=15,
                                         fontface="bold"),
                    column_title_rot=0,
                    row_dend_width=unit(8,"mm"),
                    column_dend_height=unit(22.5,"mm"),

                    top_annotation=hmap_anno_obj,
                    right_annotation = hmap_anno_obj_genes,
                    split=split,
                    row_gap = unit(5, "mm"),

                    clustering_distance_rows="euclidean",
                    clustering_method_rows=cluster.method,
                    clustering_method_columns = cluster.method,
                    cluster_columns = clust,
                    column_dend_reorder=FALSE)

  return(hmap)
}
```

```{r}
create_pf_heatmaps <- function(TMMCPM,sample_annots, gene_annots,
                               center_scale=TRUE,
                               ntop=1000,order_ids=NULL,
                               genelist=NULL,
                               split=NULL,
                               heatmap_colors=NULL,
                               GOI=NULL){
  pf_genes <- gene_annots %>% 
    as.data.frame() %>% 
    dplyr::filter(genome=="PF3D7") %>% 
    pull(gene_id)
  
  TMMCPM_all_pf <- TMMCPM[rownames(TMMCPM) %in% pf_genes, ]
  dends <- dge_dendrograms(expnData = TMMCPM_all_pf, 
                           pheno = pull(sample_annots, analysis_factor_group, name = masigpro_id), 
                           method = "ward.D2",
                           genelist = genelist, 
                           ntop=ntop,
                           log=FALSE,
                           filterTopGenes = TRUE,
                           createDGE = FALSE)
  
  # dim(dends$TMMCPM)
  hmap_anno_obj <- create_annots_hmap(expn=dends$TMMCPM,
                                      pheno_df = sample_annots,
                                      cc=NULL,
                                      colorbar.height=2,
                                      goi = GOI,
                                      cols = c("analysis_factor_group"))
  
  name <- ifelse(center_scale, "z-scores","log2_CPM")
  res <- complex_hmap(mat=dends$TMMCPM, 
               scale=center_scale,
               name=name,
               hmap_anno_obj = hmap_anno_obj$annoColumn,
               dge_dendrograms.res = dends,
               samp_dend_order=order_ids,
               color_palette=heatmap_colors,
               split=split,
               show_sample_ids=TRUE)
  
  return(list("dends"=dends,"heatmap"=res, "hmap_anno"=hmap_anno_obj))
  
}
```


```{r}
motif_enrichment_workflow <- function(targets_df, 
                                      col_to_bin="logFC",
                                      PWMs = AP2_PWMlist,
                                      proms_gr, genome_bs, 
                                      score = 10,
                                      make_bins = TRUE,
                                      use_bg_regions = TRUE,
                                      region_length=1000,
                                      seed=20221012
                                      ){
    # targets_df is a dataframe with gene_ids and column of numeric values for enrichement (ex, logFC)

    proms <- proms_gr[proms_gr$gene_id %in% targets_df$gene_id]
    genes_removed <-  width(proms) %>%  table
    proms <- proms[!width(proms) < region_length]
    
    #retrienve the DNA seqs for the promoter regions of interest
    target_seqs <- Biostrings::getSeq(genome, proms)

    # Re-order each gene's enrichment value to its matching transcript promoter region
    enr_values <- proms %>%
      as.data.frame() %>%
      #ensure correct order and account for genes with > 1 transcript ID
      left_join(., dplyr::select(targets_df, gene_id, all_of(col_to_bin)),
                by="gene_id") %>%
      arrange(!! rlang::sym(col_to_bin)) %>%
      pull(all_of(col_to_bin), name=tx_name)

    #Add the value to the metadata of the gr object
    target_seqs <- target_seqs[names(enr_values)]
    mcols(proms)[[col_to_bin]] <- enr_values
    
    # Bin the numeric values
    if(!make_bins){
      input_bins <- enr_values
    }else{
        if( col_to_bin == "betaTime" ){
          input_bins <- enr_values %>%
            bin(.,
                binmode = "equalWidth",
                nBins = 1)
        }else {
          min_val <- min(enr_values)
          max_val <- max(enr_values)
          input_bins <- enr_values %>%
            bin(.,
                binmode = "breaks",
                breaks = c(min_val,0,max_val), #compare up and down-regulated genes separately.
                nBins = 2)
        }
    }

    # Define background
    if( use_bg_regions ){
      bg <- proms_gr[!proms_gr$gene_id %in% targets_df$gene_id]
    }else{
      bg <- NULL
    }

    # Binned Motif Enrichment
    set.seed(seed)
    enr_res <- calcBinnedMotifEnrR(seqs = target_seqs,
                                    bins = input_bins,
                                    pwmL = PWMs,
                                    min.score = score,
                                    maxFracN = 0.7,
                                    test = "binomial",
                                    background = "genome",
                                    genome = genome_bs,
                                    genome.regions = bg,
                                    genome.oversample = 10.0,
                                    BPPARAM = BiocParallel::MulticoreParam(4, RNGseed = seed),
                                    verbose = TRUE)

    results <- list("genes_removed"=genes_removed,
                    "target_promoters"=proms,
                    "target_seqs"=target_seqs,
                    "input_bins"=input_bins,
                    "enrichment_res"=enr_res,
                    "random_seed"=seed)
    return(results)
}
```


```{r}
create_motif_df <- function(motif_result_list, col_to_bin="gene_cluster"){
  
  purrr::map(names(assays(motif_result_list$enrichment_res)), function(assay_name){
    assay(motif_result_list$enrichment_res, assay_name) %>% 
      as.data.frame() %>% 
      rownames_to_column("motif") %>%
      pivot_longer(cols=-motif, 
                   names_to = col_to_bin, 
                   values_to = assay_name)
  }) %>% 
    purrr::reduce(left_join, by = c(col_to_bin,"motif")) %>% 
    mutate(p_value=1/10^negLog10P,
          adj.p_value=1/10^negLog10Padj) %>% 
    select(motif, all_of(col_to_bin),
           negLog10P:negLog10Padj, log2enr,
           p_value:adj.p_value,
           everything()) %>% 
    arrange(!! rlang::sym(col_to_bin)) %>%
    group_by(!! rlang::sym(col_to_bin)) %>% 
    arrange(desc(negLog10P), .by_group = T) %>% 
    ungroup()
}
```

```{r}
create_motif_enr_plots <- function(enr_df, 
                                   factor_column="gene_cluster_factor",
                                   motif_order=NULL,
                                   summary_stats, 
                                   expn_df_long,
                                   motif_colors,
                                   motif_tile_fill='red',
                                   PWMlist=AP2_PWMlist){
  # required columns in enr_df: motif_factor_groups, motif, plot_fill, label
  # required columns in summary_stats: analysis_factor_group, motif,mean, middle_val, ymin, ymax, stat
  # required columns in expn_df_long: analysis_factor_group, median_val_per_sample
  
  # Select the most significant TF motif hit. 
  top_motifs_per_clust <- enr_df %>% 
    group_by( !! as.name(factor_column) ) %>% 
    dplyr::slice(1) %>% 
    pull(motif, name=motif_factor_groups)
  
  top_motifs_per_clust_df <- enr_df %>% 
    group_by( !! as.name(factor_column) ) %>%  
    dplyr::slice(1) 
  
  # Plot the median expression per bio-replicate of all genes in time-course as a ribbon
  # add another ribbon of the TF's expression over time. 
  num <- length(unique(enr_df[[factor_column]]))
  time_ribbon <- ggplot(data=summary_stats,
         aes(x=analysis_factor_group,
             y=mean, 
             group=motif)) +
    geom_ribbon(aes(ymin=ymin, 
                    ymax=ymax,
                    group=motif, 
                    fill=motif),
                alpha=0.10) +
    geom_line(aes(x=analysis_factor_group, 
                  y=middle_val, 
                  group=motif), 
              linetype="dashed",
              linewidth=0.5, alpha=0.5) +
    geom_point(aes(color=analysis_factor_group, shape=stat),
               size = 3, alpha=0.7) +
    facet_wrap(~label, 
               scales='free_y',
               nrow=num, 
               strip.position = "left" ) +
    labs(y="log2 (CPM)", x="") +
    scale_color_viridis_d(option="D", begin=0.2, end=0.9) +
    scale_fill_manual(values=motif_colors) +
    theme_classic() +
    guides(color = guide_legend(title = "Time point")) +
    theme(legend.text = element_text(size=14),
          axis.text.x = element_text(size=18, angle=15, vjust=1, hjust=1,color="black"),
          axis.text.y = element_text(size=18, color="black"),
          axis.title.y = element_text(size=18, color="black"),
          axis.title.x = element_blank(),
          strip.text = element_text(size=24),
          plot.margin = margin(l=5, unit="mm"), 
          strip.placement = "outside", 
          panel.border = element_rect(fill = NA,colour = c("black")))
  
  # Plot the median expression per bio-replicate as a line plot
  time_cluster <- ggplot( data=expn_df_long,
         aes(x=analysis_factor_group,
             y=mean_val_per_sample, 
             color=analysis_factor_group, 
             group=analysis_factor_group)) +
    geom_line(aes(x=analysis_factor_group,
                  y=mean_val, 
                  group=!!sym(factor_column)),
              size=0.5, color="black",
              alpha=0.75,
              linetype="longdash") +
    geom_point(size=2, alpha=0.7) +
    scale_color_viridis_d(option="D", begin=0.2, end=0.9) +
    labs(y="log2 (CPM)", x="") +
    guides(color = guide_legend(title = "Time point")) +
    facet_wrap(~label,
               scales = 'fixed',
               nrow=num,
               strip.position = "left") +
    theme_classic() +
    theme(legend.text = element_text(size=14),
          axis.text.x = element_text(size=18,angle=15, vjust=1, hjust=1,color="black"),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size=18, color="black"),
          axis.title.y = element_text(size=18, color="black"),
          strip.text = element_text(size=20),
          plot.margin = margin(l=5, unit="mm"), 
          strip.placement = "outside", 
          panel.border = element_rect(fill = NA,colour = c("black")))
  
  #Reorder the factor levels and labels so that first factor level is at the top of thte plot
   labs <- enr_df %>% 
    arrange(desc(!!sym(factor_column))) %>% 
    pull(label) %>%
    unique()
  relevels <- levels(enr_df[[factor_column]]) %>% rev()
  enr_df <- enr_df %>%  
    mutate(y_axis = factor(!!sym(factor_column), levels=relevels)) 
  
  # Tile plot of the enrichment scores
  if(motif_tile_fill == 'red'){
      tile_fill_colors <- c(high="firebrick4",low=alpha("firebrick1",0.2), na.value = alpha("grey80", 0.1))
  }else if(motif_tile_fill == 'blue'){
      tile_fill_colors <- c(high="steelblue4",low=alpha("steelblue1",0.2), na.value = alpha("grey80", 0.1))
  }else{
      tile_fill_colors <- c(high=paste(motif_tile_fill,4),low=alpha(paste0(motif_tile_fill,1),0.2), na.value = alpha("grey80", 0.1))
  }
  motif_enr <- ggplot(enr_df, aes(x=motif, y=y_axis, fill=plot_fill)) +
      geom_tile(height=0.8, width=0.85) +
      scale_x_discrete(expand = c(0,0)) +
      scale_y_discrete(expand = c(0,0),
                       labels=labs) +
      scale_color_manual(values=c("NS"=NA,"p < 0.05"="black")) +
      scale_fill_gradient(high=tile_fill_colors['high'],
                          low=tile_fill_colors['low'],
                          na.value = tile_fill_colors['na.value']) +
      guides(fill = guide_colorbar(title = "log2 enrichment")) +
      labs(y="",x="") +
      theme_classic() +
      theme(axis.text = element_text(color="black"),
            legend.text = element_text(size=14),
            axis.text.y = element_text(size=20, angle=90,hjust = 0.5),
            axis.text.x = element_text(size=16, angle=45, vjust=1, hjust=1))
  
  clust_labs <- enr_df %>% pull(label, name=!!sym(factor_column))
  incl_hits <- enr_df %>% 
                group_by(motif_factor_groups) %>% 
                filter(!is.na(plot_fill)) %>% 
                ungroup()
  
  # Create seqLogo for each of the top hits per cluster 
  top_logos <- purrr::map(top_motifs_per_clust, function(name){
    mat <- Matrix(PWMlist[[name]])
    if(ncol(mat) >= 20){
      lims <- c(4,21)
      breaks <- seq(5,20,by=5)
    }else{
      lims <- c(1,ncol(mat)+2)
      breaks <- seq(1,ncol(mat)+2,by=5)
    }
    ggseqlogo(mat) +
      labs(title=name) +
      suppressMessages(scale_x_continuous(expand = c(0,0), 
                         limits = lims,
                         breaks= breaks)) +
      theme(plot.title = element_text(size=20,vjust = 0.5, hjust = 0.5),
            axis.text = element_text(size=20),
            plot.margin = margin())
  })
  
  plots <- list(time_ribbon=time_ribbon,
                time_cluster=time_cluster,
                motif_enr=motif_enr, 
                top_logos=top_logos)
  return(plots)
}
```

```{r}
clusterProfiler_bubble_plots <- function(df,pval_column="p.adjust",
                                         label="go_term_name",Filter=TRUE){
    
    column <- glue::glue("Neg_log10_{pval_column}")
    temp <- data.frame()
    input <- df %>% 
      dplyr::mutate( !!column := -log10(!!as.name(pval_column)))
    
    if(Filter){
      temp <- input %>%
        dplyr::filter( !!as.name(pval_column) < 0.05)
    }
    
    if (nrow(temp) == 0){
        temp <- input
    }

    n <- min(10,nrow(temp))
    temp <- temp %>%
      mutate(ID=factor(ID, levels = rev(ID))) %>%
      dplyr::slice(1:n)

    bubble_plot <- ggplot(temp,
                          aes_string(x=column,
                                     y="ID")) +
      geom_point(aes(size=setSize, fill=ID), shape=21,
                 alpha=0.8) +
      scale_radius(range=c(1,10)) +
      labs(x=glue::glue("-log10({pval_column})"), y="") +
      scale_y_discrete(expand = c(0.1,0)) +
      scale_x_continuous(expand = c(0.15,0)) +
      guides(fill="none",
             size=guide_legend(title = "Number of Genes in GO Term",
                               size = 1)) +
      theme(legend.position = "left",
            legend.key = element_rect(colour = NA, fill = NA),
            plot.title = element_text(hjust = 0),
            panel.grid.major.x = element_blank(),
            panel.grid.minor.x = element_blank(),
            panel.grid.major = element_line(color="black",size = 0.1),
            panel.background = element_rect(fill="white"),
            panel.border = element_rect(color="black", fill=NA)) 


    tab <-  ggplot(temp, aes_string(x = "0", y = "ID", 
                             label=label, 
                             color="ontology")) +
        geom_text(size = 4,
                  hjust = 0,
                  nudge_x = 0)  +
        scale_y_discrete(expand = c(0.1,0)) +
        scale_color_manual(values=c("Biological_Process"="tomato","Cellular_Component"="forestgreen",
                                    "Molecular_Function"="dodgerblue","N/A"="grey")) +
        # scale_x_continuous(expand = c(0,1)) +
        scale_x_continuous(expand = expansion(mult = c(0,0.01))) +
        theme_void()

    bubble_plot +
      tab +
      plot_layout(widths = c(1, 2.5), guides='collect') &
      theme(legend.position='left')
    
}

replace_long_strings <- function(string){
  t <- string
  if(grepl(",", t)){
    at <- stringr::str_locate_all(t,",")[[1]][,"end"]
  }else{
    at <- stringr::str_locate_all(t," ")[[1]][,"end"]
  }
  at <- at[at <= 40][length(at[at <= 40])]
  stringr::str_sub(t, start=at[1],end = at[1]) <- "\n"
  t
}
```

```{r}
# from BioNet R package - just modified the x,y axis ranges
plotLLSurface <- function (x, opt = NULL, main = "Log-Likelihood Surface", color.palette = heat.colors,  nlevels = 32) {
    if (is.null(opt)) 
        opt <- list(a = -1, l = -1)
    f <- function(l, a) {
        fbumLL(c(l, a), x)
    }
    v <- seq(0.001, 0.99, 0.05)
    z <- outer(v, v, Vectorize(f))
    Lines <- list(bquote(lambda == .(round(opt$l, 4))), bquote(a == 
        .(round(opt$a, 4))))
    filled.contour(v, v, z, nlevels = nlevels, color.palette = color.palette, 
        main = main, xlab = expression(lambda), ylab = "a", plot.axes = {
            axis(1, seq(0, 1, 0.01))
            axis(2, seq(0, 1, 0.01))
            abline(v = opt$l, lty = 2, col = "darkgray")
            abline(h = opt$a, lty = 2, col = "darkgray")
            hght <- strheight("Here")
            points(opt$l, opt$a, cex = 1.5)
            # text(opt$l, opt$a - (hght * 1.5 * seq(length(Lines))), 
            #     do.call(expression, Lines), adj = c(-0.2, -2.3), 
            #     cex = c(0.8, 0.8))
        })
}
```




# Reference Data

```{r}
GO_annots <- read.delim("resources/gene_annots/PF3D7_GO_Terms_PlasmoDB_v58.txt", sep="\t") %>% 
  janitor::clean_names()

head(GO_annots)
dim(GO_annots) #37,841    11
# table(GO_annots$ontology)
```

```{r}
orthoMCL <- read.delim("results/motif_analysis/homer/groups_OrthoMCL-6.14.txt", 
                       header = FALSE) %>% 
  separate(V1, into=c("ortholog_group","accession"), 
           sep=": ", remove=FALSE) %>% 
  dplyr::select(ortholog_group,accession,full_result=V1)

head(orthoMCL)
# dim(orthoMCL) #922,534      1
```

```{r}
update_ids <- read.csv("references/TF_families/TF_famlies_Insilico_GeneByLocusTag_Summary.csv")

pf_tf_list <- read.csv("references/TF_families/TF_families_supplemental_table.csv") %>% 
  left_join(., update_ids, by=c("gene.ID"="Input.ID")) %>% 
  janitor::clean_names() %>% 
  select(old_id=gene_id,
         gene_id=gene_id_2,
         ortholog_group,
         gene_name_or_symbol,
         annotation, matches("PFAM"),
         product_description,
         everything())

pf_tf_list
# write.csv(pf_tf_list, "references/TF_families/TF_families_updateIDs_supplemental_table.csv")
```

```{r}
edb <- ensembldb::EnsDb("species_genomes/GRCh38_GRCm39_Pf3D7v58.sqlite")
# edb

genome <- Biostrings::readDNAStringSet("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_Genome.fasta")

cleaned_levs <- seqlevels(genome) %>%
  str_split_fixed(.,"\\|", n=2) %>% 
  .[,1] %>% 
  gsub("\\s", "",.)

names(genome) <- cleaned_levs
# head(genome)
```

```{r}
#subset the promoters for those -1000bp upstream to TSS and those in plasmodium 
promoters_gr <- promoters(edb, upstream=1000, downstream = 0)
```

```{r}
# See https://doi.org/10.1371/journal.ppat.1001165.s019 for info used to map old IDs to new IDs on Plasmo DB
ap2_ids <- 
  read.csv("resources/gene_annots/ap2_old_ids_updated_PF3D7_2023.10.23.csv") %>%
  janitor::clean_names()

head(ap2_ids)
dim(ap2_ids) # 18 10
```


# Sample Manifest 

```{r}
pf_samples <- read.csv("samples/Kappe_s_2022.04_PF_LiverStage_ParasitesPaper_sample_manifest_7.25.2022.csv") %>% 
  set_rownames(.$sample_id_cat)

dim(pf_samples)
table(pf_samples$analysis_group)
```

# Expression Data

```{r}
cb_se <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_SummarizedExperiment.RDS")

cb_se
```

```{r}
cb_eset <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_expressionSet.RDS")

# dim(cb_eset)
```

```{r}
as.data.frame(colSums(assay(cb_se, "raw_counts"))) %>% 
  head()
```

```{r}
# design for time-series analysis with MaSigPro
design_input <- colData(cb_se) %>%
  as_tibble() %>%
  droplevels() %>% 
  dplyr::select(sample_id_cat, 
                analysis_group_id,
                infection_status,
                analysis_factor_group,
                Time=time_point) %>% 
  # remove Day2_Heps_S3_Pf (L3Pf2_S2xL3Pf2Mix_S3) due to very low read coverage
  filter(!grepl("Day2_Heps_S3_Pf", analysis_group_id)) %>% 
  group_by(Time) %>% 
  mutate(masigpro_id=paste("Pf",  Time, 
                           paste0("Rep", 1:n()), sep="_")) %>% 
  ungroup() %>% 
  mutate(Time=as.numeric(gsub("Day", "", Time)),
         Replicate=as.numeric(analysis_factor_group),
         Group=1) %>% 
         # Control=ifelse(Time==0, 1, 0),
         # Infection=ifelse(Time==0, 0, 1)) %>% 
  ungroup() %>% 
  arrange(Time) %>% 
  dplyr::select(masigpro_id, everything()) %>% 
  as.data.frame() %>% 
  set_rownames(.$masigpro_id)
  
head(design_input)
# design_input$infection_status

expression <- log2(assay(cb_se,"cpm")+1)[,design_input$sample_id_cat]
colnames(expression) <- rownames(design_input)
# head(expression[,1:7])
```

# maSigPro

https://www.bioconductor.org/packages/release/bioc/vignettes/maSigPro/inst/doc/maSigProUsersGuide.pdf

maSigPro follows a two steps regression strategy to find genes with significant temporal
expression changes and significant differences between experimental groups.

The data.abiotic object is a matrix with normalized gene expression data.

## All Days 
                      
```{r eval=FALSE}
pf_design <- make.design.matrix(dplyr::select(design_input,Time:Group),
                                degree = 2)

# Could use edgeR to define theta, rather than using the default value
# fit <- readRDS("results/time_course/masigpro/degree1/masigpro_fit_p.vector_D2456.RDS")
# tstep <- readRDS("results/time_course/masigpro/degree1/masigpro_tstep_D2456.RDS")

fit <- p.vector(expression, pf_design,
                counts=TRUE,
                Q = 0.05,
                MT.adjust = "BH",
                min.obs = 5)


tstep <- T.fit(fit,
               step.method = "backward")

# saveRDS(tstep,"results/time_course/masigpro/masigpro_tstep_D2456.RDS")
#  saveRDS(fit,"results/time_course/masigpro/masigpro_fit_p.vector_D2456.RDS")
```

> fit$i # returns the number of significant genes
> fit$alfa # gives p-value at the Q false discovery control level
> fit$SELEC # is a matrix with the significant genes and their expression values

T.fit() executes stepwise regression. The step.method can be ”backward” or ”forward” indicating whether the step procedure starts from the model with all or none variables.

For each selected gene the following values are given:

   * p-value of the regression ANOVA
   * R-squared of the model
   * p-value of the regression coefficients of the selected variables

You can use option ”each” to analyze the type of responses present in the significant genes: 

  * significant genes at the ”intercept” term will have a significant expression value at the starting time (beta0)
  * genes associated to the variable ”Time” will have a significant linear component, which can be induction or repression depending on the sign of their coefficient (betaTime)
  * genes associated to the variable ”Time2” will show a change in
the linear response that might be indicating transitory or saturation responses

```{r eval=FALSE}
## This will obtain all sigificant genes
sigs_all <- get.siggenes(tstep,
                     rsq = 0.6, 
                     vars = "all",
                     trat.repl.spots="none")

# length(sigs_all$summary)
# head(sigs_all$summary)
# quantile(sigs_all_df$`p-value`, na.rm=TRUE)
```

Data clustering can be done on the basis of either the original expression values, the regression coefficients, or the t.scores.

```{r eval=FALSE}
# identical values b/c only 1 group
bind_cols(sigs_all$sig.genes$coefficients,
          sigs_all$sig.genes$group.coeffs) %>% 
  rownames_to_column("gene_id") %>% 
  filter(beta0 != Group_beta0 | 
           betaTime != Group_beta1)
```


```{r fig.height=10, fig.width=10, eval=FALSE}
gene_clusters <- see.genes(data = sigs_all$sig.genes,
          show.fit = FALSE,
          dis = pf_design$dis,
          cluster.data = "sig.profiles", 
          k.mclust = TRUE,
          cluster.method = "Mclust",
          show.lines = TRUE)

# gene_clusters$cut
```

```{r}
sigs_all_df <- bind_cols(c(sigs_all$sig.genes[c("coefficients","sig.pvalues")], 
                           sigs_all$sig.genes["sig.profiles"])) %>% 
  rownames_to_column("gene_id") %>% 
  add_column(cluster = gene_clusters$cut, .before = 2) %>% 
  arrange(cluster)

head(sigs_all_df)
dim(sigs_all_df) #708 genes

# write.csv(sigs_all_df, "results/time_course/masigpro/maSigPro_significant_genes_degree2_time_course_day2-day6_excludeDay2rep3_results.csv", row.names = FALSE)
```

### Expression plots 

```{r}
identical(rownames(sigs_all$sig.genes$sig.profiles), names(gene_clusters$cut)) # TRUE
```

```{r eval=FALSE}
expn <- sigs_all$sig.genes$sig.profiles  %>% 
  rownames_to_column("gene_id") %>% 
  add_column(cluster=gene_clusters$cut,.before=2) %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to="log2_CPM") %>%
  left_join(., design_input, by="masigpro_id") %>%
  
  group_by(cluster) %>%
  mutate(label=paste0("cluster", cluster, "\n (", length(unique(gene_id)), " genes)")) %>%
  ungroup() %>%
  
  group_by(cluster, analysis_factor_group) %>%
  mutate(median_val=median(log2_CPM)) %>%
  ungroup() %>%
  
  group_by(cluster, analysis_factor_group, masigpro_id) %>%
  mutate(median_val_per_sample=median(log2_CPM)) %>%
  ungroup()

# expn
```

```{r fig.height=10, fig.width=14, eval=FALSE}
# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_expression_per_cluster.pdf", height = 10, width = 14)
ggplot(data=expn,
       aes(x=masigpro_id, y=log2_CPM, color=gene_id)) +
  geom_point() +
  geom_line(aes(group=gene_id)) +
  scale_color_viridis_d(option="D") +
  guides(color = "none") +
  theme_classic() +
  facet_wrap(~label, scales = "free_x") +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))
# dev.off()
```

```{r fig.height=10, fig.width=14, eval=FALSE}
# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_expression_boxplot.pdf", height = 10, width = 14)

ggplot(data=expn,
       aes(x=analysis_factor_group, y=log2_CPM, fill=analysis_factor_group)) +
  geom_point(aes(color=analysis_factor_group),
             position = position_jitterdodge(jitter.width = 2),
             alpha=0.25) +
  geom_boxplot(outlier.color = NA, alpha=0.3, color="grey70") +
  geom_line(aes(x=analysis_factor_group, y=median_val, group=cluster), 
            size=1.5, color="black", alpha=0.75,
            linetype="solid") +
  scale_color_viridis_d(option="D", begin=0.2, end=0.9) +
  scale_fill_viridis_d(option="D") +
  theme_classic() +
  facet_wrap(~label, scales = "free_x") +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))
# dev.off()
```

```{r fig.height=10, fig.width=14, eval=FALSE}
# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_expression_median_dotplot.pdf", height = 10, width = 14)

ggplot(data=expn,
       aes(x=analysis_factor_group, y=median_val_per_sample, 
           color=analysis_factor_group)) +
  geom_line(aes(x=analysis_factor_group, y=median_val, group=cluster),
            size=1, color="red", alpha=0.75,
            linetype="longdash") +
  geom_point(size=3, alpha=0.5) +
  scale_color_viridis_d(option="D", begin=0.2, end=0.9) +
  scale_fill_viridis_d(option="D") +
  theme_classic() +
  facet_wrap(~label, scales = "free_x") +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))

# dev.off()
```


```{r fig.height=20, fig.width=12, eval=FALSE}
gene_clust_df1 <- sigs_all_df %>%
  mutate(gene_cluster_factor = as.factor(cluster)) %>% 
  select(gene_id,
         cluster=gene_cluster_factor) %>% 
  column_to_rownames("gene_id")

n <- length(unique(gene_clust_df1$cluster))
print(n)
row_ha1 <- ComplexHeatmap::rowAnnotation(df = gene_clust_df1, 
                        col = list(cluster=ggpubr::get_palette("jco", n) %>% 
                                     set_names(levels(gene_clust_df1$cluster))))


hmap_time_course <- create_pf_heatmaps(TMMCPM = sigs_all$sig.genes$sig.profiles[rownames(gene_clust_df1),],
                                       sample_annots = design_input,
                                       gene_annots = rowData(cb_se),
                                       genelist = rownames(gene_clust_df1),
                                       split = gene_clust_df1$cluster)

# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_day2-day6_exclude_Day2_rep3_clusters_expression_heatmap.pdf", height = 10, width = 14)
hmap_time_course$heatmap + row_ha1
# dev.off()
```

## Day 456

```{r}
input_d456 <- dplyr::select(design_input,Time:Group) %>% 
  dplyr::filter(Time != 2)

pf_design_d456 <- make.design.matrix(input_d456,
                                degree = 2)
```

```{r eval=FALSE}
fit_d456  <- p.vector(expression[,rownames(input_d456)], 
                      pf_design_d456,
                      counts=TRUE,
                      Q = 0.05,
                      MT.adjust = "BH",
                      min.obs = 5)


tstep_d456  <- T.fit(fit_d456, step.method = "backward")


# saveRDS(fit_d456,"results/time_course/masigpro/masigpro_fit_p.vector_degree2_D456.RDS")
# saveRDS(tstep_d456,"results/time_course/masigpro/masigpro_tstep_degree2_D456.RDS")

##"Influence: 85 genes with influential data at slot influ.info. Model validation for these genes is recommended"
```

```{r}
tstep_d456 <- readRDS("results/time_course/masigpro/masigpro_tstep_degree2_D456.RDS")

## This will obtain all sigificant genes regardless the Rsquared value.
sigs_all_d456 <- get.siggenes(tstep_d456,
                     rsq = 0.6, 
                     vars = "all",
                     trat.repl.spots="none")


# sigs_all_d456$sig.genes$sig.pvalues %>% head()
# sigs_all_d456$sig.genes$Time$coefficients
# sigs_all_d456$sig.genes$Time2$coefficients
# sigs_all_d456$sig.genes$independ$coefficients
```


You can use option ”each” to analyze the type
of responses present in the significant genes: significant genes at the ”intercept” term will have
a significant expression value at the starting time; genes associated to the variable ”Time”
will have a significant linear component, which can be induction or repression depending on
the sign of their coefficient; genes associated to the variable ”Time2” will show a change in
the linear response that might be indicating transitory or saturation reponses,


```{r fig.width=10, fig.height=10}
gene_clusters_d456 <- see.genes(data = sigs_all_d456$sig.genes,
          dis = pf_design_d456$dis,
          cluster.data = "sig.profiles", 
          k.mclust = TRUE,
          cluster.method = "Mclust",
          show.fit = TRUE,
          show.lines = TRUE)

# str(gene_clusters_d456)
```

-Panel C putative:
Group for cluster trend:
Profile 1 Highly expressed in SPZ down in LS ( Cluster 1 & 2),
Profile 2 Highly expressed in SPZ reduced in LS (Cluster 3 and 8),
Profile 3 expressed in SPZ upregulated in LS (Cluster 4 & 6),
Profile 4 Increasing over LS development (Cluster 5 & Cluster 9),
Profile 5 upregulated in mature LS forms (Cluster 7).

```{r}
names(sigs_all_d456)
names(sigs_all_d456$sig.genes)
# sigs_all_d456$sig.genes$group.coeffs
# sigs_all_d456$sig.genes$coefficients
# sigs_all_d456$sig.genes$sig.pvalues
```

```{r}
# identical(names(gene_clusters_d456$cut),sigs_all_d456$summary)
sigs_all_df_d456 <- bind_cols(c(sigs_all_d456$sig.genes[c("group.coeffs","sig.pvalues")], 
                                sigs_all_d456$sig.genes["sig.profiles"])) %>% 
  mutate(gene_cluster=gene_clusters_d456$cut) %>% 
  rownames_to_column("gene_id") %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, 
                                    levels=c(1,2,3,8,4,6,5,9,7))) %>%
  arrange(gene_cluster_factor) %>% 
  
  group_by(gene_cluster_factor) %>% 
  mutate(gene_cluster_renamed=paste0("cluster",cur_group_id())) %>% 
  ungroup() %>% 
  
  dplyr::select(gene_id,matches("gene_cluster"), everything())


# dim(sigs_all_df_d456) #1510   26
head(sigs_all_df_d456)


# table(sigs_all_df_d456$gene_cluster_renamed,
#       sigs_all_df_d456$gene_cluster_factor)


# write.csv(sigs_all_df_d456, "results/time_course/masigpro/masigpro_significant_genes_degree2_d456_results.csv", row.names = TRUE)
```

### Expression plots

```{r}
expn_d456 <- sigs_all_d456$sig.genes$sig.profiles  %>% 
  rownames_to_column("gene_id") %>% 
  # add_column(cluster=gene_clusters_d456$cut, .before=2) %>% 
  pivot_longer(cols=matches("Pf"), 
               names_to="masigpro_id",
               values_to="log2_CPM") %>%
  left_join(design_input, .,
            by="masigpro_id") %>% 
  left_join(select(sigs_all_df_d456,gene_id,matches("gene_cluster")), .,
            by="gene_id") %>% 
  group_by(gene_cluster_factor) %>%
  mutate(label=paste0("cluster",gene_cluster_factor, "\n (", length(unique(gene_id)), " genes)") %>% 
           factor(., levels=unique(.))) %>%
  ungroup() %>%
  group_by(gene_cluster_factor, analysis_factor_group) %>%
  mutate(median_val=median(log2_CPM),
         mean_val=mean(log2_CPM)) %>%
  ungroup() %>%
  group_by(gene_cluster_factor, analysis_factor_group, masigpro_id) %>%
  mutate(median_val_per_sample=median(log2_CPM),
         mean_val_per_sample=mean(log2_CPM)) %>%
  ungroup()

head(expn_d456)
# expn_d456$gene_cluster_factor
```

```{r fig.height=4, fig.width=12}
temp <- expn_d456 %>% 
  dplyr::filter(grepl("PF3D7_1413600", gene_id)) %>% 
  group_by(analysis_factor_group) %>% 
  mutate(mean_expn=mean(log2_CPM)) %>% 
  ungroup()

p1 <- ggplot(temp, aes(x=Time,y=log2_CPM, color=analysis_factor_group)) +
  geom_point() +
    geom_line(aes(x=Time, y=mean_expn, group=gene_id))  +
  theme_classic()

p2 <- ggplot(temp, aes(x=Time^2,y=log2_CPM, color=analysis_factor_group)) +
  geom_point() +
  geom_line(aes(x=Time^2, y=mean_expn, group=gene_id)) +
  theme_classic()

p1 + p2
```

```{r fig.height=10, fig.width=7}
# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_expression_per_cluster_d456.pdf", height = 10, width = 14)

# svglite("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_expression_per_cluster_d456.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 12, width = 7)

ggplot(data=expn_d456,
       aes(x=masigpro_id, y=log2_CPM, color=gene_id)) +
  geom_point() +
  geom_line(aes(group=gene_id)) +
  scale_color_viridis_d(option="D") +
  guides(color = "none") +
  theme_classic() +
  facet_wrap(~label, scales = "free_x",ncol = 2) +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))

# dev.off()
```

```{r fig.height=10, fig.width=5}
# png("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_expression_boxplot_d456.png", height = 14, width = 10, units = "in", res = 150)
sample_labels <- c("sporozorite"="sporozoites", 
                   "Day2_Heps_Pf"="Day2_Pf_LS",
                   "Day4_Heps_Pf"="Day4_Pf_LS",
                   "Day5_Heps_Pf"="Day5_Pf_LS",
                   "Day6_Heps_Pf"="Day6_Pf_LS" )

boxplots <- ggplot(data=expn_d456,
       aes(x=analysis_factor_group, y=log2_CPM, fill=analysis_factor_group)) +
  geom_point(aes(color=analysis_factor_group),
             position = position_jitterdodge(jitter.width = 2),
             alpha=0.25) +
  geom_boxplot(outlier.color = NA, alpha=0.3, color="grey70") +
  geom_line(aes(x=analysis_factor_group, y=median_val, group=gene_cluster_factor), 
            size=1.5, color="black", alpha=0.75,
            linetype="solid") +
  scale_x_discrete(labels = sample_labels) +
  scale_color_viridis_d(option="D", begin=0.2, end=0.9, labels = sample_labels) +
  scale_fill_viridis_d(option="D", labels = sample_labels) +
  theme_classic() +
  facet_wrap(~label, scales = "free_x", ncol=2) +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))

# svglite("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_expression_boxplot_d456.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 14, width = 7)
boxplots
# dev.off()
```

```{r fig.height=10, fig.width=5}
# pdf("figures/combat_seq/kappe_s_maSigPro_time_course_expression_median_dotplot_d456.pdf", height = 10, width = 14)

# svglite("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_expression_median_dotplot_d456.svg",fix_text_size=FALSE,
#         system_fonts = list(sans = "Open Sans"), height = 14, width = 7)

ggplot(data=expn_d456,
       aes(x=analysis_factor_group, y=median_val_per_sample, 
           color=analysis_factor_group)) +
  geom_line(aes(x=analysis_factor_group, y=median_val, group=gene_cluster_factor),
            size=1, color="red", alpha=0.75,
            linetype="longdash") +
  geom_point(size=3, alpha=0.5) +
  scale_color_viridis_d(option="D", begin=0.2, end=0.9) +
  scale_fill_viridis_d(option="D") +
  theme_classic() +
  facet_wrap(~label, scales = "free_x", ncol=2) +
  theme(axis.text.x = element_text(angle=35, vjust=1, hjust=1,color="black"),
        axis.title.x = element_blank(),
        plot.margin = margin(l=5, unit="mm"))

# dev.off()
```

```{r}
cc <- viridis::viridis(4, begin=0.2, end=0.9) %>% 
  set_names(sample_labels[-2])

cc2 <- ggpubr::get_palette("npg", k = 4) %>% 
  set_names(sample_labels[-2])

# barplot(rep(1,4), col = cc, names.arg = names(cc))
barplot(rep(1,4), col = cc2, names.arg = names(cc2))
```

```{r fig.height=20, fig.width=12}
# idx <- which(!duplicated(ordered_genes, fromLast=TRUE))
# quickPheatmap(expn = sigs_all_d456$sig.genes$sig.profiles[names(ordered_genes),], 
#               sample_data = design_input[,c("analysis_factor_group","infection_status")],
#               gene_index=idx)
```


```{r eval=FALSE}
gene_clust_df <- sigs_all_df_d456 %>% 
  select(gene_id,cluster=gene_cluster_factor) %>% 
  column_to_rownames("gene_id")

row_ha <- ComplexHeatmap::rowAnnotation(df = gene_clust_df, 
                        col = list(cluster=ggpubr::get_palette("jco", 9) %>% 
                                     set_names(levels(gene_clust_df$cluster))))

expn_mat <- sigs_all_d456$sig.genes$sig.profiles[rownames(gene_clust_df),]
colnames(expn_mat) <- gsub("Pf_Day0","sporozoites",colnames(expn_mat)) %>% 
  gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2", .)

new_pal <- circlize::colorRamp2(c(-4,-2, 0, 2, 4),
                      c("deepskyblue3", "deepskyblue","white", "red", "red3"),
                      space="sRGB")



sample_anno <- design_input %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), levels= c("sporozoites",
                                                                                      "Day2_Pf_LS",
                                                                                      "Day4_Pf_LS",
                                                                                      "Day5_Pf_LS",
                                                                                      "Day6_Pf_LS"))) %>% 
  set_rownames(.$masigpro_id) %>% 
  as.data.frame()

# head(expn_mat)
# head(sample_anno)
# all(colnames(expn_mat) %in% rownames(sample_anno))

hmap_time_course <- 
  create_pf_heatmaps(TMMCPM = expn_mat,
                     sample_annots = sample_anno,
                     gene_annots = rowData(cb_se),
                     genelist = rownames(gene_clust_df),
                     split = gene_clust_df)
```


```{r fig.height=10, fig.width=7, eval=FALSE}
# pdf("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_heatmap_d456.pdf",    height = 10, width = 10)
hmap_time_course$heatmap + row_ha
# dev.off()
```

```{r eval=FALSE}
plot(hmap_time_course$dends$samp.c1)
 
sample_order <- expn_mat %>% 
  # dplyr::select(Pf_Day0_Rep1:Pf_Day0_Rep5, 
  #               Pf_Day4_Rep2,Pf_Day4_Rep1,Pf_Day4_Rep3,
  #               Pf_Day5_Rep2,Pf_Day5_Rep3,Pf_Day5_Rep1,
  #               Pf_Day6_Rep1, Pf_Day6_Rep2, Pf_Day6_Rep3)  %>% 
    dplyr::select(sporozoites_Rep1:sporozoites_Rep5, 
                Day4_Pf_LS_Rep2,Day4_Pf_LS_Rep1,Day4_Pf_LS_Rep3,
                Day5_Pf_LS_Rep2,Day5_Pf_LS_Rep3,Day5_Pf_LS_Rep1,
                Day6_Pf_LS_Rep1, Day6_Pf_LS_Rep2, Day6_Pf_LS_Rep3)  %>% 
  colnames()

# sample_order

hmap_time_course2 <- 
  create_pf_heatmaps(TMMCPM = expn_mat,
                     order_ids = sample_order,
                     sample_annots = sample_anno,
                     gene_annots = rowData(cb_se),
                     genelist = rownames(gene_clust_df), #names(ordered_genes),
                     split = gene_clust_df)
```

```{r fig.height=12, fig.width=10, eval=FALSE}

# png("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_re-order_heatmap_d456.png", height = 10, width = 10, units = "in", res = 150)

# svglite("figures/combat_seq/masigpro/kappe_s_maSigPro_time_course_re-order_heatmap_d456.svg", system_fonts = list(sans = "Open Sans"), fix_text_size=FALSE,  height = 10, width = 10)

hmap_time_course2$heatmap + row_ha

# dev.off()
```


## Example Data Inputs

```{r}
data(data.abiotic)
data(edesign.abiotic)
# View(edesign.abiotic)

design <- make.design.matrix(edesign.abiotic, degree = 2)
# design

quantile(data.abiotic, na.rm=TRUE)

str(design)
# design$groups.vector

#For a single series time course
Time <- rep(c(1,5,10,24), each = 3)
Replicates <- rep(c(1:4), each = 3)
Group <- rep(1,12)
ss.edesign <- cbind(Time,Replicates,Group)
rownames(ss.edesign) <- paste("Array", c(1:12), sep = "")
# ss.edesign

# data(NBdata)
# quantile(NBdata)
# data(NBdesign)
# View(NBdesign)
```

```{r eval=FALSE}
IDs <- rbind(paste("feature", c(1:300), sep = ""),
       rep(paste("gene", c(1:150), sep = ""), each = 2))

head(IDs[,1:5])
# dim(IDs)


tc.GENE <- function(n, r,
             var11 = 0.01, var12 = 0.01,var13 = 0.01,
             var21 = 0.01, var22 = 0.01, var23 =0.01,
             var31 = 0.01, var32 = 0.01, var33 = 0.01,
             var41 = 0.01, var42 = 0.01, var43 = 0.01,
             a1 = 0, a2 = 0, a3 = 0, a4 = 0,
             b1 = 0, b2 = 0, b3 = 0, b4 = 0,
             c1 = 0, c2 = 0, c3 = 0, c4 = 0)
{

  tc.dat <- NULL
  for (i in 1:n) {
    Ctl <- c(rnorm(r, a1, var11), rnorm(r, b1, var12), rnorm(r, c1, var13))  # Ctl group
    Tr1 <- c(rnorm(r, a2, var21), rnorm(r, b2, var22), rnorm(r, c2, var23))  # Tr1 group
    Tr2 <- c(rnorm(r, a3, var31), rnorm(r, b3, var32), rnorm(r, c3, var33))  # Tr2 group
    Tr3 <- c(rnorm(r, a4, var41), rnorm(r, b4, var42), rnorm(r, c4, var43))  # Tr3 group
    gene <- c(Ctl, Tr1, Tr2, Tr3)
    tc.dat <- rbind(tc.dat, gene)
  }
  tc.dat
}

# tc.GENE

## Create 270 flat profiles
flat <- tc.GENE(n = 270, r = 3)
## Create 10 genes with profile differences between Ctl and Tr1 groups
twodiff <- tc.GENE (n = 10, r = 3, b2 = 0.5, c2 = 1.3)
## Create 10 genes with profile differences between Ctl, Tr2, and Tr3 groups
threediff <- tc.GENE(n = 10, r = 3, b3 = 0.8, c3 = -1, a4 = -0.1, b4 = -0.8, c4 = -1.2)
## Create 10 genes with profile differences between Ctl and Tr2 and different variance
vardiff <- tc.GENE(n = 10, r = 3, a3 = 0.7, b3 = 1, c3 = 1.2, var32 = 0.03, var33 = 0.03)
## Create dataset
tc.DATA <- rbind(flat, twodiff, threediff, vardiff)
rownames(tc.DATA) <- paste("feature", c(1:300), sep = "")
colnames(tc.DATA) <- paste("Array", c(1:36), sep = "")
tc.DATA [sample(c(1:(300*36)), 300)] <- NA  # introduce missing values


# head(tc.DATA[,1:10])
# head(IDs[,1:10])
```


# GO Terms

https://plasmodb.org/plasmo/app/search/transcript/GeneByLocusTag

http://current.geneontology.org/products/pages/downloads.html

-Panel C putative:
Group for cluster trend:
Profile 1 Highly expressed in SPZ down in LS ( Cluster 1 & 2),
Profile 2 Highly expressed in SPZ reduced in LS (Cluster 3 and 8),
Profile 3 expressed in SPZ upregulated in LS (Cluster 4 & 6),
Profile 4 Increasing over LS development (Cluster 5 & Cluster 9),
Profile 5 upregulated in mature LS forms (Cluster 7).


* Evaluate if the new GO terms for this profiles increase significance and data interpretation. 

```{r}
day2_de <- read.delim("results/DEGs/kappe_s_Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_DEGs.csv",
                      sep=",",comment.char = "#")
head(day2_de)
# dim(day2_de) #1978   20

time_course <- read.csv("results/time_course/masigpro/masigpro_significant_genes_degree2_d456_results.csv") %>% 
  select(-X) %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  dplyr::select(gene_id, matches("gene_cluster"), everything()) %>% 
  arrange(gene_cluster)


gene_clusts_dfs <- purrr::map(unique(time_course$gene_cluster), function(clust){
  time_course %>%
    dplyr::filter(gene_cluster==clust) 
})
names(gene_clusts_dfs) <- paste0("cluster", unique(time_course$gene_cluster))


gene_clusts_combined_dfs <- purrr::map(unique(time_course$combined_gene_clusters), function(clust){
  time_course %>%
    dplyr::filter(combined_gene_clusters==clust) 
}) %>% 
  set_names(unique(time_course$combined_gene_clusters))

# sapply(gene_clusts_dfs, dim)
# head(gene_clusts_dfs$cluster1)
# lapply(gene_clusts_dfs, function(x) quantile(x$betaTime))

table(time_course$gene_cluster_factor,
      time_course$combined_gene_clusters)
```

```{r}
GO_descriptions <- GO_annots %>% 
  dplyr::select(go_id, ontology, go_term_name) %>% 
  distinct() %>% 
  mutate_at(vars(ontology), ~gsub("\\s", "_", .))

dim(GO_descriptions)
# any(duplicated(GO_descriptions$go_id))
GO_descriptions$ontology %>%  unique()
```

```{r}
GO_annots_filtered <- GO_annots %>% 
  group_by(ontology, go_id) %>% 
  mutate(N=n()) %>% 
  ungroup() %>% 
  arrange(go_id) %>% 
  filter(N>=5, N<=600) %>% 
  select(gene_id, ontology, go_id, N)

GO_annots_filtered
# write.csv(GO_annots_filtered,"resources/gene_annots/PF3D7_GO_Terms_PlasmoDB_filteredSetSizes_v58.txt", row.names = F)

GO_annots_filtered %>% 
  select(ontology, go_id) %>% 
  distinct() %>% 
  group_by(ontology) %>% 
  dplyr::count()
```

```{r}
GO_annots_list <- purrr::map(unique(GO_annots_filtered$go_id) %>% set_names(., value = .), function(x) { 
  GO_annots_filtered %>% 
    filter(go_id == x)  %>% 
    pull(gene_id)
})

# head(GO_annots_list)
```

## ClusterProfiler 

```{r}
suppressPackageStartupMessages(library(clusterProfiler))
```

```{r}
# threshold for -log10(p) values 
# this is p ~= 0.06
threshold <- 1.2
```

### Day 2 

```{r}
day2_go_clean <- read.csv("results/GO_enrichment/day2/day2_vs_spz_ClusterProfiler_AllOntology.csv") %>% 
  mutate(Direction=ifelse(NES>0, "Enriched_Day2","Enriched_SPZ")) %>%
  mutate(labels=case_when(
    grepl(",", go_term_name) ~ gsub("^(.+),.+","\\1", go_term_name),
    grepl("modulation", go_term_name) ~ gsub("^(.+) agg.+","\\1", go_term_name),
    TRUE ~ go_term_name
  )) %>%
  dplyr::select(ID,ontology:go_term_name,
                Direction,
                pvalue, qvalues,
                everything()) %>%
  dplyr::arrange(qvalues)

head(day2_go_clean)
```

```{r eval=FALSE}
d2_gsea_input <- day2_de %>% 
  pull(logFC, name=gene_id)


cats <- unique(GO_annots$ontology)
D2_GSEA_objs <- purrr::map(cats, function(x){
  
    input_go_cats <- GO_annots %>% 
      dplyr::filter(ontology==x) %>%
      dplyr::select(ID=go_id, gene=gene_id)
    
    GSEA_res <- GSEA(gene = d2_gsea_input,
                      TERM2GENE = input_go_cats,
                      minGSSize = 5,
                      maxGSSize = 600)
})
names(D2_GSEA_objs) <- cats

day2_go_clean <- purrr::map_dfr(D2_GSEA_objs, function(x) x@result) %>% 
  left_join(., GO_descriptions, by=c("ID"="go_id")) %>%
  mutate(Direction=ifelse(NES>0, "Enriched_Day2","Enriched_SPZ")) %>% 
  mutate(labels=case_when(
    grepl(",", go_term_name) ~ gsub("^(.+),.+","\\1", go_term_name),
    grepl("modulation", go_term_name) ~ gsub("^(.+) agg.+","\\1", go_term_name),
    TRUE ~ go_term_name
  )) %>% 
  dplyr::select(ID,ontology:go_term_name,
                Direction,
                pvalue, qvalues,
                everything()) %>% 
  dplyr::arrange(qvalues)

# day2_go_clean

# day2_go_clean %>% filter(ontology == "Biological_Process")
# table(day2_go_clean$Direction)
# write.csv(day2_go_clean, results/GO_enrichment/day2/day2_vs_spz_ClusterProfiler_AllOntology.csv", row.names = FALSE)
```

```{r}
chk <- day2_go_clean %>% 
  filter(Direction == "Enriched_Day2") %>%
  filter(p.adjust < 0.05) %>% 
  mutate(neg_log10_p=-log10(pvalue))

quantile(chk$p.adjust)
table(chk$neg_log10_p > threshold)
```

```{r}
lab1 <- day2_go_clean %>% 
  filter(ID=="GO:0006412") %>% 
  mutate(label=paste(ID,go_term_name)) %>% 
  pull(label)

# svglite("figures/combat_seq/barcode_plots/day2_vs_spz_ClusterProfiler_GO0006412_barcode.svg",fix_text_size=FALSE, height = 5, width = 10, system_fonts = list(sans = "Open Sans"))
bc1 <- gseaplot(D2_GSEA_objs$`Biological Process`, geneSetID = "GO:0006412", by="runningScore") +
  labs(title=lab1)
bc1
# dev.off()
 
lab2 <- day2_go_clean %>% 
  filter(ID=="GO:0044409") %>% 
  mutate(label=paste(ID,go_term_name)) %>% 
  pull(label)

# svglite("figures/combat_seq/barcode_plots/day2_vs_spz_ClusterProfiler_GO0044409_barcode.svg", fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 5, width = 10)
bc2 <- gseaplot(D2_GSEA_objs$`Biological Process`, geneSetID = "GO:0044409", by="runningScore") +
  labs(title=lab2)
bc2
# dev.off()
```

```{r}
layout <- "
BCCC
DDD#
"

# Note - the Filter=TRUE filters the supplied pval_column < 0.05
d2_up <- day2_go_clean %>% 
  filter(Direction == "Enriched_Day2") %>% 
  clusterProfiler_bubble_plots(.,
                               label = "labels",
                               pval_column="p.adjust",
                               # label="go_term_name",
                               Filter=TRUE)  +
  labs(title = "GO Terms Enriched in Day 2")

d2_down <- day2_go_clean %>%
  filter(Direction == "Enriched_SPZ") %>%
  clusterProfiler_bubble_plots(.,
                               label = "labels",
                               pval_column="p.adjust",
                               # label="go_term_name",
                               Filter=TRUE) +
  labs(title = "GO Terms Enriched in Sporozoite")

# svglite("figures/combat_seq/bubble_plots/day2_vs_spz_ClusterProfiler_bubblePlot_barcode_Day2.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 7, width = 11.5)
d2_up + bc1 + 
  plot_layout(design = layout) 
# dev.off()

# svglite("figures/combat_seq/bubble_plots/day2_vs_spz_ClusterProfiler_bubblePlot_barcode_SPZ.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 7, width = 11.5)
d2_down + bc2 + 
  plot_layout(design = layout)
# dev.off()
```

```{r}
purrr::map(unique(day2_go_clean$ontology), function(ont, df){
  df <- day2_go_clean %>% 
    filter(ontology==ont)
  
  # pdf(paste0("figures/combat_seq/bubble_plots/day2_vs_spz_ClusterProfiler_",ont,"_bubblePlot.pdf"), height = 5, width = 8)
  print(clusterProfiler_bubble_plots(df))
  # dev.off()
})
```

### Time-course clusters 

```{r eval=FALSE}
cluster_go <- purrr::map(names(gene_clusts_dfs), function(clust){
  
  gsea_input <- gene_clusts_dfs[[clust]] %>% 
    pull(gene_id)
  
  cats <- unique(GO_annots$ontology)
  go_enr_df <- purrr::map_dfr(cats, function(x){
      input_go_cats <- GO_annots %>% 
        dplyr::filter(ontology==x) %>%
        dplyr::select(ID=go_id, gene=gene_id)
      go_enr <- enricher(gene = gsea_input,
                        TERM2GENE = input_go_cats,
                        minGSSize = 5,
                        maxGSSize = 600)
      if(!is.null(go_enr)){
        go_enr@result
      }
  })

  go_enr_df <- go_enr_df %>%
    mutate(cluster_number=clust) %>% 
    left_join(., GO_descriptions, by=c("ID"="go_id")) %>%
    dplyr::select(ID,cluster_number, ontology:go_term_name,setSize=Count,
           everything()) %>% 
    dplyr::select(ID:go_term_name,pvalue,qvalue,
                everything()) %>% 
    dplyr::arrange(pvalue)

  
  return(go_enr_df)
})

names(cluster_go) <- names(gene_clusts_dfs)

cluster_go_dfs <- bind_rows(cluster_go) 
head(cluster_go_dfs, n=20)


table(cluster_go_dfs$cluster_number)
# write.csv(cluster_go_dfs, results/GO_enrichment/time_course_clusters_clusterProfiler_allOntology.csv", row.names = FALSE)
```

```{r}
cluster_go_dfs <-  read.csv("results/GO_enrichment/timecourse/time_course_clusters_clusterProfiler_allOntology.csv") %>% 
  filter(qvalue < 0.2) %>% 
  mutate(len=nchar(go_term_name)) %>% 
  mutate(labels=case_when(
    grepl("RNA polymerase", go_term_name) ~ gsub("RNA polymerase","RNApol",go_term_name),
    grepl(".apurinic or apyrimidinic site.", go_term_name) ~ gsub(".apurinic or apyrimidinic site.","", go_term_name),
    grepl(",", go_term_name) ~ gsub("^(.+),.+","\\1",go_term_name),
    grepl("polyadenylation specificity factor complex", go_term_name) ~ gsub("polyadenylation specificity factor complex","CPSF", go_term_name),
    grepl("calcium", go_term_name) ~ gsub("calcium-","Ca2+",go_term_name),
    grepl("cellular protein", go_term_name) ~ gsub("cellular protein","", go_term_name),
    grepl("-adenosylmethionine", go_term_name) ~ gsub("-adenosylmethionine","AMe",go_term_name),
    TRUE ~ go_term_name
  )) %>% 
  dplyr::arrange(desc(len))

dim(cluster_go_dfs) # 175 GO terms
```


```{r}
cluster_go_dfs %>% 
  group_by(cluster_number) %>% 
  mutate(FDR_sig_result=p.adjust < 0.2) %>% 
  group_by(FDR_sig_result, .add=TRUE) %>% 
  dplyr::count()  %>% 
  ungroup() %>% 
  pivot_wider(names_from = FDR_sig_result,
              values_from = n) %>% 
  rename_at(c("FALSE","TRUE"), ~c("GO_terms_qvalue>0.2","GO_terms_qvalue<0.2"))  %>% 
  left_join(., data.frame(Genes_in_Cluster=gene_clusts_dfs %>% 
                            sapply(., nrow)) %>% 
              rownames_to_column("cluster_number"), 
            by="cluster_number") 
 
# write.csv(.,results/GO_enrichment/time_course_clusters_clusterProfiler_SigResults_Summary.csv", row.names = FALSE)
```


```{r}
p <- purrr::map(unique(cluster_go_dfs$cluster_number), function(clust){
  df <- cluster_go_dfs %>% 
    filter(cluster_number == clust) %>% 
    filter(ontology != "N/A", qvalue < 0.2) %>% 
    mutate(setSize=as.integer(setSize)) %>% 
    dplyr::arrange(pvalue)
  
    # svglite(paste0("figures/combat_seq/bubble_plots/",clust,"_ClusterProfiler_bubblePlot.svg"),fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 5, width = 8)
    print(clusterProfiler_bubble_plots(df,pval_column = "pvalue",label = "labels", Filter=FALSE) +
      plot_annotation(title = paste0("Timecourse Cluster ",clust)))
    # dev.off()
})
```

```{r}
# quick reminder on -log10(p) threshold
print(threshold)
```

```{r}
top_paths <- cluster_go_dfs %>% 
   mutate(neg_log10_q=-log10(qvalue),
          neg_log10_p=-log10(pvalue)) %>% 
  group_by(cluster_number) %>% 
  dplyr::filter(neg_log10_p > threshold) %>% 
  # dplyr::filter( qvalue < 0.2 ) %>% 
  # filter(p.adjust < 0.05 | pvalue < 0.05) %>% 
  dplyr::arrange(desc(neg_log10_q), .by_group = TRUE) %>% 
  dplyr::slice(1:10) %>% 
  ungroup() %>% 
  dplyr::filter(ontology != "N/A") %>% 
  mutate(ID=factor(ID, levels=unique(ID)),
         text_color=case_when(
           ontology=="Cellular_Component" ~ "green4",
           ontology=="Biological_Process" ~ "tomato2",
           TRUE ~ "dodgerblue"
         ), 
         setSize_clean=ifelse(setSize > 20, 20, setSize))

dim(top_paths) #  65 18
quantile(top_paths$p.adjust)
quantile(top_paths$pvalue)
table(top_paths$neg_log10_p > threshold)
```

```{r fig.width=12, fig.height=6}
 bubble_plot <- ggplot(top_paths, aes(x=neg_log10_p, y=ID)) +
      geom_point(aes(size=setSize_clean, fill=ID), shape=21,
                 alpha=0.8) +
      scale_radius(range=c(2,10), breaks = c(seq(0,20,by=4)),
                   labels=seq(0,20,by=4) %>% gsub("20",">= 20", .)) +
      labs(x="-log10(p-value)", y="") +
      scale_y_discrete(expand = c(0.05,0.05)) +
      scale_x_continuous(expand = c(0.05,0.05)) +
  
      theme(panel.grid.major.x = element_blank(),
            axis.text.y = element_text(color=top_paths$text_color),
            panel.grid.minor.x = element_blank(),
            panel.grid.major = element_line(color="black",size = 0.1),
            panel.background = element_rect(fill="white"),
            panel.border = element_rect(color="black", fill=NA)) +
      guides(fill="none",
             size=guide_legend(title = "Number of Genes",
                               size = 1)
             ) +
      theme(legend.position = "left",
            legend.key = element_rect(colour = NA, fill = NA))


# svglite("figures/combat_seq/bubble_plots/time_course_facet_plot_allOntology.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), width = 15, height = 7)
bubble_plot +
  facet_wrap(~cluster_number, scales = 'free', ncol = 5)
# dev.off()
```


### Time point profiles

```{r}
cluster_go_combined_dfs <- read.csv("results/GO_enrichment/timecourse/time_course_clusters_combinedProfiles_clusterProfiler_allOntology.csv") %>% 
    mutate(labels=case_when(
    grepl("RNA polymerase", go_term_name) ~ gsub("RNA polymerase","RNApol",go_term_name),
    grepl(".apurinic or apyrimidinic site.", go_term_name) ~ gsub(".apurinic or apyrimidinic site.","", go_term_name),
    grepl(",", go_term_name) ~ gsub("^(.+),.+","\\1",go_term_name),
    grepl("polyadenylation specificity factor complex", go_term_name) ~ gsub("polyadenylation specificity factor complex","CPSF", go_term_name),
    grepl("calcium", go_term_name) ~ gsub("calcium-","Ca2+",go_term_name),
    grepl("cellular protein", go_term_name) ~ gsub("cellular protein","", go_term_name),
    grepl("-adenosylmethionine", go_term_name) ~ gsub("-adenosylmethionine","AMe",go_term_name),
    grepl("complex$", go_term_name) ~ gsub("complex$", "", go_term_name),
    TRUE ~ go_term_name
  )) 

cluster_go_combined_dfs
```

```{r eval=FALSE}
cluster_go_combined <- purrr::map(names(gene_clusts_combined_dfs), function(clust){
  
  gsea_input <- gene_clusts_combined_dfs[[clust]] %>% 
    pull(gene_id)
  
  cats <- unique(GO_annots$ontology)
  go_enr_df <- purrr::map_dfr(cats, function(x){
      input_go_cats <- GO_annots %>% 
        dplyr::filter(ontology==x) %>%
        dplyr::select(ID=go_id, gene=gene_id)
      go_enr <- enricher(gene = gsea_input,
                        TERM2GENE = input_go_cats,
                        minGSSize = 5,
                        maxGSSize = 600)
      if(!is.null(go_enr)){
        go_enr@result
      }
  })

  go_enr_df <- go_enr_df %>%
    mutate(cluster_number=clust) %>% 
    left_join(., GO_descriptions, by=c("ID"="go_id")) %>%
    dplyr::select(ID,cluster_number, ontology:go_term_name,setSize=Count,
           everything()) %>% 
    dplyr::select(ID:go_term_name,pvalue,qvalue,
                everything()) %>% 
    dplyr::arrange(pvalue)

  
  return(go_enr_df)
})

names(cluster_go_combined) <- names(gene_clusts_combined_dfs)

cluster_go_combined_dfs <- bind_rows(cluster_go_combined) %>% 
  filter(qvalue < 0.2) %>% 
  mutate(len=nchar(go_term_name)) %>% 
  mutate(labels=case_when(
    grepl("RNA polymerase", go_term_name) ~ gsub("RNA polymerase","RNApol",go_term_name),
    grepl(".apurinic or apyrimidinic site.", go_term_name) ~ gsub(".apurinic or apyrimidinic site.","", go_term_name),
    grepl(",", go_term_name) ~ gsub("^(.+),.+","\\1",go_term_name),
    grepl("polyadenylation specificity factor complex", go_term_name) ~ gsub("polyadenylation specificity factor complex","CPSF", go_term_name),
    grepl("calcium", go_term_name) ~ gsub("calcium-","Ca2+",go_term_name),
    grepl("cellular protein", go_term_name) ~ gsub("cellular protein","", go_term_name),
    grepl("-adenosylmethionine", go_term_name) ~ gsub("-adenosylmethionine","AMe",go_term_name),
    grepl("complex$", go_term_name) ~ gsub("complex$", "", go_term_name)
    TRUE ~ go_term_name
  )) %>% 
  dplyr::arrange(desc(len))

# head(cluster_go_combined_dfs)

# write.csv(cluster_go_combined_dfs, results/GO_enrichment/timecourse/time_course_clusters_combinedProfiles_clusterProfiler_allOntology.csv", row.names = FALSE)
```

```{r}
p <- purrr::map(unique(cluster_go_combined_dfs$cluster_number), function(clust){
  df <- cluster_go_combined_dfs %>% 
    filter(cluster_number == clust) %>% 
    filter(ontology != "N/A", qvalue < 0.2) %>% 
    mutate(setSize=as.integer(setSize)) %>% 
    dplyr::arrange(pvalue)
  
    # svglite(paste0("figures/combat_seq/bubble_plots/",clust,"_ClusterProfiler_bubblePlot.svg"),fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 5, width = 8)
    print(clusterProfiler_bubble_plots(df,pval_column = "pvalue",label = "labels", Filter=FALSE) +
      plot_annotation(title = paste0("Timecourse Cluster ",clust)))
    # dev.off()
})
names(p) <- unique(cluster_go_combined_dfs$cluster_number)
```

```{r}
# quick reminder on -log10(p) threshold
print(threshold)
```

```{r}
top_paths_profiles <- cluster_go_combined_dfs %>% 
   mutate(neg_log10_q=-log10(qvalue),
          neg_log10_p=-log10(pvalue)) %>% 
  
  group_by(cluster_number) %>% 
  dplyr::filter( qvalue < 0.2 ) %>%
  dplyr::filter(neg_log10_p > threshold) %>%
  
  dplyr::arrange(desc(neg_log10_p), .by_group = TRUE) %>%
  dplyr::slice(1:10) %>%
  mutate(x_pos = max(neg_log10_p)+0.5) %>%
  ungroup() %>%

  dplyr::filter(ontology != "N/A") %>%
  mutate(GO_term=ID,
         ID=paste(cluster_number, ID, sep="_")) %>%
  mutate(ID=factor(ID, levels = rev(unique(ID)) )) %>%
  arrange(desc(ID)) %>%
  mutate(text_color=case_when(
           ontology=="Cellular_Component" ~ "green4",
           ontology=="Biological_Process" ~ "tomato2",
           TRUE ~ "dodgerblue"
         ),
         setSize_clean=ifelse(setSize > 25, 25, setSize))  %>%
  select(ID:ontology, text_color, labels,
         x_pos,
         neg_log10_p,
         everything())
  # arrange(cluster_number)

# top_paths_profiles

# levels(top_paths_profiles$ID) %>% length()
table(top_paths_profiles$neg_log10_p > threshold) #OK
# write.csv(top_paths_profiles, "results/GO_enrichment/timecourse/time_course_clusters_combinedProfiles_clusterProfiler_top10_sig_results.csv",
#           row.names = FALSE)
```

```{r fig.width=12, fig.height=6}
y_labs <- pull(top_paths_profiles, GO_term, name= ID)

bubble_plot_profile <- ggplot(top_paths_profiles, aes(x=neg_log10_p, y=ID)) +
      geom_point(aes(size=setSize_clean, fill=ID),
                 shape=21,
                 alpha=0.8) +
      scale_radius(range=c(5,20), breaks = c(seq(0,25,by=5)),
                   labels=seq(0,25,by=5) %>% gsub("25",">= 25", .) %>% gsub("^5$","<= 5", .)) +
      labs(x="-log10(p-value)", y="") +

      scale_y_discrete(expand = c(0.1,0.1), 
                       labels = y_labs
                       ) +
      scale_x_continuous(expand = c(0.1,0.1),
                         limits = c(0,11.8),
                         breaks = seq(0,10, by=2)
                         ) +
      coord_cartesian(clip = 'off') +
      # ggh4x::facet_grid2(~cluster_number, scales = "free", independent = "all", axes="all")
      facet_wrap(~cluster_number, scales = 'free_y', ncol = 3) +
      # ggh4x::facetted_pos_scales(x = list(
      #           cluster_number %in%  paste0("Profile",c(1,3)) ~ scale_x_continuous(breaks = seq(1,12, by=2)),
      #           cluster_number  %in%  paste0("Profile",c(2,4:5)) ~ scale_x_continuous(breaks = seq(1,4, by=0.1))
      # )) +
      geom_text(aes(label = labels, color = ontology), x = 13.5, hjust = 0, size=6) +
      # scale_fill_viridis_d(option = "turbo", begin=0.2, end=0.9) +
      scale_fill_manual(values=rainbow(n = 38, v = 0.7, s=0.6)) +
      guides(fill="none",
             size=guide_legend(title = "Number of Genes\nin GO Term", size = 1)) +
      theme_classic() +
      theme(panel.grid.major.y =  element_line(color="black",size = 0.1),
            axis.text = element_text(size=14),
            # axis.text.y = element_text(color = top_paths_profiles$text_color),
            legend.position = 'left',
            panel.spacing.x = unit(19.5, "lines"),
            plot.margin = margin(r = 11.5, unit = "cm"))



# svglite("figures/combat_seq/bubble_plots/time_course_profiles_facet_plot_allOntology.svg",fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), width = 25, height = 10)
bubble_plot_profile 
# dev.off()
```


## TopGO 

https://bioconductor.org/packages/devel/bioc/vignettes/topGO/inst/doc/topGO.pdf

```{r}
#  loaded three environments GOBPTerm, GOMFTerm and GOCCTerm
suppressPackageStartupMessages(library(topGO))
```

topGO_2.46.0 (current 2.5.0, but how current are the GO annots themselves??)

```{r}
go_terms_avail <- c(ls(GOBPTerm),ls(GOMFTerm),ls(GOCCTerm))
# length(go_terms_avail)
# length(unique(go_terms_avail))
table(unique(GO_annots$go_id) %in% go_terms_avail)
missing <- unique(setdiff(GO_annots$go_id, go_terms_avail)) #46 missing 

GO_annots %>% 
  filter(go_id %in% missing) %>%
  dplyr::select(ontology, go_id) %>% 
  distinct() %>% 
  arrange(ontology) 

length(unique(GO_annots$gene_id)) #4,868
length(unique(GO_annots$go_id)) #3230

table(gene_clusts_dfs$cluster8$gene_id %in% unique(GO_annots$gene_id))
table(all_genes %in% unique(GO_annots$gene_id))
```

```{r}
gene2GO <- GO_annots %>% pull(go_id,name=gene_id)

topGO_eset <- cb_eset
masigpro_id <- pData(topGO_eset) %>% 
  group_by(time_point) %>% 
  mutate(masigpro_id=paste("Pf",  time_point, 
                           paste0("Rep", 1:n()), sep="_")) %>% 
  ungroup() %>% 
  pull(masigpro_id)

pData(topGO_eset)$masigpro_id <- masigpro_id
colnames(topGO_eset) <- masigpro_id
topGO_eset <- topGO_eset[,!grepl("Day2", colnames(topGO_eset))]
```

```{r}
all_genes <- fData(topGO_eset) %>% pull(gene_id)
gene_list <- as.numeric(all_genes %in% gene_clusts_dfs$cluster8$gene_id , 1,0) %>% 
  as.factor() %>% 
  set_names(all_genes)

# table(gene_list)

godata_BP <- new("topGOdata", 
              ontology = c("BP"), 
              nodeSize = 5,
              allGenes = gene_list,
              annot = annFUN.gene2GO, 
              gene2GO = gene2GO)

# godata_BP
```


```{r}
whichTests()
whichAlgorithms()
```

## goseq

```{r}
library(goseq)
library(GenomicRanges)
conflicted::conflict_prefer("select", "dplyr")
conflicted::conflict_prefer("filter", "dplyr")
conflicted::conflict_prefer("mutate", "dplyr")
```

```{r message=FALSE}
edb <- ensembldb::EnsDb("species_genomes/GRCh38_GRCm39_Pf3D7v58.sqlite")
edb
```

```{r}
genes_gr <- ensembldb::genes(edb)[rownames(cb_se),]
genes_gr %>%  head()
# length(genes_gr) #5188

gene_lens <- width(genes_gr) %>% 
  set_names(names(genes_gr))
```

```{r}
all_genes <- rownames(cb_se)
gene_list <- as.numeric(all_genes %in% gene_clusts_dfs$cluster8$gene_id , 1,0) %>% 
  as.factor() %>% 
  set_names(all_genes)

table(gene_list)
```

```{r}
supportedGenomes() %>% 
  filter(grepl("mal|plas|fal", species, ignore.case = T) | 
           grepl("PF", db, ignore.case = TRUE))
# supportedOrganisms()
supportedGeneIDs() %>%  head()
```

Warning: initial point very close to some inequality constraintsWarning: collapsing to unique 'x' values

```{r}
GO_annots_clean <- GO_annots %>% 
  group_by(ontology,go_id) %>% 
  dplyr::mutate(n_genes = n())  %>% 
  ungroup() %>% 
  select(gene_id:go_id, n_genes) %>% 
  dplyr::arrange(ontology, go_id,n_genes) %>% 
  dplyr::filter(n_genes >= 5, n_genes <= 1000)

GO_annots_clean %>% 
  select(ontology, go_id) %>% 
  distinct() %>% 
  group_by(ontology) %>% 
  dplyr::count()
```

```{r}
GO_annots %>% 
  filter(ontology=="Biological Process") %>% 
  select(gene=gene_id,go_id)
```


```{r}
input_go <- 
  GO_annots_clean %>%
  filter(ontology=="Biological Process") %>% 
  dplyr::arrange(gene_id) %>% 
  select(gene=gene_id,go_id) %>% 
  as.data.frame()


pwf <- nullp(DEgenes = gene_list, 
             genome = "PF3D7",
             id = "ensGene",
             bias.data = gene_lens)
```

```{r}
go_res <- goseq(pwf, 
                genome = "PF3D7",
                id = "ensGene",
                gene2cat = input_go, 
                method = "Wallenius",
                # method = "Sampling",
                repcnt = 2000,
                use_genes_without_cat = F)
```

Error: node stack overflow
Error during wrapup: node stack overflow
Error: no more error handlers available (recursive errors?); invoking 'abort' restart

```{r}
go_res_clean <- go_res %>% 
  as.data.frame() %>% 
  dplyr::mutate(p_adj = p.adjust(p = as.numeric(over_represented_pvalue),
                                 method = "BH")) %>% 
  dplyr::select(category, p_adj, everything())

# quantile(go_res_clean$p_adj)
go_res_clean
```


## PlasmoDB 

```{r}
Day2 <- read.csv("results/DEGs/kappe_s_Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_DEGs.csv", comment.char = "#")

Day2 %>%
  dplyr::filter(logFC > 0) %>%
  pull(gene_id) %>%
  write.table(.,results/GO_enrichment/day2/Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_upregulated.csv",
              sep="\t", quote=FALSE, row.names = FALSE)

Day2 %>%
  dplyr::filter(logFC < 0) %>%
  pull(gene_id) %>% 
  write.table(.,results/GO_enrichment/day2/Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_downregulated.csv",
              sep="\t", quote=FALSE, row.names = FALSE)

sigs_all_df_d456 <-  read.csv("results/time_course/masigpro/masigpro_significant_genes_d456_results.csv") %>% 
  rename_at(vars(X), ~c("gene_id"))

gene_clusts_dfs <- purrr::map(unique(sigs_all_df_d456$gene_cluster), function(clust){
  out <- here::here(file.path("GO_enrichment",paste0("cluster",clust)))
  dir.create(out,recursive = T, showWarnings = FALSE)
  
  sigs_all_df_d456 %>%
    dplyr::filter(gene_cluster==clust) %>% 
    pull(gene_id) %>%
    write.table(., file.path(out,
                             paste0("Heps_Pf_vs_sporozorite_combat-seq_corrected_masigpro_cluster",clust, ".csv")),
                sep="\t", quote=FALSE, row.names = FALSE)
})
```

https://static-content.veupathdb.org/documents/GO-Enrichment.pdf

```{r}
# From PLASMODB GO analysis - need to find the GO terms themselves...
res_path <- dir"results/GO_enrichment/",
                recursive = TRUE,
                pattern = ".tsv",
                full.names = TRUE)
res <- purrr::map(res_path, function(x) read.delim(x) %>% 
                    janitor::clean_names())
names(res) <- basename(res_path) %>% gsub("hiddenGoEnrichmentResult_|.tsv", "",.)
lapply(res, head)
```

```{r}
# BiocManager::install("org.Pf.plasmo.db", 
#                      lib=.libPaths()[1])
#org.Pf.plasmo.db is based on old data that are no longer updated. org.Pf.plasmo.db is deprecated and will be removed from Bioconductor version 3.15.
# library(org.Pf.plasmo.db)
# BiocManager::install("malaria.db0", lib=.libPaths()[1])
```

```{r}
create_bubble_plots <- function(df){
  
    input <- df %>% 
      mutate(neg_log10_P=-log10(p_value),
             log2_FC=log2(fold_enrichment+1)) %>% 
      arrange(benjamini)
    
    temp <- input %>% 
      dplyr::filter(benjamini < 0.05) %>% 
      mutate(id=factor(id, levels = rev(id)))
  
    if(nrow(temp) == 0){
      temp <- input %>% 
        mutate(id=factor(id, levels = rev(id)))
    }
    
    n <- min(20,nrow(temp))
    temp <- temp %>%  
      dplyr::slice(1:n)
    
    bubble_plot <- ggplot(temp, aes(x=neg_log10_P, y=id)) +
      geom_point(aes(size=result_count, fill=id), shape=21,
                 alpha=0.8) +
      scale_radius(range=c(1,10)) +
      labs(x="-log10(p-value)", y="") +
      scale_y_discrete(expand = c(0.05,0)) +
      theme(panel.grid.major.x = element_blank(),
            panel.grid.minor.x = element_blank(),
            panel.grid.major = element_line(color="black",size = 0.1),
            panel.background = element_rect(fill="white"),
            panel.border = element_rect(color="black", fill=NA)) +
      guides(fill="none",
             # size="none",
             size=guide_legend(title = "Number of DEGs with GO Term",
                               size = 1)
             ) +
      theme(legend.position = "left",
            legend.key = element_rect(colour = NA, fill = NA))

    # return(bubble_plot)
    tab <-  ggplot(temp, aes(x = 0, y = id, label=name)) +
        geom_text(size = 3,
                  hjust = 0,
                  nudge_x = 0)  +

        scale_y_discrete(expand = c(0.05,0)) +
        scale_x_continuous(expand = expansion(mult = c(0,1e2))) +
        coord_cartesian(clip="off",
                        xlim=c(0, 500)) +
        theme_void()

    return(bubble_plot +
      tab + 
      plot_layout(widths = c(1, 2.5), guides='collect') &
      theme(legend.position='top'))
}
```

```{r fig.height=5, fid.width=10}
purrr::map(names(res),function(df_ID){ 
  svglite(paste0("kappe_s_",df_ID,"_bubbleplot.svg"), height = 5, width = 10)
  print(create_bubble_plots(res[[df_ID]]))
  dev.off()
})

# tab
```


# Regulatory Networks 

https://www.bioconductor.org/packages/release/bioc/vignettes/RTN/inst/doc/RTN.html
- sample size limit??

https://bioconductor.org/packages/release/bioc/vignettes/qpgraph/inst/doc/qpTxRegNet.pdf
- Need Tfs and target genes! 

https://www.bioconductor.org/packages/release/bioc/html/CBNplot.html
- Only run on POSIT
- uses GSEA results
- no support for undirected graphs?? 

EnrichmentBrowser
- Need a network 
- Could try kegg metabolic paths?

**BioNET**
https://bioconductor.org/packages/release/bioc/vignettes/BioNet/inst/doc/Tutorial.pdf
- Need PPI 
- https://www.nature.com/articles/nature04104 
- https://www.sciencedirect.com/science/article/pii/S221112471930909X
  - P. falciparum schizonts were freed from erythrocytes by incubation with 0.1% saponin in PBS for 10 minutes and then washed with PBS to remove saponin.
  - Mouse-derived P. berghei infected erythrocytes were put into culture for 22-24 hours to generate schizont
  - P. knowlesi schizonts were [...] freed from erythrocyte
  

We identified maximally interconnected networks from PPI derived from blood stage schizonts. 
- more functional information!

```{r}
suppressPackageStartupMessages(library(BioNet))
suppressPackageStartupMessages(library(igraph))
# library(CBNplot) 
# library(RTN)
# library(EnrichmentBrowser)

# BiocManager::install("EnrichmentBrowser", update = FALSE)
# hsa.grn <- compileGRN(org="ko", db="kegg")
# head(hsa.grn)
```

```{r}
# read in the time-course results
if(!exists("gene_clusts_combined_dfs")){
  time_course <- read.csv("results/time_course/masigpro/masigpro_significant_genes_degree2_d456_results.csv") %>% 
    select(-X) %>% 
    mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>% 
    mutate(combined_gene_clusters=case_when(
      grepl("1$|2$",gene_cluster) ~ "Profile1",
      grepl("3$|8$", gene_cluster) ~ "Profile2",
      grepl("4$|6$", gene_cluster) ~ "Profile3",
      grepl("5$|9$", gene_cluster) ~ "Profile4",
      grepl("7", gene_cluster) ~ "Profile5"
    )) %>% 
    dplyr::select(gene_id, matches("gene_cluster"), everything()) %>% 
    arrange(gene_cluster)
}

# Need the fit object from MaSigPro for p-values for genes not selected in time-course analysis
fit_d456 <- readRDS("results/time_course/masigpro/masigpro_fit_p.vector_degree2_D456.RDS")

# Need gene_names info
pf_genes <- as.data.frame(rowData(cb_se)) %>% 
  mutate(gene_name = gsub("Name=","", gene_name))

# collect DE analysis results into a single file 
combat_de_nofilter <- readRDS("results/DEGs/kappe_sHeps_Pf_vs_sporozorite_combat-seq_corrected_DEGs_nogenefilter_list.RDS")
DEGs <- purrr::map(combat_de_nofilter[grep("Day[456]", names(combat_de_nofilter))], function(res){
  grp <- dimnames(res$eBayesFit$contrasts)$Levels[2]
  print(grp)
  df <- res$all_genes %>%
    rownames_to_column("gene_id") %>%
    janitor::clean_names() %>% 
    mutate(group = grp) %>%
    rename_at(vars(-gene_id), ~paste(., grp, sep = "_")) %>% 
    select(matches("group"), everything())
})  %>% 
  purrr::reduce(.x = .,
                .f = function(x,y){
    full_join(x, y, by = "gene_id")
  }) %>%
  left_join(., pf_genes, by = "gene_id") %>%
  rowwise() %>% 
  mutate(num_upreg = sum(c_across(cols = matches("log_fc_")) > 0) ) %>% 
  ungroup() %>% 
  select(gene_id,
         gene_name,
         num_upreg,
         matches("group"),
         matches("p_value_"),
         everything())

head(DEGs)
# dim(DEGs) #5188   34
DEGs$num_upreg %>% table()
#    0    1    2    3 
# 2087  292  602 2207 
```

```{r}
PPI_GO_terms <- openxlsx::read.xlsx("resources/PPI/1-s2.0-S221112471930909X-mmc8.xlsx") %>% 
  janitor::clean_names() %>% 
    group_by(cluster_number) %>% 
  mutate(go_processes = paste(go_process, collapse = "; ")) %>% 
  ungroup() %>% 
  select(cluster_number,go_processes) %>% 
  distinct() %>% 
  arrange(cluster_number)

# PPI_GO_terms

PPI_clusters <- openxlsx::read.xlsx("resources/PPI/1-s2.0-S221112471930909X-mmc7.xlsx") %>% 
  janitor::clean_names() %>% 
  mutate_at(vars(corum_complex), ~ifelse(is.na(.), "", .)) %>% 
  group_by(cluster_number) %>% 
  mutate(corum_complex_clean = str_split_fixed(unique(corum_complex), pattern = "[;]", n = 3)[,1:2] %>% 
           paste(., collapse = "; ")) %>% 
  ungroup() %>% 
  mutate_at(vars(corum_complex_clean), ~gsub("^;|;\\s{1,}$", "", .) %>% 
              gsub(", cytoplasmic", "", .) %>% 
              gsub("\\s?\\(.+\\)\\s?", "", .)) %>% 
  select(cluster_number, gene_id, symbol, corum_complex_clean)  %>% 
  left_join(., PPI_GO_terms, by = "cluster_number") %>% 
  mutate_at(vars(go_processes,corum_complex_clean), ~ifelse(is.na(.), "", .)) 

# head(PPI_clusters)
# dim(PPI_clusters) #1785    3
````

```{r}
PPI <- openxlsx::read.xlsx("resources/PPI/1-s2.0-S221112471930909X-mmc6.xlsx") %>% 
  janitor::clean_names() %>% 
  
  # skip protein nodes that have multiple gene_ids
  # and retain those found in the expression data (filtered to remove lowly expressed genes)
  filter(!grepl(",", source)) %>% 
  filter(source %in% pf_genes$gene_id, 
         target %in% pf_genes$gene_id) 

head(PPI)
# dim(PPI) #23544    36
```

```{r}
gene_names_annot <- pf_genes %>% 
                      pull(gene_name, name = gene_id)

# head(gene_names_annot)

protein_complex_annots <- PPI %>% 
  select(source, target) %>% 
  pivot_longer(cols = everything(), 
               names_to =  "cols", 
               values_to = "gene_id") %>% 
  left_join(., PPI_clusters, by= c("gene_id")) %>% 
  mutate(protein_complex_id = ifelse(is.na(cluster_number), "unassigned", cluster_number)) %>% 
  select(-cluster_number, -cols) %>% 
  distinct() %>% 
  
  group_by(protein_complex_id) %>% 
  mutate(number_proteins_in_complex = n()) %>% 
  ungroup()  %>% 
  mutate_at(vars(go_processes,corum_complex_clean), ~ifelse(is.na(.), "", .)) %>% 
  select(gene_id, protein_complex_id, number_proteins_in_complex,
         everything())


# head(protein_complex_annots)
# dim(protein_complex_annots) #2141    5
``` 


### BioNet

broken link https://software.cwi.nl/software/heinz
https://academic.oup.com/bioinformatics/article/26/8/1129/207852

```{r}
ppi_modules <- function(time_course_res, cluster_name, input_network, msp_fit, DEGs = NULL, percentile=0.1){
  
  # combine time-course and DEGs p-values into a single vector
  if(!is.null(DEGs)){
      message("combining DE results with time-course results p-values")
      # join time-course and DE results 
      pvals <- time_course_res %>% 
        filter(combined_gene_clusters == cluster_name) %>% 
        select(gene_id, p.value) %>% 
        inner_join(.,select(DEGs, gene_id, matches("p_value_")),
                  by = "gene_id") %>% 
        column_to_rownames("gene_id") %>% 
        as.matrix()
      
      # combine time-course p-values for all genes with all DE results (including non-significant hits)
      all_pf_pvals <- msp_fit$p.vector %>% 
        as.data.frame() %>% 
        rownames_to_column("gene_id") %>% 
        inner_join(.,select(DEGs, gene_id, matches("p_value_")),
                  by = "gene_id") %>% 
        column_to_rownames("gene_id") %>% 
        as.matrix()

      # remove the time-course cluster genes from the "full" `all_pf_pvals` dataset
      input_pf_vals <- all_pf_pvals[setdiff(rownames(all_pf_pvals), time_course_res$gene_id),]
      
      chk1 <- !any(rownames(input_pf_vals) %in% rownames(pvals)) # TRUE
      chk2 <- !any(rownames(input_pf_vals) %in% time_course_res$gene_id) #TRUE
      chk3 <- identical(colnames(input_pf_vals), colnames(pvals)) #TRUE
      # testit::assert("", )
      if(all(c(chk1,chk2,chk3))){
         in_pvalues <- rbind(pvals, input_pf_vals)
         print(glue("the input number of genes is {nrow(in_pvalues)}."))
      }
 
    } else {
      message("Using time-course results p-values")
      # gene ids of all time-course selected sig genes
      all_sig_genes <- time_course_res %>% 
        pull(gene_id)
      
      # pvalues of time-course cluster
      pvalues <- time_course_res %>% 
        filter(combined_gene_clusters == cluster_name) %>%  
        pull(p.value, name = gene_id) 
      
      # need p-values for the non-sig genes too
      pf_genes_pvals <- msp_fit$p.vector[,1] %>% 
        set_names(., value = rownames(msp_fit$p.vector))
    
      # remove genes from other time-course clusters
      not_in_profile <- pf_genes_pvals[setdiff(names(pf_genes_pvals), all_sig_genes)]
      in_pvalues <- c(pvalues, not_in_profile)
      in_pvalues <- in_pvalues[order(in_pvalues)]
    }
  

  # aggregate if its a matrix input
  if(is.matrix(in_pvalues)){
      in_pvalues <- aggrPvals(in_pvalues, plot=FALSE)
      in_pvalues <- in_pvalues[order(in_pvalues)]
  }
  # beta-uniform mixture model to a given p-value distribution
  message("fitting p-values")
  fb <- fitBumModel(x = in_pvalues, plot=FALSE, starts = 10)
  
  # subset the network graph
  message("subset the network graph")
  subnet <- subNetwork(nodeList = names(in_pvalues),
                        network = input_network)
  subnet <- rmSelfLoops(subnet)
  subnet <- largestComp(subnet)
  # A graphNEL graph with undirected edges
  
  # Add scores to the subnetworks
   message(glue("Add scores to the subnetworks using the top {percentile} percentile threshold"))
  scores <- scoreNodes(network = subnet,
                       fb = fb, 
                       fdr = quantile(fb$pvalues, probs = percentile))
  
  # identify the enriched subnetwork(s) 
  message("identify the enriched subnetwork(s)")
  module <- runFastHeinz(network = subnet, 
                         scores = scores)
  
  res <- list("module" = module,
              "scores" = scores,
              "pvalues" = in_pvalues,
              "model" = fb)
}
```

```{r}
ppi_network_figures <- function(input_module, protein_complex_annots, gene_names_annot, DEGs, title = ""){
  
    # node (gene_id) names 
    sel_nodes <- nodes(input_module)
    
    # subset to genes in the module 
    protein_complexes <- protein_complex_annots %>% 
        filter(gene_id %in% sel_nodes) %>% 
        
        # count number of genes within protein complexes
        # ppic == protein-protein interaction complex
        group_by(protein_complex_id) %>%
        mutate(number_genes_in_ppic = n()) %>% 
        ungroup()  %>% 
        mutate(percent_in_complex = round( number_genes_in_ppic / number_proteins_in_complex * 100, digits = 2)) %>%
        
        # remove duplicate gene_ids and select those with protein complex that has > 3 genes
        group_by(gene_id) %>%
        mutate(num_gene_annots = n(),
               keep = case_when(
                      num_gene_annots == 1 ~ TRUE,
                      num_gene_annots > 1 & all(number_genes_in_ppic == 1) ~ !duplicated(gene_id),
                      num_gene_annots > 1 & any(grepl("^[0-9]", protein_complex_id)) ~ number_genes_in_ppic == max(number_genes_in_ppic)
        )) %>% 
        ungroup() %>% 
        arrange(gene_id) %>% 
        filter(keep) %>% 
    
        # check for any remaining duplicate annotations
        mutate(dups = duplicated(gene_id) | duplicated(gene_id, fromLast = TRUE)) %>% 
        
        # retain annots for graph visualization
        mutate(protein_complex_id = case_when(
                          protein_complex_id == "unassigned" ~ "other", 
                          number_genes_in_ppic == 1 ~ "other", 
                          TRUE ~ protein_complex_id) %>% 
                      as.factor(.)
        ) %>% 
        
        # generate color codes for the graph figure
        group_by(protein_complex_id) %>%
        mutate(node_color = case_when(
           protein_complex_id == "other" ~ "grey80",
           TRUE ~ colors_vector[dplyr::cur_group_id()]
        )) %>%
        ungroup() %>%
        
        # order the data to be identical to nodes in graph
        mutate(gene_id = factor(gene_id, levels = sel_nodes)) %>% 
        arrange(gene_id) %>% 
        
        select(dups, 
               number_genes_in_ppic,
               percent_in_complex,
               keep,
               num_gene_annots,
               everything()) 

    chk1 <- all(!protein_complexes$dups) # TRUE
    if(!chk1){
      # a second round of filtering to address ties
      protein_complexes <- protein_complexes %>% 
        group_by(gene_id) %>%
        mutate(keep = ifelse(dups == TRUE & percent_in_complex != max(percent_in_complex), FALSE, TRUE)) %>% 
        ungroup() %>% 
        filter(keep) %>% 
        # check for any remaining duplicate annotations
        mutate(dups = duplicated(gene_id) | duplicated(gene_id, fromLast = TRUE))
    }
    
    chk2 <- all(!protein_complexes$dups) # TRUE
    chk3 <- nrow(protein_complexes) == length(sel_nodes) # TRUE
    chk4 <- all(sel_nodes %in% protein_complexes$gene_id) # TRUE
    if(all(c(chk2, chk3, chk4))){
    
        # combine the DEGs results and protein complex annotations
        fcs <- DEGs %>% 
          filter(gene_id %in% sel_nodes) %>% 
          pivot_longer(cols = matches("log_fc_"), 
                       names_to = "time_point",
                       values_to = "log_fc") %>% 
          select(gene_id, gene_name, time_point, log_fc) %>% 
          left_join(., protein_complexes, by = "gene_id") %>% 
          mutate(time_point = gsub(".+(Day[456])_.+","\\1", time_point)) %>%
          group_by(gene_name) %>%
          mutate(mean_fc = mean(log_fc)) %>%
          ungroup() %>%
          # arrange heatmap/tileplot by protein cluster ID and decreasing fold-change 
          arrange(protein_complex_id, desc(mean_fc)) %>% 
          mutate(gene_name = factor(gene_name, levels = rev(unique(gene_name))))
        
        # color codes vector for the protein cluster numbers 
        pcid <- select(protein_complexes, 
                      node_color, 
                      protein_complex_id) %>%
          distinct() %>% 
          mutate(x = "cluster")
        
        ###### Generate heatmap of Fold-changes for genes in the module
        fc_plot <- ggplot(fcs, aes(x = time_point, y = gene_name, fill = log_fc)) +
          labs(y = "", x = "", title = "Fold-Changes per LS Timepoint") +
          geom_tile() + 
          coord_fixed() +
          scale_x_discrete(expand = c(0,0)) +
          scale_fill_viridis_c(option = "A") +
          theme_classic() +
          theme(legend.position = "top", 
                axis.text = element_text(color = "black", size = 8),
                axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1),
                plot.title = element_text(size = 10))
        
        protein_complex_anno <- ggplot(fcs, aes(x = "protein_complex", y = gene_name)) +
          geom_text(aes(label = protein_complex_id, color = protein_complex_id),
                    size = 4) +
          scale_color_manual(values = pull(pcid, node_color, name = protein_complex_id),
                             guide = NULL) +
          labs(title = "ID") +
          theme_void() +
          theme(plot.title = element_text(size = 10))
        
        anno_df <- fcs %>% 
          select(gene_name, protein_complex_id, go_processes, corum_complex_clean) %>% 
          mutate_at(vars(go_processes, corum_complex_clean), 
                    ~ifelse(protein_complex_id == "other", "", .)) %>% 
          mutate(anno = paste(go_processes, corum_complex_clean, sep = "; ")) %>% 
          mutate(anno = gsub("^;|^NA;|NA$", "", anno)) %>% 
          select(gene_name, protein_complex_id, anno)
        
        ppic_anno <- ggplot(anno_df, aes(x = "protein_complex", y = gene_name)) +
          geom_text(aes(label = anno, color = protein_complex_id),
                    size = 2) +
          scale_color_manual(values = pull(pcid, node_color, name = protein_complex_id), 
                             guide = NULL) +
          labs(title = "Protein Complex GO/CORUM Annotation") +
          coord_fixed(clip = "off") +
          theme_void() +
          theme(plot.title = element_text(size = 10))
        
        fold_changes_heatmap <- fc_plot + 
                                protein_complex_anno +
                                ppic_anno +
                                patchwork::plot_layout(guides = "collect",
                                                       ncol = 3, 
                                                       width = unit(c(1, 2, 1), c("null", "cm","null"))) &
                                theme(legend.position = 'left')
        
      
        #### Generate a figure legend for the network graph
    
        # barplot of the color codes
        legend <- ggplot(pcid, aes(x = x, y = protein_complex_id)) +
          geom_tile(aes(fill = protein_complex_id)) +
          scale_fill_manual(values = pull(pcid, node_color, name = protein_complex_id)) +
          labs(title = "Protein Complex IDs") +
          coord_fixed() +
          theme_void() +
          theme(legend.position = "none", 
                axis.text.y = element_text(color = "black", size = 14), 
                axis.ticks.y = element_line(color = "black", linewidth = 1, linetype = "solid"),
                axis.ticks.length.y = unit(2,"mm"))
        
        #### Generate  network graph
        set.seed(20240410)
        g <- graph_from_graphnel(input_module)
        layout <- layout_with_kk(g, dim = 2)
        # make gene symbol as labels
        idx <- match(V(g)$name, names(gene_names_annot))
        # table(is.na(idx))
        V(g)$gene_names <-  gene_names_annot[idx]
        V(g)$node_color <- protein_complexes %>% 
          pull(node_color, name = gene_id) %>% 
          alpha(., alpha = 0.5)
        
        module_network_graph <- plot(g, 
                                     main = title,
                                     layout = layout,
                                     vertex.label.dist = 0, 
                                     label.color = "black",
                                     edge.color = "grey50",
                                     vertex.label.cex = 0.5,
                                     vertex.label.color = "black",
                                     vertex.size = 12,
                                     vertex.label = V(g)$gene_names,
                                     vertex.color = V(g)$node_color)
        
        figs <- list("fcs" = fold_changes_heatmap, 
                      "graph_legend" = legend,
                      "network_graph" = module_network_graph)
        return(figs)
    } else {
      message("protein complex annotations has unresolved duplicates.")
      return(protein_complexes)
    }
}
```

```{r}
pf_ppi_graph <- ftM2graphNEL(ft = as.matrix(PPI[,c("source","target")]), 
                             W = PPI[["r_fscore"]],
                             edgemode="undirected")

pf_ppi_graph
# A graphNEL graph with undirected edges
# Number of Nodes = 1642 
# Number of Edges = 23544
```


TO DO: make the breaks on the heatmap identical to ALL plots!! 
- TO CONSIDER: if there are ties after the first pass of filtering protein annots, how to select the "best" annotation to view
- the one with the most members in the module or the one with GO/CORUM annots if available?


- color the graph nodes by average expression 
- provide the heatmap of the 3 replicates FC's in supplement  

https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-023-05376-z

BioNet fits a beta-uniform mixture model using p values from a differential expression (DE) analysis to score each node in the network and then uses integer linear programming to optimally solve the **Maximum-weight Connected Subgraph (MWCS)** problem [11]. The MWCS problem aims to identify a subset of nodes of maximum node weight that are connected, where connected means there exists a path between any two nodes in the subgraph. 

Network based analyses
https://link.springer.com/protocol/10.1007/978-1-0716-0239-3_23


```{r}
timecourse_profiles <- unique(time_course$combined_gene_clusters) %>% 
  set_names(., value = .)

enriched_upreg_ppi_modules <- purrr::map(timecourse_profiles, function(profile){
  print(profile)
  
  ppi_modules(time_course_res = time_course, 
              cluster_name = profile,
              msp_fit = fit_d456,
              input_network = pf_ppi_graph, 
              DEGs = filter(DEGs_df, num_upreg >= 2),
              # DEGs = DEGs_df,
              percentile = 0.25)

})

enriched_dnreg_ppi_modules <- purrr::map(timecourse_profiles, function(profile){
  print(profile)
  
  ppi_modules(time_course_res = time_course, 
              cluster_name = profile,
              msp_fit = fit_d456,
              input_network = pf_ppi_graph, 
              DEGs = filter(DEGs_df, num_upreg < 2),
              # DEGs = DEGs_df,
              percentile = 0.25)

})
```

```{r fig.height=7, fig.width=7}
# enriched_ppi_modules$Profile1$module
# names(enriched_ppi_modules$Profile1)
# checkout model fit
 lapply(names(enriched_ppi_modules), function(x){
   input <- enriched_ppi_modules[[x]]
  #  p <- enriched_ppi_modules[["pvalues"]]
  #  mod <- enriched_ppi_modules[["model"]]
  #  
  plotLLSurface(x = enriched_ppi_modules[[x]]$pvalues, 
              opt = enriched_ppi_modules[[x]]$model)
 })
# checkout model fit
lapply(enriched_ppi_modules, function(x) plot(x[["model"]]))
```   


```{r}
sel_modules <- lapply(enriched_upreg_ppi_modules, function(x) x[["module"]])
sapply(names(sel_modules), function(x){
  p <- time_course %>% 
    filter(combined_gene_clusters == x) 
  table(nodes(sel_modules[[x]]) %in% p$gene_id)
})
```


```{r fig.height=5, fig.width=5}
# module_figs <- purrr::map(enriched_ppi_modules, function(x){
for ( profile in names(enriched_upreg_ppi_modules)[3:5]){
    print(profile)
    print(
      ppi_network_figures(input_module = enriched_upreg_ppi_modules[[profile]]$module, 
                          protein_complex_annots = protein_complex_annots,
                          gene_names_annot = gene_names_annot,
                          DEGs = DEGs, 
                          title = profile)
          )
}

# })
```

```{r}
data(interactome,package = "DLBCL") #A graphNEL graph with undirected edges
data(dataLym, package = "DLBCL")
data(exprLym, package = "DLBCL") 

pvals <- cbind(t=dataLym$t.pval, s=dataLym$s.pval)
rownames(pvals) <- dataLym$label
pval <- aggrPvals(pvals, order=2, plot=TRUE)


subnet <- subNetwork(dataLym$label, interactome)
subnet <- rmSelfLoops(subnet)
subnet <- largestComp(subnet)
subnet


fb <- fitBumModel(pval, plot=FALSE)
scores <- scoreNodes(subnet, fb, fdr=0.001)

module <- runFastHeinz(subnet, scores)
logFC <- dataLym$diff
names(logFC) <- dataLym$label

plotModule(module, scores=scores, diff.expr=logFC)
  # layout_with_fr()
    # layout_with_sugiyama(g) 
                                 # layers = c(1:10))
    # layout_with_mds(g)
    # layout_nicely(g,
    #                         # weights = NA,
    #                         weights = 1/E(g)$weight,
    #                         # use.seed = TRUE,
    #                         # seed = 20240410, 
    #                         dim = 3)
    # layout_in_circle(g)
    # layout_nicely(g)
  
```

```{r}
exclude <- time_course_res %>% 
  filter(combined_gene_clusters != "Profile5") %>% 
  pull(gene_id)

temp <- time_course_res %>% 
  filter(combined_gene_clusters == "Profile5") %>% 
  select(gene_id, p.value,betaTime:betaTime2) %>% 
  left_join(.,select(DEGs, gene_id, matches("p_value_"), matches("log_fc_")),
            by = "gene_id")


temp
# length(exclude) #1386

# day4_de <- read.csv(de_res[1], comment.char = "#")
# 
# table(is.na(temp$p_value_Day4_Heps_Pf))
# # FALSE  TRUE 
# #    70    54 
# table(is.na(in_pvalues), names(in_pvalues) %in% temp$gene_id)

na_ids <- temp %>% 
  filter(is.na(p_value_Day4_Heps_Pf)) %>% 
  pull(gene_id)


in_pvalues[na_ids]
```

```{r}
DEGs_pvals <- DEGs %>% 
  select(gene_id, matches("p_value_"), matches("log_fc_")) %>% 
  filter(!gene_id %in% exclude) %>% 
  mutate(group = ifelse(gene_id %in% temp$gene_id, "p3","background")) %>% 
  pivot_longer(cols = matches("p_value_")) #%>% 
  # filter(!(group == "background" & value < 0.05))


length(unique(DEGs_pvals$gene_id)) #2376
# table(pval_dist$group)

ggplot(DEGs_pvals, aes(x = group, y = value)) +
  geom_boxplot(aes(color = group))


# dim(DEGs_pvals)
# DEGs_pvals$group %>% table()
```

```{r fig.width=5}
pval_dist <- in_pvalues %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  mutate(group = ifelse(gene_id %in% temp$gene_id, "p3","background")) %>% 
  pivot_longer(cols = p.value:p_value_Day6_Heps_Pf) %>% 
  filter(!(group == "background" & value < 0.01))

# length(unique(pval_dist$gene_id))
# table(pval_dist$group)

ggplot(pval_dist, aes(x = group, y = value)) +
  geom_boxplot(aes(color = group))
```

```{r}
pvals <- time_course_res %>% 
    filter(combined_gene_clusters == cluster_name) %>% 
    select(gene_id, p.value) %>% 
    left_join(.,select(DEGs, gene_id, matches("p_value_")),
              by = "gene_id") %>% 
    mutate(across(.cols = matches("^p.value"), .fns = ~ifelse(is.na(.), 1, .)))

pvals
```



### Enrichment Browser 

```{r}
nbea.res <- nbea(method="ggea",
                 se=cb_se,
                 gs=GO_annots_list, 
                 grn=GRN) 
gsRanking(nbea.res)


par(mfrow=c(1,2))
ggeaGraph( gs=hsa.gs[["hsa05217_Basal_cell_carcinoma"]], grn=hsa.grn, se=allSE) 
ggeaGraphLegend()
```

### RTN 

```{r}
data(tniData)
tniData$rowAnnotation %>% head()
```

```{r}
rtni <- tni.constructor(expData = tniData$expData, 
                        regulatoryElements = GO_annots_TF$gene_id, 
                        rowAnnotation = tniData$rowAnnotation, 
                        colAnnotation = tniData$colAnnotation)

rtni <- tni.permutation(rtni, nPermutations = 100)
rtni <- tni.bootstrap(rtni)
rtni <- tni.dpi.filter(rtni)
regulons <- tni.get(rtni, what = "regulons.and.mode", idkey = "SYMBOL")
```


# Motif Analysis

https://bioconductor.org/packages/release/bioc/vignettes/universalmotif/inst/doc/MotifManipulation.pdf

https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1001165#s5

https://rdrr.io/github/robertamezquita/marge/man/write_homer_motif.html

http://homer.ucsd.edu/homer/motif/creatingCustomMotifs.html


```{r message=FALSE}
library(monaLisa)
library(JASPAR2020)
library(TFBSTools)
library(MotifDb)
library(universalmotif)
library(ggseqlogo)
```

https://en.wikipedia.org/wiki/Position_weight_matrix

A PWM has one row for each symbol of the alphabet (4 rows for nucleotides in DNA sequences or 20 rows for amino acids in protein sequences) and one column for each position in the pattern. 

In the first step in constructing a PWM, a basic position frequency matrix (PFM) is created by counting the occurrences of each nucleotide at each position. 

From the PFM, a position probability matrix (PPM) can now be created by dividing that former nucleotide count at each position by the number of sequences, thereby normalising the values. 

## Input Datasets 

```{r eval=TRUE}
Day2 <- read.csv("results/DEGs/kappe_s_Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_DEGs.csv", comment.char = "#")

time_course <- read.csv("results/time_course/masigpro/masigpro_significant_genes_degree2_d456_results.csv", row.names = 1) %>% 
  # rename_at(vars(X), ~c("gene_id")) %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  dplyr::select(gene_id, matches("gene_cluster"), everything()) %>% 
  arrange(gene_cluster)

gene_clusts_dfs <- purrr::map(unique(time_course$gene_cluster), function(clust){
  out <- here::here(file.path("results/motif_analysis/homer",paste0("cluster",clust)))
  dir.create(out,recursive = T, showWarnings = FALSE)
  subset <- time_course %>%
    dplyr::filter(gene_cluster==clust)
  
    outfile <- file.path(out,paste0("Heps_Pf_vs_sporozorite_combat-seq_corrected_masigpro_cluster",clust, ".csv"))
    if(!file.exists(outfile)){
      print(c("Saving", outfile))
      write.table(subset, outfile, sep="\t", quote=FALSE, row.names = FALSE)
    }
    return(subset)
})
names(gene_clusts_dfs) <- paste0("cluster",unique(time_course$gene_cluster))

# gene_clusts_dfs

gene_clusts_combined_dfs <- purrr::map(unique(time_course$combined_gene_clusters), function(clust){
  subset <- time_course %>%
    dplyr::filter(combined_gene_clusters==clust)
  
  out <- here::here(file.path("results/motif_analysis/homer",paste0("cluster",clust)))
  dir.create(out,recursive = T, showWarnings = FALSE)
  outfile <- file.path(out,paste0("Heps_Pf_vs_sporozorite_combat-seq_corrected_masigpro_",clust, ".csv"))
  if (!file.exists(outfile)) {
    print(c("Saving", outfile))
    write.table(subset,outfile, sep="\t", quote=FALSE, row.names = FALSE)
  } 
  
  return(subset)
}) %>% 
  set_names(unique(time_course$combined_gene_clusters))

table(time_course$gene_cluster_factor,
      time_course$combined_gene_clusters)
```

## Promoter Sequences 

```{r warning=FALSE, message=FALSE}
#ensure same seqlevels and seqlengths are present
pf_proms_gr <- promoters_gr[promoters_gr$gene_id %in% rowData(cb_se)[["gene_id"]]]
seqlevels(pf_proms_gr) <- names(genome)
seqlengths(pf_proms_gr) <- width(genome)
pf_proms_gr <- trim(pf_proms_gr) #some ranges are outside the expected lengths of Pf3D7_MIT_v3

length(pf_proms_gr) #5,258

# create gene ID mappings
pf_names <- rowData(cb_se) %>% 
  as.data.frame() %>% 
  select(gene_name, gene_id)

human_mouse_genes <- AnnotationDbi::select(edb,
                                    keys=AnnotationDbi::keys(edb),
       columns=c("GENEID","ENTREZID","SYMBOL"), 
       keytype = "GENEID") %>% 
  janitor::clean_names() %>% 
  mutate(genome=case_when(
    grepl("^ENSG[0-9]+", geneid) ~ "human",
    grepl("^ENSMUSG[0-9]+", geneid) ~ "mouse",
    grepl("^PF3D7_", geneid) ~ "plasmodium",
  )) %>% 
  left_join(., pf_names, by=c("geneid"="gene_id")) %>% 
  mutate_at(vars(symbol),~ifelse(is.na(symbol) & !is.na(gene_name), gene_name, .)) %>% 
  mutate_at(vars(symbol), ~toupper(.))

# human_mouse_genes %>% head()
# human_mouse_genes %>% tail()
# dim(human_mouse_genes)
table(human_mouse_genes$genome)

# annotation DBI clobbers the dplyr functions, so set those as default 
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("select","dplyr")
conflicted::conflict_prefer("arrange","dplyr")
conflicted::conflict_prefer("mutate","dplyr")
conflicted::conflict_prefer("intersect","base")
```


## AP2 Motif Set 

*PWMs:*
- PWM scores are defined as the log-ratio of the PPM and the background frequencies:
- the position weight matrix (PWM), which uses the PPM and a background model to get an idea of “expected” sequences, and thereby can assign weights to each letter at each position. 
- “Background” refers here to the base composition at non-specific sites (i.e. here, sequences that do not necessarily bind the TF).
- The background is often assumed to be either uniform (i.e. 1/4 for each of A, T, C, and G), or to be the genome-wide nucleotide frequencies. PWM scores are defined as the log-ratio of the PPM and the background frequencies

*ICMs*

- highlight conserved over non-conserved sites.
- The information content matrix (ICM) is a scaled version of the PPM, where each position gets scaled by its information content (IC) measured in bits.
- Most conserved sites have an IC of 2 bits.
- This form can be visualised as a “sequence logo”, 
  - Reference: https://ivanek.github.io/analysisOfGenomicsDataWithR/13_Motifs_html.html
  - https://bioconductor.org/packages/devel/bioc/vignettes/monaLisa/inst/doc/monaLisa.html

See: matchPWM()
https://rdrr.io/github/Bioconductor/Biostrings/man/matchPWM.html
https://stat.ethz.ch/pipermail/bioconductor/2008-April/022029.html

MotifDB
https://bioconductor.org/packages/3.15/bioc/html/MotifDb.html


### Supplemental Data

Dataset S1.  Position weight matrices for all AP2 domains that bound DNA in PBM experiments.
The data is from protein-binding microarray experiments performed using purified, GST-tagged AP2 domains.
The "Seed-and-Wobble" algorithm was applied to combine the data from two separate experiments and create position weight matrices (PWMs) (Berger et al. 2006).

  - An enrichment score cut-off of 0.45 was used to distinguish high affinity binding data from low affinity and non-specific binding.

  - Each column in the matrices represent the nucleotide position from 5' to 3'.
  - Reference: doi:10.1371/journal.ppat.1001165
  
```{bash, eval=FALSE}
cd references/AP2_motifs/
TFs=$(cat ppat.1001165.s019.csv | grep -E "^PF|^MAL" | tr -d "," )
for tf in $(echo "$TFs"); do grep -E -A 4 "$tf" ppat.1001165.s019.csv | grep -v $tf > ${tf}.csv  ; done
```

```{r}
tf_dfs <- 
  # dir("references/AP2_motifs", pattern = "^P", full.names = TRUE)
  dir("references/AP2_motifs", pattern = "^P|^M", full.names = TRUE)

print(tf_dfs)

AP2_PWMlist <- TFBSTools::PWMatrixList()
for(df in tf_dfs){
  motif_name <- str_split_fixed(df, pattern="/", n=3)[,3] %>% 
    gsub(".csv","", .)
  pwm <- read.delim(df, sep=",", header=FALSE,row.names = 1)
  rownames(pwm) <- gsub(":","",rownames(pwm))
  pwm <- pwm[,!sapply(pwm, function(x) any(is.na(x)))] 
  pwm <- TFBSTools::as.matrix(pwm)

  pwm_mat <- PWMatrix(ID=motif_name,
                      name=motif_name,
                      strand = "*",
                      profileMatrix = pwm)
  AP2_PWMlist[[motif_name]] <- pwm_mat
}
names(AP2_PWMlist) <- str_split_fixed(tf_dfs, pattern="/", n=3)[,3] %>%  gsub(".csv","", .)

# AP2_PWMlist
```

```{r eval=FALSE}
names(AP2_PWMlist) %>% 
  gsub("_D[0-9]|_DLD","",.) %>% 
  cat(., file="ap2_old_ids.txt", sep="\n")
```

```{r}
# maxScore(Matrix(AP2_PWMlist$PF07_0126_DLD))
sapply(AP2_PWMlist, function(x) maxScore(Matrix(x))) %>% 
  data.frame(score=.) %>% 
  rownames_to_column("motif") %>% 
  arrange(score) %>% 
  mutate(percent_75=score*0.75, 
         percent_90=score*0.9)
```

```{r}
# VERY Similar to the Motifs on the PlasmoDB, though PlasmoDB lacks a y-axis
# view_motifs(AP2_PWMlist$PF07_0126_DLD,
#             use.type='PWM')
# view_motifs(AP2_PWMlist$PF11_0404_D1, use.type = 'PWM')
```



### AP2 MotifDB

MotifDB:
 - More than 9900 annotated position frequency matrices (PFM) from 14 public sources, for multiple organisms.

```{r}
# can't get this to be a correct df without substantial cleaning, so NVM
# was going to use the motif counts to convert the PFMs to PWMs (the n=XXX paramater)
AP2_table_S2 <- tabulizer::extract_tables("references/AP2_motifs/AP2_TF_motif_counts.pdf",
                                           output="character")

# AP2_table_S2
```

```{r}
MotifDb@elementMetadata %>% 
  as.data.frame() %>% 
  filter(grepl("^P", organism)) %>% 
  group_by(organism) %>% 
  dplyr::count() %>% 
  arrange(desc(n))
```

```{r}
#The MotifDb query function performs a broad, case-neutral text search through all the metadata (all the annotation) for all of the motifs. It returns a list of matching motifs. More information about the metadata is provided below.
# pf_motifs <- query(MotifDb, andStrings=c("Pfalciparum"))
pf_motifs <- query(MotifDb, andStrings=c("vivax"))
pf_motif_meta <- pf_motifs@elementMetadata
# head(pf_motif_meta)

# PMID: 21060817 is the same data from the supplemental files from  Campbell et al,..,.Manuel Llinás, 2010
# table(pf_motif_meta$geneSymbol)
# table(pf_motif_meta$pubmedID)
table(pf_motif_meta$tfFamily) #AP2 and 2 NAs

#PFM position frequency matrix
# pf_motifs@listData %>% head(.,n=2)
```

```{r}
pf_motif_meta %>% 
  as.data.frame() %>% 
```

```{r}
# See also: convert_type(motif, type="PPM")
# this assumes a count of n=100 sequences, and I'm not sure if that is introducing too much bias
#could potentially use the data in the pdf supplemental table
ap2_res <- convert_motifs(pf_motifs, class="TFBSTools-PWMatrix")
names(ap2_res) <- names(pf_motifs@listData)
# head(ap2_res)

# view_motifs(ap2_res[1:3], dedup.names = TRUE)
# class(ap2_res[[1]])
```

```{r}
sel_motifs <- pf_motifs@elementMetadata %>% 
  as.data.frame() %>% 
  rownames_to_column("id") %>% 
  filter(pubmedID=="21060817")

pf_motifs_sel <- pf_motifs[sel_motifs$id]
ap2_PFM <- convert_motifs(pf_motifs_sel,class = "TFBSTools-PFMatrix")
ap2_PFMList <- do.call(PFMatrixList,ap2_PFM)
ap2_similarity <- motifSimilarity(ap2_PFMList)

dim(ap2_similarity)
# rowData(day2_enr$enrichement_res)$motif.pfm
# rowData(day2_enr$enrichement_res)$motif.pwm
```

```{r}
# pdf("figures/motif_analysis/ap2_motif_similarity_corrplot.pdf", height = 7, width = 7)
corrplot::corrplot(ap2_similarity)
# dev.off()
```


## AP2 Expression 

```{r}
ap2_expn <- expression[ap2_ids$gene_id,rownames(input_d456)] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to = "log2_CPM") %>% 
  left_join(., design_input, by="masigpro_id") %>% 
  left_join(., ap2_ids, by="gene_id") %>%
  mutate(new_rownames=paste(input_id,gene_id, sep=":")) %>% 
  select(gene_id,input_id, everything())


head(ap2_expn)
dim(ap2_expn) #252  10
```

```{r fig.height=10, fig.width=10, eval=FALSE}
# pdf("figures/combat_seq/AP2_gene_expression_freescales_boxplot.pdf", height = 15, width=10)
ggplot(ap2_expn, aes(x=analysis_factor_group, y=log2_CPM)) +
  geom_boxplot(aes(fill=analysis_factor_group),
               alpha=0.3,
               outlier.colour = NA) +
  geom_point(aes(color=analysis_factor_group),
             size=2, alpha=0.7) +
  facet_wrap(~new_rownames, scales = "free", ncol=3) +
  scale_fill_viridis_d(direction = -1) +
  scale_color_viridis_d(end=0.925,direction = -1) +
  theme_classic() +
  theme(legend.position = "top", 
        axis.text.x = element_text(angle=25, vjust=1, hjust=1))
# dev.off()
```


```{r}
id_map <- ap2_expn %>% 
  select(gene_id, input_id) %>% 
  distinct()

ap2_mat <- expression[ap2_ids$gene_id,rownames(input_d456)]
# identical(id_map$gene_id, rownames(ap2_mat)) #TRUE
rownames(ap2_mat) <- id_map$input_id

ap2_mat_rename <- ap2_mat
colnames(ap2_mat_rename) <- gsub("Pf_Day0","sporozoites",colnames(ap2_mat)) %>% 
  gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2", .)

# head(ap2_mat_rename)

new_pal <- circlize::colorRamp2(c(-4,-2, 0, 2, 4),
                      c("deepskyblue3", "deepskyblue","white", "red", "red3"),
                      space="sRGB")


sample_anno <- design_input %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  set_rownames(.$masigpro_id) %>% 
  as.data.frame()

gene_anno <- select(ap2_expn, 
                    gene_id=input_id,
                    gene_name_or_symbol) %>% 
  distinct() %>% 
  mutate(genome="PF3D7") %>% 
  as.data.frame() %>% 
  set_rownames(.$new_rownames)

s_ord <- c(paste0("sporozoites_Rep",1:5),
           # "sporozoites_Rep3", "sporozoites_Rep5", "sporozoites_Rep1", "sporozoites_Rep2", "sporozoites_Rep4",
           "Day4_Pf_LS_Rep3",  "Day4_Pf_LS_Rep1",  "Day4_Pf_LS_Rep2",
           "Day5_Pf_LS_Rep3",  "Day5_Pf_LS_Rep2", "Day5_Pf_LS_Rep1" , 
           "Day6_Pf_LS_Rep3",  "Day6_Pf_LS_Rep1" , "Day6_Pf_LS_Rep2")
```


```{r fig.height=10, fig.width=10, eval=FALSE}
conflicted::conflict_prefer("rowVars","matrixStats")
conflicted::conflicts_prefer(base::intersect)

heat_colors <- colorRamp2(c(-3,-2, -1, 0, 1, 2,3),c("deepskyblue4","deepskyblue2","deepskyblue","white", "red","red2","red4"), space="LAB")
# heat_colors <- colorRamp2(c(-3,-1,0,1,3),c("deepskyblue3","deepskyblue","white", "red", "red3"), space="LAB")
# 
ap2_complexHmap <- create_pf_heatmaps(TMMCPM = ap2_mat_rename, 
                   sample_annots = sample_anno,
                   gene_annots = gene_anno,
                   order_ids = s_ord,
                   heatmap_colors = heat_colors,
                   GOI = rownames(ap2_mat_rename),
                   center_scale = TRUE)


# png(paste0("figures/combat_seq/AP2_expression/AP2_heatmap_row_zscores_heatmap.png"),units="in", height = 15, width = 15, res=150)
# svglite(paste0("figures/combat_seq/AP2_expression/AP2_heatmap_row_zscores_heatmap.svg"),fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 15, width = 15)
ap2_complexHmap$heatmap + ap2_complexHmap$hmap_anno$geneLabels
# dev.off()
```


## AP2 monaLisa

calcBinnedMotifEnrR:

Motif enrichments are then calculated for each bin, normalizing for differences in sequence composition in a very similar way as originally done by Homer (Heinz et al. 2010).

We recommend using this function to do the binned motif enrichment analysis, since it corrects for sequence composition differences similarly to Homer, but is implemented more efficiently.

 we would recommend to use regions of similar or even equal lengths to avoid a length bias, for example by using a fixed-size region around the midpoint of each region of interest. 

GC-Bias:
GC-poorer than in other bins. A strong bias of this kind could give rise to false positives in that bin, e.g. enrichments of AT-rich motifs.
bias *may* be ignored because the built-in sequence composition correction, but check of AT rich motifs. 

Generally speaking, we recommend a minimum of ~100 sequences per bin as fewer sequences may lead to small motif counts and thus either small or unstable enrichment. 

So for the time course clusters, I'll keep them all as 1 bin

```{r}
# threshold for -log10(p) values 
# this is p ~= 0.06
threshold <- 1.2
```

```{r}
save_monaL_plots <- function(monalisa_res, prefix){
  
    pdf(paste0(prefix,"_GC_content.pdf"), height=7, width=7)
    print(plotBinDiagnostics(monalisa_res$target_seqs, monalisa_res$input_bins, aspect="GCfrac"))
    dev.off()
    
    pdf(paste0(prefix,"_bin_distribution.pdf"), height=7, width=7)
    print(plotBinDensity(mcols(monalisa_res$target_promoters)[[7]], monalisa_res$input_bins, legend = "topleft"))
    dev.off()
    
    
    maxEnr <- ceiling(max(assay(monalisa_res$enrichment_res, "log2enr"), na.rm=TRUE))
    maxSig <- ceiling(max(assay(monalisa_res$enrichment_res,"negLog10P"), na.rm = TRUE))
    
    pdf(paste0(prefix,"_log2Enr_p.value_heatmap.pdf"), height = 10, width = 14)
    plotMotifHeatmaps(x = monalisa_res$enrichment_res, 
                      which.plots = c("log2enr", "negLog10P"), 
                      width = 2.0,
                      cluster = TRUE,
                      maxEnr = maxEnr, 
                      maxSig = maxSig, 
                      show_motif_GC = TRUE)
    dev.off()
}
```


### Day 2

A negLog10 P or FDR > 1.3 will ~= 0.050118
A negLog10 P or FDR > 4.0 will ~= 0.0001


day2_enr <- readRDS("results/motif_analysis/monaLisa/day2_vs_spz_monaLisa_results.RDS")
day2_enr[["enrichment_res"]] <- day2_enr$enrichement_res # spelling error 
day2_AP2_results <- read.csv("results/motif_analysis/monaLisa/day2_vs_spz_monaLisa_results.csv")

```{r}
day2_de <- read.delim("results/DEGs/kappe_s_Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_DEGs.csv",
                      sep=",",comment.char = "#")
# head(day2_de)
dim(day2_de) #1978   20
```

```{r}
day2_enr <- readRDS("results/motif_analysis/monaLisa/day2_vs_spz_monaLisa_results.RDS")

# day2_enr <- motif_enrichment_workflow(targets_df = day2_de,
#                             col_to_bin = "logFC",
#                             PWMs = AP2_PWMlist,
#                             proms_gr = pf_proms_gr,
#                             genome_bs = genome,
#                             make_bins = TRUE,
#                             use_bg_regions = TRUE)

# saveRDS(day2_enr, "results/motif_analysis/monaLisa/day2_vs_spz_monaLisa_results.RDS")
```

```{r}
day2_AP2_results <- create_motif_df(motif_result_list = day2_enr, col_to_bin = "logFC")

dim(day2_AP2_results) #46
# write.csv(day2_AP2_results, "results/motif_analysis/monaLisa/day2_vs_spz_monaLisa_results.csv", row.names = FALSE)
```

```{r}
# assays(day2_enr$enrichement_res)
d2p <- save_monaL_plots(day2_enr, prefix = "figures/motif_analysis/Day2_vs_SPZ")
```

```{r}
p_vals <- assay(day2_enr$enrichment_res,"negLog10P")
idx <- which(p_vals[,1] > threshold | p_vals[,2] > threshold)

# pdf("figures/motif_analysis/day2_ap2_motif_heatmap.pdf", height = 5, width = 7)
plotMotifHeatmaps(x = day2_enr$enrichment_res[idx,], 
                  which.plots = c("log2enr"), 
                  width = 1.8, 
                  cluster = TRUE,
                  show_dendrogram = TRUE, 
                  width.seqlogo = 1,
                  show_seqlogo = FALSE)
# dev.off()
```

```{r}
day2_sig <- day2_AP2_results %>% 
  mutate(motif = as.factor(motif),
         logFC = as.factor(logFC)) %>% 
  group_by(motif,.drop = FALSE) %>% 
  filter(any(negLog10P >= threshold)) %>% 
  ungroup() %>% 
  mutate(log2enr_clean=ifelse(negLog10P >= threshold, log2enr, NA)) %>% 
  select(motif:log2enr, log2enr_clean, everything()) %>% 
  arrange(log2enr_clean) %>% 
  mutate(motif=factor(motif, levels = unique(motif)))

# day2_sig

day2_top_logos <- purrr::map(unique(day2_sig$motif), function(name){
    mat <- Matrix(AP2_PWMlist[[name]])
    if(ncol(mat) >= 20){
      lims <- c(4,21)
      breaks <- seq(5,20,by=5)
    }else{
      lims <- c(1,ncol(mat)+2)
      breaks <- seq(1,ncol(mat)+2,by=5)
    }
    ggseqlogo(mat) +
      labs(title=name) +
      suppressMessages(scale_x_continuous(expand = c(0,0), 
                         limits = lims,
                         breaks= breaks)) +
      theme(plot.title = element_text(size=14,vjust = 0.5, hjust = 0.5),
            axis.text = element_text(size=14),
            axis.title = element_text(size=14),
            plot.margin = margin())
  })

# length(day2_top_logos)
names(day2_top_logos) <- unique(day2_sig$motif)
```

```{r}
day2_enr <- ggplot(day2_sig, aes(y=motif, x=logFC, fill=log2enr_clean)) +
  geom_tile(color="black") +
  coord_fixed() +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(labels=c("upregulated","downregulated"), expand = c(0,0)) +
  scale_fill_viridis_c(na.value = 'white', option = "A", 
                       begin = 0.15, end = 0.9, direction = -1) +
  labs(y="") +
  theme_classic() +
  theme(axis.text = element_text(size=12),
        axis.title = element_text(size=14))

# day2_enr
```


```{r fig.height=10, fig.width=10}
layout_rows <- "
AB
AC
AD
AE
AF
"

# wdts <- c(0.6,0.6,0.75)
#matrix(data=c(rep(1,5), rep(0.5,5)), nrow = 5, ncol = 2)
# hts <- c(rep(0.5,5), 1)

# svglite("figures/motif_analysis/day2_ap2_motif_seqlogos_heatmap.svg", height = 7, width=11,  system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
day2_enr + 
  day2_top_logos$PF10_0075_D3 +
  day2_top_logos$PF13_0097 +
  day2_top_logos$PFE0840c_D2 +
  day2_top_logos$PF14_0533 + 
  day2_top_logos$PFD0985w_D2 + 
  plot_layout(design = layout_rows,
              guides = 'collect')
# dev.off()
```




### Time-Course 

NOTE: the BiocParallel framework does not utilized the seed in set.seed(), so the results of this with the time-course clusters lacks the reproducibility of setting RNGseed in biocparallel. 

Now the question is - how does one select the most appropriate seed? Or simply forget p-values and select hits as those that are present in top sig hits X% of the time when varying the seed? Average the p-values and log2 enrichment values?

is this a function of the small genome?? that the random sampling does appear to have an effect on the results. 

```{r}
tc_clusters <- time_course %>% 
  mutate(gene_cluster = paste0("cluster", gene_cluster) %>% as.factor()) %>% 
  select(gene_id, gene_cluster)

tc_profiles <- time_course %>% 
  mutate(combined_gene_clusters=as.factor(combined_gene_clusters)) %>% 
  select(gene_id, combined_gene_clusters) %>% 
  arrange(combined_gene_clusters)
```

```{r}
cluster_motifs <- readRDS("results/motif_analysis/monaLisa/time_course_ap2_results_list.RDS")

profile_motifs <- readRDS("results/motif_analysis/monaLisa/time_course_profiles_ap2_results_list.RDS")
```

```{r eval=FALSE}
cluster_nums <- unique(tc_clusters$gene_cluster)
cluster_motifs <- list()
for (cluster in cluster_nums){
  
  input <- tc_clusters[tc_clusters$gene_cluster == cluster, ] %>% 
    droplevels()

  enr <- motif_enrichment_workflow(targets_df = input,
                            col_to_bin = "gene_cluster",
                            PWMs = AP2_PWMlist,
                            proms_gr = pf_proms_gr,
                            genome_bs = genome,
                            make_bins = FALSE,
                            use_bg_regions = TRUE)
  
  cluster_motifs[[cluster]] <- enr
}

# saveRDS(cluster_motifs,"results/motif_analysis/monaLisa/time_course_ap2_results_list.RDS")

profile_nums <- unique(tc_profiles$combined_gene_clusters)
profile_motifs <- list()
for (profile in profile_nums){
  
  input <- tc_profiles[tc_profiles$combined_gene_clusters == profile, ] %>% 
    droplevels()

  enr <- motif_enrichment_workflow(targets_df = input,
                            col_to_bin = "combined_gene_clusters",
                            PWMs = AP2_PWMlist,
                            proms_gr = pf_proms_gr,
                            genome_bs = genome,
                            make_bins = FALSE,
                            use_bg_regions = TRUE)
  
  profile_motifs[[profile]] <- enr
}

# saveRDS(profile_motifs, "results/motif_analysis/monaLisa/time_course_profiles_ap2_results_list.RDS")
```

```{r}
# Clusters 
with_prom_bg <- purrr::map_dfr(cluster_motifs, create_motif_df)
# write.csv(with_prom_bg,"results/motif_analysis/monaLisa/time_course_ap2_results.csv", row.names = FALSE)

with_prom_bg %>%
  filter(negLog10P > 1.2) %>%
  group_by(gene_cluster) %>%
  dplyr::slice(1)

# Profiles 
profiles_with_prom_bg <- purrr::map_dfr(profile_motifs, create_motif_df, col_to_bin="combined_gene_clusters")
# write.csv(profiles_with_prom_bg, "results/motif_analysis/monaLisa/time_course_profiles_ap2_results.csv", row.names=FALSE)

profiles_with_prom_bg %>% 
  filter(negLog10P > 1.2) %>% 
  group_by(combined_gene_clusters) %>% 
  dplyr::slice(1)
```

### Count Frequencies and hits 

```{r}
cluster_hits <- purrr::map_dfr(names(cluster_motifs), function(motif_res){
  
  hits <- findMotifHits(query = AP2_PWMlist,
              subject = cluster_motifs[[motif_res]]$target_seqs,
              method = "matchPWM",
              min.score = 10)
  hits %>%
    as.data.frame() %>%
    group_by(pwmid) %>%
    summarise(mean_score=mean(score),
              transcript_hits = paste0(unique(seqnames) %>% as.character(), collapse = "; "),
              .groups = 'keep') %>%
    ungroup() %>%
    mutate(gene_cluster=motif_res) %>% 
    select(motif=pwmid,
           gene_cluster,
           transcript_hits) 
})

# cluster_hits
```

```{r}
cluster_hits_enr <- with_prom_bg %>% 
    left_join(., cluster_hits,
              by=c("motif", "gene_cluster")) %>% 
    select(motif,gene_cluster:expForegroundWgtWithHits,
           sumForegroundWgtWithHits:sumBackgroundWgtWithHits,
           everything()) %>%
    arrange(gene_cluster, desc(negLog10P))
    
dim(cluster_hits_enr)
# write.csv(cluster_hits_enr, "results/motif_analysis/monaLisa/time_course_ap2_withTranscriptIDs_results.csv", row.names = FALSE)
```

```{r}
motif_hits_profiles <- purrr::map_dfr(names(profile_motifs), function(motif_res){
  print(motif_res)
  hits <- findMotifHits(query = AP2_PWMlist,
              subject = profile_motifs[[motif_res]]$target_seqs,
              method = "matchPWM",
              min.score = 10)
  hits %>%
    as.data.frame() %>%
    group_by(pwmid) %>%
    summarise(mean_score=mean(score),
              transcript_hits = paste0(unique(seqnames) %>% as.character(), collapse = "; "),
              .groups = 'keep') %>%
    ungroup() %>%
    mutate(combined_gene_clusters=motif_res) %>% 
    select(motif=pwmid,
           combined_gene_clusters,
           transcript_hits)
})
```

```{r}
profile_hits_enr <- profiles_with_prom_bg %>% 
    left_join(., motif_hits_profiles,
              by=c("motif", "combined_gene_clusters")) %>% 
    select(motif,combined_gene_clusters:expForegroundWgtWithHits,
           sumForegroundWgtWithHits:sumBackgroundWgtWithHits,
           everything()) %>%
    arrange(combined_gene_clusters, desc(negLog10P))

dim(profile_hits_enr)
head(profile_hits_enr)

# write.csv(profile_hits_enr, "results/motif_analysis/monaLisa/time_course_profiles_ap2_withTranscriptIDs_results.csv", row.names = FALSE)
```

### Visualization 

#### Time course Clusters 

```{r}
with_prom_bg <- read.csv("results/motif_analysis/monaLisa/time_course_ap2_results.csv")
profiles_with_prom_bg <- read.csv("results/motif_analysis/monaLisa/time_course_profiles_ap2_results.csv")
```

```{r}
# quick reminder -log10(p) > threshold
print(threshold)
```

```{r}
sel_tc_hits <- with_prom_bg %>% 
  
  mutate(gene_cluster_factor=factor(gsub("cluster", "", gene_cluster), 
                                    levels=c(1,2,3,8,4,6,5,9,7))) %>% 
  mutate(is_significant=ifelse(negLog10P > 1.3, "p < 0.05", "NS"), 
         label = gene_cluster,
         old_gene_id = gsub("_D[0-9]|_DLD","",motif)) %>%
  
  group_by(gene_cluster) %>% 
  mutate(plot_fill=case_when(
    negLog10P > threshold ~ log2enr,
    TRUE ~ NA_real_)) %>% 
  ungroup() %>% 
  
  group_by(motif) %>% 
  filter(!all(is.na(plot_fill))) %>% 
  ungroup() %>% 
  
  mutate(motif_factor_groups=paste0(label,":",motif)) %>% 
  group_by(gene_cluster) %>%
  # arrange by desc log2 enrichment values 
  arrange(desc(plot_fill), .by_group = T) %>% 
  ungroup() %>% 
  mutate(motif_factor_groups=factor(motif_factor_groups, levels=rev(motif_factor_groups))) %>% 
  select(gene_cluster,
         gene_cluster_factor,
         motif_factor_groups,
         plot_fill,log2enr, everything())


sel_tc_hits %>% head()
dim(sel_tc_hits)
length(unique(sel_tc_hits$motif)) #15
```

```{r fig.height=4, fig.width=10}
vals <- sel_tc_hits %>% 
  mutate(plot_fill=ifelse(is.na(plot_fill), 0, plot_fill)) %>% 
  pull(plot_fill, name=motif)

clust_motifs <- hclust(dist(vals, method="euclidean"))
plot(clust_motifs)

order_motifs <- c("PF13_0026") %>% 
  c(., clust_motifs$labels[clust_motifs$order][1:2]) %>% 
  c(.,  tail(clust_motifs$labels[clust_motifs$order], n=20)) %>% 
  unique()

length(order_motifs) #15
table(order_motifs %in% unique(sel_tc_hits$motif))
table(unique(sel_tc_hits$motif) %in% order_motifs)
```

```{r}
top_motifs_per_clust_df <- sel_tc_hits %>% 
  group_by(gene_cluster) %>% 
  dplyr::slice(1) 

# dim(top_motifs_per_clust_df)

ap2_expn_top_hits <- purrr::map(top_motifs_per_clust_df$gene_cluster_factor,function(x){
  subset <- top_motifs_per_clust_df %>% 
    filter(gene_cluster_factor==x)

  gene <- subset[["old_gene_id"]]
  # grep(gene,ap2_expn$input_id, value=T)
  ap2_expn %>%
    filter(grepl(gene,input_id)) %>%
    mutate(gene_cluster=x,
           motif=subset[["motif"]]) %>%
    select(1:2,gene_cluster,motif, everything())
})
# ap2_expn_top_hits
names(ap2_expn_top_hits) <- top_motifs_per_clust_df$motif_factor_groups

# lapply(ap2_expn_top_hits, function(x) table(x$gene_id, x$input_id))
# lapply(ap2_expn_top_hits, function(x) table(x$motif, x$input_id))
```


```{r}
ap2_summary_stats <- ap2_expn_top_hits %>% 
  bind_rows() %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>%
  group_by(gene_cluster_factor, analysis_factor_group,motif) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd,
         stat="transcription factor",
        .groups = 'keep') %>% 
  ungroup() %>% 
  select(gene_cluster_factor, analysis_factor_group,
         motif,masigpro_id,
         stat,ymin:ymax, mean=log2_CPM) 

tc_summary_stats <- expn_d456 %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>%
  group_by(gene_cluster_factor, analysis_factor_group) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd) %>% 
  group_by(gene_cluster_factor,analysis_factor_group, masigpro_id) %>% 
  mutate(stat="time-course cluster",
          mean=mean(log2_CPM),
          motif="time-course cluster",
         .groups = 'keep') %>% 
  ungroup() %>%
  select(gene_cluster_factor, analysis_factor_group, masigpro_id,
         motif, stat, mean, ymin:ymax, label) %>% 
  distinct() %>% 
  bind_rows(ap2_summary_stats) %>% 

  group_by(gene_cluster_factor) %>%
  mutate(label=as.character(label)) %>%
  mutate_at(vars(label), ~ifelse(is.na(.), label[!is.na(label)], .)) %>%
  ungroup() %>% 
  
  group_by(gene_cluster_factor,analysis_factor_group,stat) %>%
  mutate(middle_val=median(mean)) %>%
  ungroup() %>%

  arrange(gene_cluster_factor,analysis_factor_group) %>%
  mutate(label=factor(label, levels=unique(label)))
```

```{r fig.height=10, fig.width=10}
n <- length(unique(sel_tc_hits$motif))
motif_colors <-  rep("grey50", n) %>% 
  set_names(unique(sel_tc_hits$motif))
motif_colors[c("time-course cluster")] <- "green2"

# barplot(rep(1,length(motif_colors)), col=motif_colors, names.arg = names(motif_colors), las=2, cex.names = 0.7)

# Update the labels for the manuscript
tc_lab_summary_stats <- tc_summary_stats %>% 
    mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
    mutate(analysis_factor_group = gsub("_Rep[0-9]", "", masigpro_id) %>% 
                                    factor(., levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))
           ) %>% 
  select(analysis_factor_group, everything())

expn_labs_d456 <- expn_d456 %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  select(analysis_factor_group, everything())

# Generate individual plots
tc_AP2_plots <- create_motif_enr_plots(enr_df = sel_tc_hits,
                       factor_column="gene_cluster_factor",
                       motif_order =  order_motifs,
                       summary_stats = tc_lab_summary_stats,
                       expn_df_long = expn_labs_d456,
                       PWMlist = AP2_PWMlist,
                       motif_colors = motif_colors)

layout_rows <- "
ABC
ABD
ABE
ABF
ABG
ABH
ABI
ABJ
ABK
"

wdts <- c(0.6,0.6,0.75)


# svglite("figures/motif_analysis/time_point_cluster_heatmap_seqlogos_tileplot_rows.svg", height = 20, width=20,
# system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
tc_AP2_plots$time_cluster + 
  tc_AP2_plots$motif_enr + #theme(axis.text.x = element_text(size=7, angle=45)) +
  tc_AP2_plots$top_logos[[1]] +
  tc_AP2_plots$top_logos[[2]] +
  tc_AP2_plots$top_logos[[3]] +
  tc_AP2_plots$top_logos[[4]] + 
  tc_AP2_plots$top_logos[[5]] + 
  tc_AP2_plots$top_logos[[6]] +
  tc_AP2_plots$top_logos[[7]] + 
  tc_AP2_plots$top_logos[[8]] + 
  tc_AP2_plots$top_logos[[9]] +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()


# svglite("figures/motif_analysis/time_point_cluster_ap2_heatmap_seqlogos_tileplot_rows.svg", height = 20, width=20,
# system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
tc_AP2_plots$time_ribbon + 
  tc_AP2_plots$motif_enr + #theme(axis.text.x = element_text(size=7, angle=45)) +
  tc_AP2_plots$top_logos[[1]] +
  tc_AP2_plots$top_logos[[2]] + 
  tc_AP2_plots$top_logos[[3]] + 
  tc_AP2_plots$top_logos[[4]] + 
  tc_AP2_plots$top_logos[[5]] +
  tc_AP2_plots$top_logos[[6]] + 
  tc_AP2_plots$top_logos[[7]] + 
  tc_AP2_plots$top_logos[[8]] + 
  tc_AP2_plots$top_logos[[9]] +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```

```{r fig.height=10, fig.width=10}
hits_for_lineplot <- sel_tc_hits  %>% 
  filter(!is.na(plot_fill)) %>% 
  mutate(input_id=gsub("_D[12L].{0,}", "", motif)) %>% 
  select(1:2, input_id, everything())

summary_stats_means <- tc_summary_stats %>% 
  filter(motif=="time-course cluster") %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0", "sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = gsub("_Rep[0-9]", "", masigpro_id) %>% 
                                    factor(., levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  mutate(gene_id=motif, 
         label=as.character(label)) %>% 
  select(gene_id, gene_cluster_factor, 
         analysis_factor_group,
         motif,
         mean=middle_val, label) %>% 
  distinct()

temp_summ <- ap2_expn %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0", "sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = gsub("_Rep[0-9]", "", masigpro_id) %>% 
                                    factor(., levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  inner_join(., hits_for_lineplot, by="input_id") %>%
  arrange(gene_cluster_factor) %>%
  group_by(gene_cluster_factor, gene_id, motif, analysis_factor_group) %>% 
  summarize(mean=mean(log2_CPM), .groups = 'keep') %>% 
  ungroup() %>% 
  bind_rows(summary_stats_means) %>% 
  group_by(gene_cluster_factor) %>% 
  mutate(label=ifelse(is.na(label), unique(label[!is.na(label)]), label)) %>% 
  ungroup() %>% 
  mutate(label=factor(label, levels=unique(label)))

# temp_summ

tf_genes <- unique(temp_summ$gene_id[grep("PF", temp_summ$gene_id)])
color_code <- viridis::inferno(n = length(tf_genes), end=0.925,direction = -1) %>% 
  set_names(., tf_genes) %>% 
  c(., "time-course cluster"='grey70')

gene_labels <- temp_summ %>% 
  select(motif, gene_id) %>% 
  unique() %>% 
  pull(motif, name = gene_id) %>% 
  .[names(color_code)]

max_val <- floor(max(temp_summ$mean))
tc_line_plot_TFs <- ggplot(temp_summ %>% mutate(gene_id=factor(gene_id, levels = unique(gene_id))), 
                           aes(x=analysis_factor_group, y=mean)) +
  # motif expression over time 
  geom_line(mapping = aes(group=gene_id, 
                          color=gene_id), 
            linetype = 'dashed') +
  geom_point(aes(shape = analysis_factor_group, color=gene_id, fill=gene_id),
                 # fill=analysis_factor_group), 
             # color='white', 
             size=3, alpha=0.75) +
  facet_wrap(~label, ncol = 2, scales = 'free_x', 
             strip.position = "left") +
  scale_x_discrete(expand=c(0.05,0.05)) +
  scale_y_continuous(breaks = seq(0, max_val, by=2)) +
  #color and shape and theme
  scale_shape_manual(values = c(21,22,23,24,25)) +
  scale_fill_viridis_d(end=0.925,direction = -1) +
  scale_color_manual(values=color_code, labels=gene_labels) +
  scale_fill_manual(values=color_code, labels=gene_labels) +
  labs(y="mean log2(CPM+1)", x = '') +
  theme_classic() +
  theme(strip.placement = 'outside', 
        axis.text.x = element_text(angle = 15, vjust =1, hjust =1 ))

# svg("figures/motif_analysis/time_point_cluster_ap2_enriched_motifs_linePlot.svg", height = 14, width = 10)
tc_line_plot_TFs
# dev.off()
```


#### Time course Profiles 

```{r}
# -log10(p) > 1.2
print(threshold)
```

```{r}
sel_profile_hits <- profiles_with_prom_bg %>% 
  
  mutate(gene_cluster_factor=as.factor(combined_gene_clusters), 
         gene_cluster=combined_gene_clusters) %>% 
  mutate(is_significant=ifelse(negLog10P > 1.3, "p < 0.05", "NS"), 
         label = gene_cluster,
         old_gene_id = gsub("_D[0-9]|_DLD","",motif)) %>%
  
  group_by(gene_cluster) %>% 
  mutate(plot_fill=case_when(
    negLog10P > threshold ~ log2enr,
    TRUE ~ NA_real_)) %>% 
  ungroup() %>% 
  
  group_by(motif) %>% 
  filter(!all(is.na(plot_fill))) %>% 
  ungroup() %>% 
  
  mutate(motif_factor_groups=paste0(label,":",motif)) %>% 
  group_by(gene_cluster) %>%
  arrange(desc(plot_fill), .by_group = T) %>% 
  ungroup() %>% 
  mutate(motif_factor_groups=factor(motif_factor_groups,
                                    levels=rev(motif_factor_groups))) %>% 
  select(gene_cluster,
         gene_cluster_factor,
         motif_factor_groups,
         plot_fill,log2enr, everything())


dim(sel_profile_hits)
# head(sel_profile_hits)
# table(sel_profile_hits$gene_cluster, sel_profile_hits$is_significant) %>% as.data.frame()

# sel_profile_hits %>%
#   group_by(gene_cluster, is_significant) %>%
#   dplyr::count() %>%
#   pivot_wider(names_from = is_significant,
#               values_from = n)
```

```{r fig.height=4, fig.width=10}
vals <- sel_profile_hits %>% 
  mutate(plot_fill=ifelse(is.na(plot_fill), 0, plot_fill)) %>% 
  pull(plot_fill, name=motif)

clust_profile_motifs <- hclust(dist(vals, method="euclidean"))
# plot(clust_profile_motifs)
```

```{r}
order_motifs_profile <- tail(clust_profile_motifs$labels[clust_profile_motifs$order], n=15) %>% 
  c(., clust_profile_motifs$labels[clust_profile_motifs$order][1:3]) %>% 
  unique()

ap2_expn <- expression[ap2_ids$gene_id,rownames(input_d456)] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to = "log2_CPM") %>% 
  left_join(., design_input, by="masigpro_id") %>% 
  left_join(., ap2_ids, by="gene_id") %>%
  select(gene_id,input_id, everything())


head(ap2_expn)
dim(ap2_expn) #252  10
```

```{r}
top_motifs_per_profile_df <- sel_profile_hits %>% 
  group_by(gene_cluster) %>% 
  dplyr::slice(1) 

ap2_expn_profiles <- purrr::map(top_motifs_per_profile_df$gene_cluster,function(x){
  subset <- top_motifs_per_profile_df %>% 
    filter(gene_cluster==x)

  gene <- subset[["old_gene_id"]]
  # grep(gene,ap2_expn$input_id, value=T)
  ap2_expn %>%
    filter(grepl(gene,input_id)) %>%
    mutate(gene_cluster=x,
           motif=subset[["motif"]]) %>%
    select(1:2,gene_cluster,motif, everything())
})
# ap2_expn_top_hits
names(ap2_expn_profiles) <- top_motifs_per_profile_df$motif_factor_groups

# lapply(ap2_expn_profiles, function(x) table(x$gene_id, x$input_id))
# lapply(ap2_expn_profiles, function(x) table(x$motif, x$input_id))
```

```{r}
profile_labs <- time_course %>% 
  group_by(combined_gene_clusters) %>% 
  mutate(label=paste0(combined_gene_clusters,"\n(",length(unique(gene_id))," genes)")) %>% 
  ungroup() %>% 
  select(combined_gene_clusters, label) %>% 
  distinct()

# head(profile_labs)

ap2_summary_stats <- ap2_expn_profiles %>% 
  bind_rows() %>% 
  mutate(gene_cluster_factor=as.factor(gene_cluster)) %>% 
  group_by(gene_cluster_factor, analysis_factor_group,motif) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd,
         stat="transcription factor",
        .groups = 'keep') %>% 
  ungroup() %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  select(gene_cluster_factor, analysis_factor_group,
         motif,masigpro_id,
         stat,ymin:ymax, mean=log2_CPM) 


profile_expn_d456 <- expn_d456 %>% 
  select(-label) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  mutate(gene_cluster_factor=as.factor(combined_gene_clusters)) %>% 
  left_join(., profile_labs, by=c("combined_gene_clusters")) %>% 
  
  group_by(gene_cluster_factor, analysis_factor_group) %>%
  mutate(median_val=median(log2_CPM),
         mean_val=mean(log2_CPM)) %>%
  ungroup() %>%
  group_by(gene_cluster_factor, analysis_factor_group, masigpro_id) %>%
  mutate(median_val_per_sample=median(log2_CPM),
         mean_val_per_sample=mean(log2_CPM)) %>%
  ungroup() %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS")))

# head(profile_expn_d456)


profile_summary_stats <- expn_d456 %>% 
  select(-label) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  mutate(gene_cluster_factor=as.factor(combined_gene_clusters)) %>% 
  left_join(., profile_labs, by=c("combined_gene_clusters")) %>% 
  
  group_by(gene_cluster_factor,
           analysis_factor_group) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd) %>% 
  
  group_by(gene_cluster_factor,
           analysis_factor_group,
           masigpro_id) %>% 
  mutate(stat="time-course profile",
          mean=mean(log2_CPM),
          motif="time-course profile",
         .groups = 'keep') %>% 
  ungroup() %>%
  select(gene_cluster_factor,
         analysis_factor_group, 
         masigpro_id,
         motif, stat, mean, ymin:ymax, label) %>% 
  distinct() %>% 
  bind_rows(ap2_summary_stats) %>% 
  
  group_by(gene_cluster_factor) %>% 
  mutate(label=as.character(label)) %>% 
  mutate_at(vars(label), ~ifelse(is.na(.), label[!is.na(label)], .)) %>% 
  ungroup() %>% 
  
  group_by(gene_cluster_factor,analysis_factor_group,stat) %>% 
  mutate(middle_val=median(mean)) %>% 
  ungroup() %>% 
  
  arrange(gene_cluster_factor,analysis_factor_group) %>% 
  mutate(label=factor(label, levels=unique(label))) %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   
                            gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS")))


# summary_stats

n <- length(unique(sel_profile_hits$motif))
motif_colors <-  rep("grey50", n) %>% 
  set_names(unique(sel_profile_hits$motif))
motif_colors[c("time-course profile")] <- "green2"

profile_AP2_plots <- create_motif_enr_plots(enr_df = sel_profile_hits,
                       factor_column="gene_cluster_factor",
                       motif_order =  order_motifs_profile,
                       summary_stats = profile_summary_stats,
                       expn_df_long = profile_expn_d456,
                       PWMlist = AP2_PWMlist,
                       motif_colors = motif_colors)
```

```{r fig.height=10, fig.width=10}
layout_rows <- "
ABC
ABD
ABE
ABF
ABG
"

wdts <- c(0.6,0.6,0.75)



# svglite("figures/motif_analysis/time_point_profile_heatmap_seqlogos_tileplot_rows.svg", height = 16, width=20, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
profile_AP2_plots$time_cluster +
  profile_AP2_plots$motif_enr + 
  profile_AP2_plots$top_logos[[1]] +
  profile_AP2_plots$top_logos[[2]] + 
  profile_AP2_plots$top_logos[[3]] + 
  profile_AP2_plots$top_logos[[4]] + 
  profile_AP2_plots$top_logos[[5]] + 
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()

wdts <- c(0.7,0.85,0.75)

# svglite("figures/motif_analysis/time_point_profile_TFexpn_ap2_heatmap_seqlogos_tileplot_rows.svg", height = 16, width=20, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
profile_AP2_plots$time_ribbon + 
  profile_AP2_plots$motif_enr + #theme(axis.text.x = element_text(size=7, angle=25)) +
  profile_AP2_plots$top_logos[[1]] +
  profile_AP2_plots$top_logos[[2]] + 
  profile_AP2_plots$top_logos[[3]] + 
  profile_AP2_plots$top_logos[[4]] +
  profile_AP2_plots$top_logos[[5]] +  
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```

```{r fig.height=10, fig.width=10}
hits_for_lineplot <- sel_profile_hits  %>% 
  filter(!is.na(plot_fill)) %>% 
  mutate(input_id=gsub("_D[12L].{0,}", "", motif)) %>% 
  select(1:2, input_id, everything())

summary_stats_means <- profile_summary_stats %>% 
  filter(motif=="time-course profile") %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0", "sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = gsub("_Rep[0-9]", "", masigpro_id) %>% 
                                    factor(., levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  mutate(gene_id=motif, 
         label=as.character(label)) %>% 
  select(gene_id, gene_cluster_factor, 
         analysis_factor_group,
         motif,
         mean=middle_val, label) %>% 
  distinct()

temp_summ <- ap2_expn %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0", "sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = gsub("_Rep[0-9]", "", masigpro_id) %>% 
                                    factor(., levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  inner_join(., hits_for_lineplot, by="input_id") %>%
  arrange(gene_cluster_factor) %>%
  group_by(gene_cluster_factor, 
           gene_id, 
           motif,
           analysis_factor_group) %>% 
  summarize(mean=mean(log2_CPM), .groups = 'keep') %>% 
  ungroup() %>% 
  bind_rows(summary_stats_means) %>% 
  group_by(gene_cluster_factor) %>% 
  mutate(label=ifelse(is.na(label), unique(label[!is.na(label)]), label)) %>% 
  ungroup() %>% 
  mutate(label=factor(label, levels=unique(label)))

# temp_summ
tf_genes <- unique(temp_summ$gene_id[grep("PF", temp_summ$gene_id)])
color_code <- viridis::inferno(n = length(tf_genes), end=0.925,direction = -1) %>% 
  set_names(., tf_genes) %>% 
  c(., "time-course profile"='grey70')

gene_labels <- temp_summ %>% 
  select(motif, gene_id) %>% 
  unique() %>% 
  pull(motif, name = gene_id) %>% 
  .[names(color_code)]

max_val <- floor(max(temp_summ$mean))
profile_line_plot_TFs <- ggplot(temp_summ %>% mutate(gene_id=factor(gene_id, levels = unique(gene_id))), 
                                aes(x=analysis_factor_group, y=mean)) +
  # motif expression over time 
  geom_line(aes(group=gene_id, color=gene_id), 
            linetype = 'dashed') +
  geom_point(aes(shape = analysis_factor_group, color=gene_id, fill=gene_id),
             size=3, alpha=0.75) +
  facet_wrap(~label, ncol = 2, scales = 'free_x', 
             strip.position = "left") +
  scale_x_discrete(expand=c(0.05,0.05)) +
  scale_y_continuous(breaks = seq(0,max_val, by=2)) +
  #color and shape and theme
  scale_shape_manual(values = c(21,22,23,24,25)) +
  scale_color_manual(values=color_code, labels = gene_labels) +
  scale_fill_manual(values=color_code, labels = gene_labels) +
  labs(y="mean log2(CPM+1)", x = '') +
  theme_classic() +
  theme(strip.placement = 'outside', 
        axis.text.x = element_text(angle = 15, vjust =1, hjust =1 ))

# svg("figures/motif_analysis/time_point_profile_ap2_enriched_motifs_linePlot.svg", height = 10, width = 12)
profile_line_plot_TFs
# dev.off()
```



## Homer Command line 


**Important** Homer is installed locally in my home drive. The PF3D7 database was generated using file paths in create_db.sh for the GTF and FASTA that pointed to this analysis directory. 

Later, this analysis directory was moved/reorganized and the paths for GTF and FASTA changed. Later, when re-running HOMER using the custom database, there was a recurring error:
`Progress: Step 5 - adjusting background sequences for GC/CpG content... Illegal division by zero at assignGeneWeights.pl line 63.` 

Rebuilding the database with update paths for GTF and FASTA appears to resolve the issue... so in the future - all custom DBs should have copies of GFT and FASTA files INSIDE the homer directory, and never move them? 


```{r eval=FALSE}
dir.create("results/motif_analysis/homer/day2/up",recursive = T,showWarnings = FALSE)
dir.create("results/motif_analysis/homer/day2/down",recursive = T, showWarnings = FALSE)

Day2 %>%
  dplyr::filter(logFC > 0) %>%
  pull(gene_id) %>%
  write.table(.,"results/motif_analysis/homer/day2/up/Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_upregulated.csv",
              sep="\t", quote=FALSE, row.names = FALSE)

Day2 %>%
  dplyr::filter(logFC < 0) %>%
  pull(gene_id) %>%
  write.table(.,"results/motif_analysis/homer/day2/down/Day2_Heps_Pf_vs_sporozorite_combat-seq_corrected_downregulated.csv",
              sep="\t", quote=FALSE, row.names = FALSE)
```


```{bash eval=FALSE}
cd motif_analysis/homer/
./create_db.sh #only run 1x
qsub homer_motif_analysis.pbs 
```

https://www.rdocumentation.org/packages/TFBSTools/versions/1.10.3/topics/getMatrixSet
name: The name of the transcription factor. As far as possible, the name is based on the standardized Entrez gene symbols. 
family: Structural sub-class of the transcription factor, based on the TFCaT system.
Acc: A representative protein accession number in Genbank for the transcription factor. Human takes precedence if several exists. 

## Orthologous Hits


```{r}
orthoGroups <- read.csv("resources/gene_annots/PF3D7_OrthoMCL_IDs_Summary.csv") %>% 
  janitor::clean_names() %>% 
  filter(!grepl("N\\/A", ortholog_group)) %>% 
  dplyr::select(ortholog_group,gene_id, uni_prot_id_s, everything())
  
head(orthoGroups)
dim(orthoGroups) #5389   12

# write.table(orthoGroups, "PF3D7_ortholog_group_IDs.txt", sep = "\t")
```

```{r}
orthoGroups_species <- read.delim("results/motif_analysis/orthogroups_pf_to_all_species_v2.txt", sep = "\t") %>% 
  janitor::clean_names() %>% 
  filter(!grepl("-old", accession)) %>% 
  mutate(taxon_clean=gsub("^([A-Z][a-z]+\\s[a-z]+)\\s.+", "\\1", taxon_name),
         gene_id=gsub("^[a-z].+\\|","", accession)) %>% 
  dplyr::select(accession,gene_id,
                taxon_name, taxon_clean, 
         everything())

head(orthoGroups_species)
length(unique(orthoGroups_species$taxon_clean)) #554 
# dim(orthoGroups_species) #1,533,059
```


### Uniprot IDs

Required the newest version of R Bioconductor for Uniprot.ws to work properly. 

```
Rscript pbs_jobs/uniprot.ws_lookup.R
```

```{r}
ortholog_search_res <- dir("results/motif_analysis/uniprot_ids",
                           pattern = "*.RDS", 
                           full.names = TRUE)

ortholog_uniprotIDs <- purrr::map_dfr(ortholog_search_res, function(df){
  readRDS(df) %>% 
    janitor::clean_names() %>% 
    distinct()
}) %>% 
  inner_join(., orthoGroups_species, by=c("from"="gene_id")) %>% 
  distinct()


dim(ortholog_uniprotIDs) #772,570     18
head(ortholog_uniprotIDs)
``` 

```{r}
ortholog_uniprotIDs %>% 
  group_by(organism) %>% 
  dplyr::count() %>% 
  nrow() # 571 taxons 
```


### Motifs with possible Orthologs

```{r eval=FALSE}
library(JASPAR2022)
opts <- list(collection="CORE",
             matrixtype="PWM")

opts2 <- list(collection="CORE",
              matrixtype="PFM")
pwm <- TFBSTools::getMatrixSet(JASPAR2022, opts)
pfm <- TFBSTools::getMatrixSet(JASPAR2022, opts2)

length(pwm) #1956
length(pfm) #1956
# saveRDS(pwm, results/motif_analysis/JASPAR2022_PWMs.RDS")
# saveRDS(pfm, results/motif_analysis/JASPAR2022_PFMs.RDS")
```

```{r}
all_PWMs <- readRDS("results/motif_analysis/JASPAR2022_PWMs.RDS")
all_PFMs <- readRDS("results/motif_analysis/JASPAR2022_PFMs.RDS")

length(all_PFMs) #1956
```

```{r}
motif_annots <- tags(all_PWMs) %>% 
  c(name(all_PWMs)) %>% 
  c(matrixClass(all_PWMs) %>%
      set_names(paste0(names(matrixClass(all_PWMs)), ".class"))) %>% 
  unlist() %>% 
  data.frame(values=.)  %>% 
  rownames_to_column("motif") %>% 
  arrange(motif) %>%
  mutate(info=str_split_fixed(motif, "\\.", n=3)[,3]) %>%
  mutate(info=ifelse(info=="" & grepl("[0-9]$", motif), "gene_name", info)) %>%
  mutate(motif=gsub("(^MA.+\\.[0-9]{1})\\.[a-z0-9].+$", "\\1", motif) %>%
           gsub("_.+", "", .)) %>%
  mutate(info=gsub("\\.[0-9]+","", info))


# motif_annots
dim(motif_annots)
```

```{r}
ortholog_motifs_ids <- motif_annots %>% 
  filter(values %in% ortholog_uniprotIDs$entry) %>% 
  pull(motif)


sel_motif_annots <- motif_annots %>% 
  filter(motif %in% ortholog_motifs_ids) %>% 
  pivot_wider(names_from = info,
              values_from = values) %>% 
  left_join(., ortholog_uniprotIDs, by=c("acc"="entry")) %>% 
  left_join(., orthoGroups %>% 
                  # filter(ortholog_group %in% sel_motif_annots$group_id) %>% 
                  rename_at(vars(gene_id:p_fam_description), ~paste0("pfal_", .)),
            by = c("group_id"="ortholog_group")) %>% 
  select(motif:acc, group_id,
         pfal_gene_id,
         pfal_product_description,
         everything()) %>% 
  arrange(group_id)


# sel_motif_annots
dim(sel_motif_annots)
# write.csv(sel_motif_annots, results/motif_analysis/jaspar2022_orthologs.csv", row.names = F)
```



```{r}
sapply(all_PWMs[sel_motif_annots$motif], function(x) maxScore(Matrix(x))) %>% 
  data.frame(score=.) %>% 
  rownames_to_column("motif") %>% 
  arrange(score) %>% 
  mutate(percent_75=score*0.75, 
         percent_90=score*0.9)
```

```{r}
ortho_day2_enr <- motif_enrichment_workflow(targets_df = Day2,
                                      PWMs = all_PFMs[sel_motif_annots$motif],
                                      proms_gr = pf_proms_gr,
                                      genome_bs = genome,
                                      col_to_bin = "logFC",
                                      use_bg_regions = TRUE)

# saveRDS(ortho_day2_enr,results/motif_analysis/monaLisa/day2_vs_spz_orthologs_monaLisa_results.RDS")
```

```{r}
# ortho_enr$enrichement_res
ortho_enr_df <- create_motif_df(ortho_day2_enr,col_to_bin = "logFC") %>% 
  left_join(., sel_motif_annots, by="motif") %>% 
  select(motif, gene_name, class, family, everything()) %>% 
   arrange(desc(negLog10P))

# dim(ortho_enr_df)
ortho_enr_df
# write.csv(ortho_enr_df,results/motif_analysis/monaLisa/day2_vs_spz_orthologs_monaLisa_results.csv", row.names = FALSE)
```

```{r}
tc_clusters <- time_course %>% 
  mutate(gene_cluster = paste0("cluster", gene_cluster) %>% as.factor()) %>% 
  select(gene_id, gene_cluster)

tc_profiles <- time_course %>% 
  mutate(combined_gene_clusters=as.factor(combined_gene_clusters)) %>% 
  select(gene_id, combined_gene_clusters) %>% 
  arrange(combined_gene_clusters)
```

```{r eval=FALSE}
cluster_nums <- unique(tc_clusters$gene_cluster)
ortho_cluster_motifs <- list()
for (cluster in cluster_nums){
  
  input <- tc_clusters[tc_clusters$gene_cluster == cluster, ] %>% 
    droplevels()

  enr <- motif_enrichment_workflow(targets_df = input,
                            col_to_bin = "gene_cluster",
                            PWMs = all_PWMs[sel_motif_annots$motif],
                            proms_gr = pf_proms_gr,
                            genome_bs = genome,
                            make_bins = FALSE,
                            use_bg_regions = TRUE)
  
  ortho_cluster_motifs[[cluster]] <- enr
}

# saveRDS(ortho_cluster_motifs,results/motif_analysis/monaLisa/time_course_ortholog_results_list.RDS")
```

```{r}
ortho_cluster_motifs_dfs <- purrr::map_dfr(ortho_cluster_motifs, create_motif_df) %>% 
  left_join(., sel_motif_annots, by="motif") %>% 
  select(motif,gene_cluster, gene_name,pfal_gene_id, class, family, everything()) %>% 
  arrange(desc(negLog10P))

# write.csv(ortho_cluster_motifs_dfs,results/motif_analysis/monaLisa/time_course_orthologs_results.csv", row.names = FALSE)

ortho_cluster_motifs_dfs %>%
  filter(negLog10P > 1.2) %>%
  group_by(gene_cluster)
  # dplyr::slice(1)
```

```{r eval=FALSE}
profile_nums <- unique(tc_profiles$combined_gene_clusters)
ortho_profile_motifs <- list()
for (profile in profile_nums){
  
  input <- tc_profiles[tc_profiles$combined_gene_clusters == profile, ] %>% 
    droplevels()

  enr <- motif_enrichment_workflow(targets_df = input,
                            col_to_bin = "combined_gene_clusters",
                            PWMs = all_PWMs[sel_motif_annots$motif],
                            proms_gr = pf_proms_gr,
                            genome_bs = genome,
                            make_bins = FALSE,
                            use_bg_regions = TRUE)
  
  ortho_profile_motifs[[profile]] <- enr
}

# saveRDS(ortho_profile_motifs, results/motif_analysis/monaLisa/time_course_profiles_orthologs_results_list.RDS")
```

```{r}
ortho_profile_motifs_dfs <- purrr::map_dfr(ortho_profile_motifs, create_motif_df) %>% 
  left_join(., sel_motif_annots, by="motif") %>% 
  select(motif,gene_cluster, gene_name,pfal_gene_id, class, family, everything()) %>% 
  arrange(gene_cluster, desc(negLog10P))

# write.csv(ortho_profile_motifs_dfs,results/motif_analysis/monaLisa/time_course_profiles_orthologs_results.csv", row.names = FALSE)

ortho_profile_motifs_dfs %>%
  filter(negLog10P > 1.2) %>%
  group_by(gene_cluster)
  # dplyr::slice(1)
```

### Visualization Motif Enrichment / Seq Logos 

#### Time course Clusters

```{r}
# negLog10P > threshold
print(threshold)
```

```{r}
ortho_cluster_motifs_dfs <- read.csv("results/motif_analysis/monaLisa/time_course_orthologs_results.csv")

input <- ortho_cluster_motifs_dfs  %>%
  mutate(gene_cluster_factor=factor(gsub("cluster", "", gene_cluster),
                                    levels=c(1,2,3,8,4,6,5,9,7)),
         label = gene_cluster) %>%
  group_by(gene_cluster) %>%
  mutate(plot_fill=case_when(
    negLog10P > threshold ~ log2enr,
    TRUE ~ NA_real_)) %>%
  ungroup() %>%

  group_by(motif) %>%
  filter(!all(is.na(plot_fill))) %>%
  ungroup() %>%

  mutate(motif_factor_groups=paste0(label,":",motif)) %>%
  group_by(gene_cluster) %>%
  arrange(desc(plot_fill), .by_group = T) %>%
  ungroup() %>%
  mutate(motif_factor_groups=factor(motif_factor_groups,
                                    levels=rev(motif_factor_groups))) %>%
  select(gene_cluster,
         gene_cluster_factor,
         motif_factor_groups,
         plot_fill,log2enr, everything())

dim(input)
# head(input)
length(unique(input$motif)) #5
```


```{r}
chk <- sel_tc_hits %>% 
  bind_rows(input) %>% 
  filter(negLog10P > threshold) 

dim(chk)
quantile(chk$p_value)
quantile(chk$adj.p_value)
# format(quantile(chk$p_value), scientific = TRUE, digits = 2)
```


```{r}
sel_motif_annots <- read.csv("results/motif_analysis/jaspar2022_orthologs.csv")

ortho_motif_PWMs <- all_PFMs[sel_motif_annots$motif] %>% 
  toPWM(., type="prob")

vals <- input %>% 
  mutate(plot_fill=ifelse(is.na(plot_fill), 0, plot_fill)) %>% 
  pull(plot_fill, name=motif)
  
clust_motifs <- hclust(dist(vals, method="euclidean"))
# plot(clust_motifs)

order_motifs_tc <-
  clust_motifs$labels[clust_motifs$order][1] %>%
  c(.,  tail(clust_motifs$labels[clust_motifs$order], n=7)) %>%
  unique()

input_withOrders <- input %>%
  mutate(motif=factor(motif, 
                      levels=rev(order_motifs_tc)),
         gene_cluster_factor=factor(gene_cluster_factor, 
                                    levels=rev(levels(gene_cluster_factor))))

```

```{r}
pf_ortho_geneids <- input %>% 
  group_by(gene_cluster) %>% 
  arrange(desc(negLog10P), .by_group = TRUE) %>% 
  dplyr::slice(1) %>% 
  ungroup %>% 
  filter(!is.na(plot_fill))


expn_top_hits <- expression[unique(pf_ortho_geneids$pfal_gene_id), rownames(input_d456)] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to = "log2_CPM") %>% 
  left_join(., design_input, by="masigpro_id") %>% 
  left_join(., select(pf_ortho_geneids, gene_cluster:pfal_gene_id),
            by=c("gene_id"="pfal_gene_id"), 
            multiple = "all")



# head(expn_top_hits)
# dim(expn_top_hits)
# head(expn_top_hits)
# unique(expn_top_hits$motif)
```

```{r}
expn_labs_d456 <- expn_d456 %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) %>% 
  select(analysis_factor_group, everything())

summary_stats_TF1 <- expn_top_hits %>% 
  group_by(gene_cluster_factor, 
           analysis_factor_group,
           motif) %>% 
  mutate(mean=mean(log2_CPM),
           sd=sd(log2_CPM),
           cv=sd/mean, 
           ymin=mean-sd,
           ymax=mean+sd,
           stat="transcription factor"
          # .groups = 'keep'
         ) %>% 
  ungroup() %>% 
  select(gene_cluster_factor,
         analysis_factor_group,
         masigpro_id,
         motif, stat, mean, 
         ymin:ymax) %>% 
  distinct() %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) 

tc_ortho_summary_stats <- expn_d456 %>% 
  mutate(gene_cluster_factor=factor(gene_cluster, levels=c(1,2,3,8,4,6,5,9,7))) %>%
  group_by(gene_cluster_factor, analysis_factor_group) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd) %>% 
  group_by(gene_cluster_factor,analysis_factor_group, masigpro_id) %>% 
  mutate(stat="time-course cluster",
          mean=mean(log2_CPM),
          motif="time-course cluster",
         .groups = 'keep') %>% 
  ungroup() %>%
  select(gene_cluster_factor,
         analysis_factor_group,
         masigpro_id,
         motif, stat, mean, 
         ymin:ymax, label) %>%  
  distinct() %>% 
  bind_rows(summary_stats_TF1) %>% 

  group_by(gene_cluster_factor) %>%
  mutate(label=as.character(label)) %>%
  mutate_at(vars(label), ~ifelse(is.na(.), label[!is.na(label)], .)) %>%
  ungroup() %>%

  group_by(gene_cluster_factor,analysis_factor_group,stat) %>%
  mutate(middle_val=median(mean)) %>%
  ungroup() %>%

  arrange(gene_cluster_factor,analysis_factor_group) %>%
  mutate(label=factor(label, levels=unique(label))) %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) 


motif_colors <- c(ggthemes::canva_pal("Simple but bold")(4), ggthemes::canva_pal("Warm and bold")(4))
motif_colors <- motif_colors[c(2:4,6:8)] %>% 
  set_names(c("time-course cluster",unique(input$motif)))

# barplot(rep(1,length(motif_colors)), col=motif_colors)

tc_ortho_plots <- create_motif_enr_plots(enr_df = input, 
                       factor_column="gene_cluster_factor",
                       motif_order =  order_motifs_tc,
                       summary_stats = tc_ortho_summary_stats, 
                       expn_df_long = expn_labs_d456, 
                       PWMlist = ortho_motif_PWMs,
                       motif_tile_fill = 'blue',
                       motif_colors = motif_colors)
```

```{r fig.height=12, fig.width=10}
layout_rows <- "
ABC
ABD
ABE
ABF
ABG
ABH
ABI
ABJ
ABK
"

wdts <- c(0.6,0.6,0.75)


# svglite("figures/motif_analysis/time_point_cluster_ortholog_heatmap_seqlogos_tileplot_rows.svg", height = 15, width=15, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
tc_ortho_plots$time_cluster + tc_ortho_plots$motif_enr + 
  tc_ortho_plots$top_logos$`cluster1:MA0502.2` +
  plot_spacer() + 
  tc_ortho_plots$top_logos$`cluster3:MA0314.2` + 
  plot_spacer() +
  tc_ortho_plots$top_logos$`cluster4:MA0316.1` + 
  tc_ortho_plots$top_logos$`cluster6:MA1644.1` + 
  plot_spacer() +  
  tc_ortho_plots$top_logos$`cluster9:MA1644.1` + 
  plot_spacer() +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```

```{r fig.height=10, fig.width=10}
# svglite("figures/motif_analysis/time_point_cluster_ortholog_TF_expn_heatmap_seqlogos_tileplot_rows.svg", height = 18, width=17, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
tc_ortho_plots$time_ribbon + tc_ortho_plots$motif_enr + 
  tc_ortho_plots$top_logos$`cluster1:MA0502.2` +
  plot_spacer() + 
  tc_ortho_plots$top_logos$`cluster3:MA0314.2` + 
  plot_spacer() +
  tc_ortho_plots$top_logos$`cluster4:MA0316.1` + 
  tc_ortho_plots$top_logos$`cluster6:MA1644.1` + 
  plot_spacer() +  
  tc_ortho_plots$top_logos$`cluster9:MA1644.1` + 
  plot_spacer() +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```

#### Time Course Profiles 

```{r}
# negLog10P > 1.2
print(threshold)
```

```{r}
ortho_profile_motifs_dfs <- read.csv("results/motif_analysis/monaLisa/time_course_profiles_orthologs_results.csv")

in_profiles <- ortho_profile_motifs_dfs  %>%
  mutate(gene_cluster_factor=as.factor(gene_cluster),
         label = gene_cluster) %>%
  group_by(gene_cluster) %>%
  mutate(plot_fill=case_when(
    negLog10P > thresholds ~ log2enr,
    TRUE ~ NA_real_)) %>%
  ungroup() %>%

  group_by(motif) %>%
  filter(!all(is.na(plot_fill))) %>%
  ungroup() %>%

  mutate(motif_factor_groups=paste0(label,":",motif)) %>%
  group_by(gene_cluster) %>%
  arrange(desc(plot_fill), .by_group = T) %>%
  ungroup() %>%
  mutate(motif_factor_groups=factor(motif_factor_groups,
                                    levels=rev(motif_factor_groups))) %>%
  select(gene_cluster,
         gene_cluster_factor,
         motif_factor_groups,
         plot_fill,log2enr, everything())


length(unique(in_profiles$motif))
unique(in_profiles$motif)
```


```{r}
ortho_motif_PWMs <- all_PFMs[sel_motif_annots$motif] %>% 
  toPWM(., type="prob")

vals <- in_profiles %>% 
  mutate(plot_fill=ifelse(is.na(plot_fill), 0, plot_fill)) %>% 
  pull(plot_fill, name=motif)
  
clust_motifs <- hclust(dist(vals, method="euclidean"))
# plot(clust_motifs)

order_motifs_profiles <-
  tail(clust_motifs$labels[clust_motifs$order], n=3) 


# input_withOrders <- in_profiles %>%
#   mutate(motif=factor(motif, 
#                       levels=rev(order_motifs_profiles)),
#          gene_cluster_factor=factor(gene_cluster_factor, 
#                                     levels=rev(levels(gene_cluster_factor))))
```


```{r}
profile_ortho_geneids <- in_profiles %>% 
  group_by(gene_cluster) %>% 
  arrange(desc(negLog10P), .by_group = TRUE) %>% 
  dplyr::slice(1) %>% 
  ungroup %>% 
  filter(!is.na(plot_fill))


p_expn_top_hits <- expression[unique(profile_ortho_geneids$pfal_gene_id), rownames(input_d456)] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to = "log2_CPM") %>% 
  left_join(., design_input, by="masigpro_id") %>% 
  left_join(., select(profile_ortho_geneids, gene_cluster:pfal_gene_id),
            by=c("gene_id"="pfal_gene_id"), 
            multiple = "all")


unique(p_expn_top_hits$motif)
```

```{r}
summary_stats_TF2 <- p_expn_top_hits %>% 
  group_by(gene_cluster_factor, 
           analysis_factor_group,
           motif) %>% 
  mutate(mean=mean(log2_CPM),
           sd=sd(log2_CPM),
           cv=sd/mean, 
           ymin=mean-sd,
           ymax=mean+sd,
           stat="transcription factor"
          # .groups = 'keep'
         ) %>% 
  ungroup() %>% 
  select(gene_cluster_factor,
         analysis_factor_group,
         masigpro_id,
         motif, stat, mean, 
         ymin:ymax) %>% 
  distinct() %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) 

profile_labs <- time_course %>% 
  group_by(combined_gene_clusters) %>% 
  reframe(label=paste0(combined_gene_clusters,"\n(",length(unique(gene_id))," genes)")) %>% 
  ungroup() %>% 
  distinct()

profile_summary_stats <- expn_d456 %>% 
  select(-label) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  mutate(gene_cluster_factor=as.factor(combined_gene_clusters)) %>% 
  left_join(., profile_labs, by=c("combined_gene_clusters")) %>% 
  
  group_by(gene_cluster_factor,
           analysis_factor_group) %>% 
  mutate(mean=mean(log2_CPM),
         sd=sd(log2_CPM),
         cv=sd/mean, 
         ymin=mean-sd,
         ymax=mean+sd) %>% 
  
  group_by(gene_cluster_factor,
           analysis_factor_group,
           masigpro_id) %>% 
  mutate(stat="time-course profile",
          mean=mean(log2_CPM),
          motif="time-course profile",
         .groups = 'keep') %>% 
  ungroup() %>%
  select(gene_cluster_factor,
         analysis_factor_group, 
         masigpro_id,
         motif, stat, mean, ymin:ymax, label) %>% 
  distinct() %>% 
  bind_rows(summary_stats_TF2) %>% 
  
  group_by(gene_cluster_factor) %>% 
  mutate(label=as.character(label)) %>% 
  mutate_at(vars(label), ~ifelse(is.na(.), label[!is.na(label)], .)) %>% 
  ungroup() %>% 
  
  group_by(gene_cluster_factor,analysis_factor_group,stat) %>% 
  mutate(middle_val=median(mean)) %>% 
  ungroup() %>% 
  
  arrange(gene_cluster_factor,analysis_factor_group) %>% 
  mutate(label=factor(label, levels=unique(label))) %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS"))) 

profile_median_expn <- expn_d456 %>% 
  select(-label) %>% 
  mutate(combined_gene_clusters=case_when(
    grepl("1$|2$",gene_cluster) ~ "Profile1",
    grepl("3$|8$", gene_cluster) ~ "Profile2",
    grepl("4$|6$", gene_cluster) ~ "Profile3",
    grepl("5$|9$", gene_cluster) ~ "Profile4",
    grepl("7", gene_cluster) ~ "Profile5"
  )) %>% 
  mutate(gene_cluster_factor=as.factor(combined_gene_clusters)) %>% 
  left_join(., profile_labs, by=c("combined_gene_clusters")) %>% 
  
  group_by(gene_cluster_factor, masigpro_id) %>% 
  mutate(median_val_per_sample=median(log2_CPM),
         mean_val_per_sample=median(log2_CPM)) %>% 
  ungroup() %>% 
  
  group_by(gene_cluster_factor, analysis_factor_group) %>% 
  mutate(median_val=median(log2_CPM),
         mean_val=mean(log2_CPM)) %>% 
  ungroup() %>% 
  mutate(masigpro_id = gsub("^.+_(Day[0-9])(_R.+)","\\1_Pf_LS\\2",   gsub("Pf_Day0","sporozoites",masigpro_id))) %>% 
  mutate(analysis_factor_group = factor(gsub("_Rep[0-9]", "", masigpro_id), 
                                          levels= c("sporozoites","Day2_Pf_LS",
                                                    "Day4_Pf_LS","Day5_Pf_LS","Day6_Pf_LS")))

```

```{r fig.height=10, fig.width=10}
# ggthemes::canva_palettes %>% names() %>% grep("bold", ., value = T)
possible_colors <- c(ggthemes::canva_pal("Simple but bold")(4), ggthemes::canva_pal("Warm and bold")(4))
# barplot(rep(1,length(possible_colors)), col=possible_colors)

p_motif_colors <- possible_colors[c(2:4,6)] %>% 
  set_names(c("time-course profile",unique(in_profiles$motif)))
# barplot(rep(1,length(p_motif_colors)), col=p_motif_colors)


profile_ortho_plots <- create_motif_enr_plots(enr_df = in_profiles,
                       factor_column="gene_cluster_factor",
                       motif_order =  order_motifs_profiles,
                       summary_stats = profile_summary_stats,
                       expn_df_long = profile_median_expn,
                       PWMlist = ortho_motif_PWMs,
                       motif_tile_fill = 'blue',
                       motif_colors = p_motif_colors)
```


```{r fig.height=10, fig.width=10}
layout_rows <- "
ABC
ABD
ABE
ABF
ABG
"

wdts <- c(0.6,0.6,0.75)


# svg("figures/motif_analysis/time_point_profiles_ortholog_heatmap_seqlogos_tileplot_rows.svg", height = 8, width=10)
# svglite("figures/motif_analysis/time_point_profiles_ortholog_heatmap_seqlogos_tileplot_rows.svg", height = 15, width=17, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
profile_ortho_plots$time_cluster + profile_ortho_plots$motif_enr + 
  profile_ortho_plots$top_logos$`Profile1:MA1644.1` +
  plot_spacer() + 
  profile_ortho_plots$top_logos$`Profile3:MA1644.1` +
  profile_ortho_plots$top_logos$`Profile4:MA0540.1` +
  plot_spacer() +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```

```{r fig.height=5, fig.width=10}
# svg("figures/motif_analysis/time_point_profile_ortholog_TF_expn_heatmap_seqlogos_tileplot_rows.svg", height = 10, width=12)

# svglite("figures/motif_analysis/time_point_profile_ortholog_TF_expn_heatmap_seqlogos_tileplot_rows.svg", height = 15, width=17, system_fonts = list(sans = "Open Sans"),fix_text_size=FALSE)
profile_ortho_plots$time_ribbon +
  profile_ortho_plots$motif_enr +
  profile_ortho_plots$top_logos$`Profile1:MA1644.1` +
  plot_spacer() +
  profile_ortho_plots$top_logos$`Profile3:MA1644.1` +
  profile_ortho_plots$top_logos$`Profile4:MA0540.1` +
  plot_spacer() +
  plot_layout(design = layout_rows, guides = 'collect', heights = NULL, widths = wdts)
# dev.off()
```



### Ortholog Expression 

```{r}
ortholog_pf_ids <- orthoGroups %>% 
  inner_join(., select(sel_motif_annots, motif,motif_gene_name=gene_name, motif_acc=acc, 
                       group_id,motif_class=class),
             by=c("ortholog_group"="group_id"),
             multiple="all") %>% 
  group_by(gene_id) %>% 
  mutate(across(.cols = c(motif:motif_acc,motif_class),
                .fns =  ~paste(unique(.), collapse = " / "))) %>% 
  ungroup() %>% 
  distinct() %>% 
  mutate(label=paste(gsub(" ","", motif), gsub(" ","", motif_gene_name), gene_id,sep=": ")) %>% 
  select(gene_id,ortholog_group,label,
         gene_name_or_symbol,p_fam_description,motif:motif_class,
         everything())

# dim(ortholog_pf_ids) # 9 18
head(ortholog_pf_ids)
```

```{r}
orth_tfs_expn <- expression[ortholog_pf_ids$gene_id, rownames(input_d456)] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols=matches("Pf"),
               names_to="masigpro_id",
               values_to = "log2_CPM") %>% 
  left_join(., design_input, by="masigpro_id") %>% 
  left_join(.,ortholog_pf_ids, by="gene_id") %>% 
  select(gene_id,label, everything())


head(orth_tfs_expn)
dim(orth_tfs_expn)  # 126  27
# length(unique(orth_tfs_expn$gene_id)) # 9 OK
```


```{r fig.height=10, fig.width=10}
# pdf("figures/combat_seq/ortholog_motif_gene_expression_boxplot.pdf", height = 10, width=10)
ggplot(orth_tfs_expn, aes(x=analysis_factor_group, y=log2_CPM)) +
  geom_boxplot(aes(fill=analysis_factor_group),
               alpha=0.3,
               outlier.colour = NA) +
  geom_point(aes(color=analysis_factor_group),
             size=2, alpha=0.7,
             position = position_jitter()) +
  facet_wrap(~label, scales = "free_x", ncol=3) +
  labs(x="") +
  scale_fill_viridis_d(direction = -1) +
  scale_color_viridis_d(end=0.925,direction = -1) +
  theme_classic() +
  theme(legend.position = "top", 
        axis.text.x = element_text(angle=25, vjust=1, hjust=1))
# dev.off()
```

```{r}
id_map <- select(orth_tfs_expn, gene_id, label) %>% 
  distinct()

heat_colors <- colorRampPalette(viridis::plasma(10))(299)
annot_col_df <- design_input[rownames(input_d456), ] %>% 
  droplevels()
n <- length(levels(annot_col_df$analysis_factor_group))
annot_colors <- list(analysis_factor_group = ggpubr::get_palette("jco", n) %>% 
                       set_names(levels(annot_col_df$analysis_factor_group)))

# annot_colors

ortho_tfs_mat <- expression[id_map$gene_id, rownames(input_d456)]
rownames(ortho_tfs_mat) <- id_map$label

# pdf("figures/combat_seq/ortholog_motif_heatmap_log2CPM_heatmap.pdf", height = 10, width = 15)
pheatmap::pheatmap(ortho_tfs_mat,
                   annotation_col = select(annot_col_df, analysis_factor_group),
                   annotation_colors = annot_colors,
                   color = heat_colors)
# dev.off()
```

```{r}
paletteLength <- 299
range_expn <- range(scale(t(ortho_tfs_mat)))
myBreaks <- c(seq(range_expn[1], 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(range_expn[2]/paletteLength, range_expn[2], length.out=floor(paletteLength/2)))

# pdf("figures/combat_seq/ortholog_motif_heatmap_row_zscores_heatmap.pdf", height = 10, width = 15)
pheatmap::pheatmap(ortho_tfs_mat,
                   annotation_col = select(design_input, analysis_factor_group),
                   scale = "row",breaks = myBreaks,
                   annotation_colors = annot_colors,
                   color = colorRampPalette(viridis::turbo(10))(299))
# dev.off()
```

### Search Homer Known Results

```{r}
homer_paths <- dir"results/motif_analysis/homer", pattern = "knownResults.txt", full.names = TRUE, recursive = T)

ids <- str_split_fixed(homer_paths, pattern = "/", n=5)[,3:4] %>% 
  as.data.frame() %>% 
  mutate(id=ifelse(grepl("results", V2), V1, paste(V1,V2, sep="_")))

names(homer_paths) <- ids$id
homer_dfs <- purrr::map(names(homer_paths),function(x){

  read.delim(homer_paths[[x]], sep="\t") %>% 
    mutate(result_file=homer_paths[[x]]) %>% 
    mutate(analysis=x) %>% 
    select(analysis, everything())
})

names(homer_dfs) <- ids$id
```

```{r}
homer_fuzzy_search <- function(homer_known_results, pf_groups){
    res <- homer_known_results %>% 
      janitor::clean_names() %>% 
      separate(motif_name, into=c("transcription_factor","source","database"),
               sep="\\/", remove = F,extra = "merge",fill = "right") %>% 
      separate(transcription_factor, into=c("transcription_factor","tf_type"),
               sep="\\(", remove = F, extra = "merge",fill = "right") %>% 
      filter(p_value < 0.05) %>% 
      mutate_at(vars(tf_type), ~gsub("\\)","", .)) %>% 
      mutate_at(vars(transcription_factor), ~toupper(.)) %>% 
      mutate_at(vars(transcription_factor), ~gsub("-|:","",.)) %>% 
      select(transcription_factor, p_value,q_value_benjamini, everything()) 

  regex <- paste(res$transcription_factor, collapse = "|")
  subset <- pf_groups %>%
    group_by(ortholog_group) %>% 
    filter(any(grepl(regex, symbol, ignore.case = T))) %>% 
    ungroup()
  
  if(!nrow(subset) == 0){
    regex2 <- paste(subset$symbol, collapse = "|")
    filt_res <- res %>% 
      filter(grepl(regex2, transcription_factor, ignore.case = T))
  }else{
    filt_res <- data.frame()
  }

  hits <- list("all_hits" = res,
               "regex" = regex,
               "pf_groups" = subset, 
               "filt_hits" = filt_res)
}
```

```{r}
possible_hits <- lapply(homer_dfs,homer_fuzzy_search, pf_groups=pf_groups_clean)
```

```{r}
# lapply(possible_hits, function(x) x$regex)
```







## JASPAR directly 

```{r}
dumpJaspar(filename = "homer/jaspar2020.motif",
  pkg = "JASPAR2020",
  opts = list(tax_group = c("vertebrates", "insects", "urochordat", "nematodes", "fungi"),
              collection = c("CORE")),
  pseudocount = 1,
  relScoreCutoff = 0.8,
  verbose = FALSE
)
```

```{bash eval=FALSE}
#https://pathcards.genecards.org/card/transcriptional_regulation_by_the_ap-2_(tfap2)_family_of_transcription_factors
AP2_Genes=(TFAP2A TFAP2C TFAP2B TFAP2E WWOX TFAP2D CITED2 KCTD1 UBE2I SUMO1 KCTD15 CDKN1A KDM5B CGA CGB3 CREBBP CITED4 EGFR EP300 ERBB2 ESR1 ATAD2 HSPD1 KIT CITED1 MYBL2 MYC NOP2 NPM1 PITX2 TGFA VEGFA YY1 YEATS4 CGB5 CGB8 APOE DEK)

for gene in ${AP2_Genes[*]}; 
do  
  grep -E -A 20 "MA.+:::$gene" jaspar2020.motif;
done > AP2_family_jaspar2020.motif


./homer/create_db.sh
```

Test `-find` with a known hit using the Homer motifs
http://homer.ucsd.edu/homer/custom.motifs

```{bash}
PROJDIR="/active/taylor_s/people/jsmi26/RSC/kappe_s_2022.04_PF_LiverStage/homer"
FILES=$(find $PROJDIR -type f -name "*.csv")
for file in $(echo "$FILES")
do
  ./homer_motif_analysis.pbs $file 
done 
```




# Share the Data

```{bash}
destination="/active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/"
./rsync.sh
```


# Session Info

```{r}
sessionInfo()
```

