---
title: "CVG and Translocome Gene Expression"
author: "Jenny L Smith"
date: "`r Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(tidyr)
library(tibble)
library(magrittr)
library(dplyr)
library(stringr)
library(ComplexHeatmap)
library(ggplot2)
library(gridExtra)
library(ggpubr)
library(patchwork)
library(limma)
library(edgeR)


library(SummarizedExperiment)
library(patchwork)
library(maSigPro)
library(svglite)
suppressPackageStartupMessages(library(mclust))

library(Biostrings)
```


Note about fonts:

```{r}
#https://www.tidyverse.org/blog/2021/02/svglite-2-0-0/
# system_fonts = list(sans = "Open Sans")
# systemfonts::system_fonts()
```

# Define Functions 

```{r}
#from https://github.com/mikelove/DESeq2/blob/master/R/plots.R
#Want to return the whole scores matrix so can examine 3d pca plots. 
plotPCA.DESeq.mod <- function(object, intgroup="condition", ntop=500, returnData=FALSE, PC3=FALSE)
{
  library(matrixStats)
  # calculate the variance for each gene
  rv <- rowVars(assay(object))
  
  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))
  
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )
  
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }
  
  # assembly the data for the plot - first 10 PCs
  # d <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], group=group, intgroup.df, name=colnames(object))
  n <- min(10, ncol(as.data.frame(pca$x)))
  d <- data.frame(as.data.frame(pca$x)[,1:n], group=group, intgroup.df, name=colnames(object))
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:10]
    rot <- pca$rotation[,1:10] #for first 10 PCs
    dat <- list("scores"=d,"rotation"=rot)
    return(dat)
  }
}

#Updated on 6/9/17 to use variance stabilized transformed data as input (not center scaled log2, like in princomp)
PCA <- function(expnData,phenovector,
                title="",round=TRUE,colorCodes=NULL,
                ntop=500, GOI=NULL){
  
  suppressPackageStartupMessages(library(DESeq2))
  library(ggplot2)
  #expnData is the raw counts (not normalized) has patient IDs as colnames and genes as rownames. 
  
  samples <- intersect(names(phenovector), colnames(expnData))
  countData <- expnData[,samples]
  phenovector <- phenovector[samples]
  colData <- as.data.frame(phenovector)
  
  if(round){
    countData <- round(countData, digits = 0)
  }
  
  #Create as DESeq data set object (dds)
  dds <- DESeqDataSetFromMatrix(countData = countData,
                                  colData = colData,
                                  design = ~ 1)
  
  #perform variance stabilized transformation 
  dds <- dds[ rowSums(counts(dds)) > 10, ]
  varianceStab <- vst(dds, blind = TRUE)
  
  #if given a list of genes of interest
  if (! is.null(GOI)){
    GOI <- intersect(GOI, rownames(assay(varianceStab)))
  }else{
    GOI <- 1:nrow(varianceStab)
  }
  
  #PCA data frame with the wieghts/loadings and eigen vectors
  pca.dat <- plotPCA.DESeq.mod(varianceStab[GOI,], 
                               intgroup = "phenovector", 
                               ntop = ntop,
                               returnData=TRUE)
  
  plots <- lapply(c(2:3), function(pc){
      percentVar <- attr(pca.dat$scores, which="percentVar")
      y_var <- paste0("PC",pc)
      
      p <- ggplot(data=pca.dat$scores, 
             aes_string(x="PC1", y=y_var, color="phenovector")) + 
        geom_point(size=3, alpha=0.75) +
        xlab(paste0("PC1: ",round(percentVar[1] * 100),"% variance")) +
        ylab(paste0(y_var,": ",round(percentVar[pc] * 100),"% variance")) +
        labs(title=title) +
        theme_classic() +
        theme(legend.position = "top")
      if(!is.null(colorCodes)){
        p <- p + 
          scale_color_manual(values=colorCodes)
      }
      return(p)
  })
  
  #Final Results object
  res <- list(dds, varianceStab,pca.dat, plots)
  names(res) <- c("dds", "vst","pca_data","pca_plots")
  
  if(is.character(GOI)){
    res[["GOI"]] <- GOI
  }
  
  return(res)
}
```

```{r}
#changed on 2/14/18, see bottom of Heatmaps_Function.r for the original one used.
dge_dendrograms <- function(expnData, pheno, method,
                            genelist=NULL,add.count=1, percent=0.05,
                            ntop=500,filterTopGenes=FALSE, createDGE=TRUE,log=FALSE){
  #df with count data, patient IDs are column names and rows are genes.
  #pheno is a character vector with patient IDs as names, and the status for each in each group (eg pos,neg)
  #genelist is a character vector with the genes of interest
  #percent is the % of samples in the input expn matrix that must express a gene at 1 CPM. Filter to remove low count genes.
  #set log=TRUE if providing a log2 expression dataframe.
  #filterTopGenes shuold be a logical. If TRUE filter top 1000 most varied genes.
  suppressPackageStartupMessages(require(edgeR))
  suppressPackageStartupMessages(library(dendextend))
  suppressPackageStartupMessages(library(matrixStats))

  expnData <- expnData[, intersect(names(pheno), colnames(expnData))] #ensure correct order, drop rows with nas just in case
  if(createDGE){
    dge <- DGEList(counts = expnData)
    #keep the rows (genes) with at least 1 CPM in a minimum of 2 samples or X% of samples. 
    keep.dge <- rowSums(cpm(dge) >= 1) >= max(2,(percent*ncol(expnData))) 
    # subset for those genes with cmp >= 1 per gene in samples
    dge <- dge[keep.dge,] #
    dge <- calcNormFactors(dge)
    TMMCPM <- cpm(dge, normalized.lib.sizes = TRUE,
                  log = TRUE, prior.count = add.count)

  }else{
    TMMCPM <- as.matrix(expnData)
    if(!log){
      TMMCPM <- log2(expnData+add.count) #log2 transform counts
    }
  }

  if(is.null(genelist) & filterTopGenes){
      # calculate the variance for each gene
      rv <- rowVars(as.matrix(TMMCPM))
      # select the ntop genes by variance
      select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
      #select the top 1000 most varied genes
      TMMCPM <- TMMCPM[select,]
  }else if(!is.null(genelist)){
    TMMCPM <- TMMCPM[which(rownames(TMMCPM) %in% genelist), ] #subset the matrix to genes of interest
  }


  d1 <- dist(t(TMMCPM), method = "euclidean", diag = FALSE,
             upper = FALSE) #sample distances WITHOUT SCALING
  d2 <- dist(TMMCPM, method = "euclidean", diag = FALSE,
             upper = TRUE) #gene distances WITHOUT SCaling
  samp.c1 <- hclust(d1, method = method, members = NULL) #sample clustering
  gene.c2 <- hclust(d2, method = method, members = NULL) #gene clustering
  list <- list(TMMCPM,samp.c1,gene.c2)
  names(list) <- c("TMMCPM","samp.c1", "gene.c2")

  return(list)
}

```


```{r}
#Pheatmap
quickPheatmap <- function(expn,geneList, sample_data,
                          gene_index=NULL,
                          annots_col_colors=NULL){
  library(pheatmap)

  len <- 299
  col <- colorRampPalette(c("deepskyblue4", "deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=len)

  e <- expn[geneList, intersect(colnames(expn), rownames(sample_data))]
  sample_data <- sample_data[intersect(colnames(expn), rownames(sample_data)), ]
  anno.col <- sample_data
  #if scaling, this doesnt work. would need to scale the expression matrix myself
  # Breaks <- c(seq(min(e), 0, length.out=ceiling(len/2) + 1),
  #             seq(max(e)/len, max(e), length.out=floor(len/2)))
 
  if(is.null(annots_col_colors)){
    annots_col_colors <- lapply(anno.col, function(column){
        grps <- unique(column)
        n <- length(grps)
        pal <- RColorBrewer::brewer.pal(n,name="Set1")
        names(pal) <- grps
        return(pal)
      })
  }


  p <- pheatmap::pheatmap(mat= e ,
                     color=col,
                     border_color="black",
                     scale = "row", 
                     cluster_rows=FALSE,
                     cluster_cols=TRUE,
                     clustering_method="ward.D2",
                     clustering_distance_cols="correlation",
                     annotation_names_col=TRUE,
                     annotation_col = anno.col,
                     annotation_colors = annots_col_colors,
                     gaps_row=gene_index,
                     show_rownames=TRUE,
                     fontsize_row=5,
                     fontsize_col = 10,
                     # kmeans_k=9,
                     show_colnames=TRUE)


  return(p)

}
```

```{r}
create_annots_hmap <- function(expn, pheno_df, cols,
                               goi=NULL,cc=NULL, 
                               colorbar.height=5){
  #expn is the normalized expression values with genes as rownames
  #gene list a character vector
  #goi are genes of interest to highlight on the Heatmap. Character vector of gene symbols
  #cc are color codes in WHICH FORMAT?
  #cols is a character vector of column names

  #Example on how to use the reuslts:
  #hmap <- ComplexHmap(XXXX, hmap_anno_obj=res$annoColumn, XXX)
  # draw(hmap + res$geneLabels, heatmap_legend_side="right", annotation_legend_side="right")

    library(dplyr)
    suppressPackageStartupMessages(library(ComplexHeatmap))

    #subset the expression matix and the phenotype dataframe
    anno <- pheno_df %>%
      dplyr::filter(rownames(.) %in% colnames(expn)) %>% 
      rownames_to_column("id") %>% 
      dplyr::select(id, all_of(cols)) %>%
      mutate(id = factor(id, levels = colnames(expn))) %>%
      arrange(id) %>%  #ensure same order as the expn matrix
      column_to_rownames("id")
  
    #if no color codes provided, create one for each column
    if(is.null(cc)){
      pals <- c("npg", "aaas", "lancet", "jco", 
                    "ucscgb", "uchicago", "simpsons",
                    "rickandmorty")
      #if there are more columns than color palettes, just repeat the set of available color palettes
      if(length(ncol(anno)) > length(pals)){
        pals <- rep(pals,  ceiling(ncol(anno)/length(pals)))
      }
      #define unique colors for each column
      cc <- purrr::map(1:ncol(anno), function(i){
          grps <- unique(anno[[i]])
          ggpubr::get_palette(pals[i], k=length(grps)) %>% 
            magrittr::set_names(grps)
      })
      names(cc) <- colnames(anno)
    }
    
    #legend graphical parameters
    n <- max(sapply(cc, length))
    nrow <- ifelse(n <= 6, n, 6)
    ncol <- ceiling(n/nrow)
    params <- list(show_legend=TRUE,
                  labels_gp= gpar(fontsize=12),
                  title_gp= gpar(fontsize=16),
                  nrow = nrow,
                  ncol=ncol,
                  by_row=TRUE)
    
    #Create the complex heatmap annotation column
    annoCol <- suppressWarnings(HeatmapAnnotation(df = dplyr::select(anno, all_of(cols)),
                                   name="Main Groups",
                                   col=cc,
                                   which="column",
                                   gap=unit(1,"mm"),
                                   border = T,
                                   show_annotation_name = TRUE,
                                   annotation_name_gp=gpar(fontsize=12),
                                   annotation_name_offset=unit(1,"mm"),
                                   annotation_name_side="left",
                                   annotation_height=unit(colorbar.height, "cm"),
                                   annotation_legend_param = params,
                                   simple_anno_size_adjust=TRUE))
    res <- list("annoColumn"=annoCol)
    if(!is.null(goi)){
      regex <- paste0("^",goi,"$", collapse = "|")
      goi.idx <- grep(regex, rownames(expn))
      labels <- rownames(expn)[goi.idx] #needs to be numeric indexes for complexheatmap
      #create the row (gene) labels object
      labs <- circlize::rowAnnotation(link = anno_mark(at=goi.idx,
                                             labels=labels,
                                             which="row",
                                             link_width=unit(1, "mm")),
                            width= unit(1, "mm") + max_text_width(labels),
                            gp=gpar(fontsize=4))

      res[["geneLabels"]] <-  labs
    }
    return(res)
}


complex_hmap <- function(mat,
                        hmap_anno_obj,
                        hmap_anno_obj_genes=NULL,
                        name="z-scores",
                        scale=TRUE,
                        space.type="sRGB",
                        color_palette=NULL,
                        split=NULL, 
                        cluster.method="ward.D2",
                        show_sample_ids=FALSE,
                        dge_dendrograms.res=NULL,
                        samp_dend_order=NULL){
  #mat is the normalized, log2 (usually) transformed counts
  #name is the title
  #scale is whether to scale by row
  #color palette is a colorRamp2() object.
  #threshold is whether to make all z-scores in a certain range.
  #hmap_anno_obj is from HeatmapAnnotation() function
  #space.type is for the color/shades on the heatmap. See ?Heatmap for all the choices.
  #dge_dendrograms.res is the list object output from the dge_dendrograms() function.
  #samp_dend_order is the numeric vector or character vector of column names from the matrix (mat) or the dge_dengrograms.res$TMMCPM matix, in the desired order.

  suppressPackageStartupMessages(library(ComplexHeatmap))
  suppressPackageStartupMessages(require(circlize))
  library(RColorBrewer)
  suppressPackageStartupMessages(library(dendextend))
  ht_opt$message = FALSE

   if(scale){
      mat <- t(scale(t(mat))) ##rowwise scaling
   }
  
  print(range(mat))
  min <- round(min(mat))
  max <- round(max(mat))
  mid <- round(max/2, digits = 2)
    
  if(is.null(color_palette)){
    pal <- colorRamp2(c(min,-mid, 0, mid, max),
                      c("deepskyblue3", "deepskyblue","white", "red", "red3"),
                      space=space.type)
  }else{
    pal <- color_palette
  }
  # col <- colorRampPalette(c("cyan1", "cyan2", "cyan3", "cyan4","azure4","magenta4", "magenta3", "magenta2", "magenta1"))(n=299)
  #colorRamp2(c(-2, 0, 4), c("deepskyblue","white", "red"), space="RGB") #use for breaks.
  # colorPal <- colorRampPalette(c("deepskyblue4", "deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=299)

  #legend graphical parameters
  params <-  list(color_bar="continuous",
       legend_direction="horizontal",
       title_position="topcenter",
       legend_width=unit(5,"cm"),
       legend_height=unit(5,"cm"),
       title_gp=gpar(fontsize=10,
                     fontface="bold"),
      at = c(min,-mid, 0,mid, max),
      just = c("left", "top"))
  
    if(!is.null(dge_dendrograms.res)){
        if(is.null(samp_dend_order)){
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                           order=c(ncol(mat):1))
        }else{
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                                  order=samp_dend_order)
      }
    }else{
      clust=TRUE
    }

    #create the heatmap plot
    hmap <- Heatmap(mat,
                    name=name,
                    col=pal,

                    heatmap_legend_param=params,
                    row_title="Genes",
                    row_title_side="left",
                    row_title_gp=gpar(fontsize=15,
                                      fontface="bold"),
                    show_row_names=FALSE,
                    show_column_names=show_sample_ids,
                    row_names_gp=gpar(fontsize=3),

                    column_title="Samples",
                    column_title_side="bottom",
                    column_title_gp=gpar(fontsize=15,
                                         fontface="bold"),
                    column_title_rot=0,
                    row_dend_width=unit(8,"mm"),
                    column_dend_height=unit(22.5,"mm"),

                    top_annotation=hmap_anno_obj,
                    right_annotation = hmap_anno_obj_genes,
                    split=split,
                    row_gap = unit(5, "mm"),

                    clustering_distance_rows="euclidean",
                    clustering_method_rows=cluster.method,
                    clustering_method_columns = cluster.method,
                    cluster_columns = clust,
                    column_dend_reorder=FALSE)

  return(hmap)
}
```

```{r}
create_pf_heatmaps <- function(TMMCPM,sample_annots, gene_annots,
                               center_scale=TRUE,
                               ntop=1000,order_ids=NULL,
                               genelist=NULL,
                               split=NULL,
                               heatmap_colors=NULL){
  pf_genes <- gene_annots %>% 
    as.data.frame() %>% 
    dplyr::filter(genome=="PF3D7") %>% 
    pull(gene_id)
  
  TMMCPM_all_pf <- TMMCPM[rownames(TMMCPM) %in% pf_genes, ]
  dends <- dge_dendrograms(expnData = TMMCPM_all_pf, 
                           pheno = pull(sample_annots, analysis_factor_group, name = masigpro_id), 
                           method = "ward.D2",
                           genelist = genelist, 
                           ntop=ntop,
                           log=FALSE,
                           filterTopGenes = TRUE,
                           createDGE = FALSE)
  
  # dim(dends$TMMCPM)
  hmap_anno_obj <- create_annots_hmap(expn=dends$TMMCPM,
                                      pheno_df = sample_annots,
                                      cc=NULL,
                                      colorbar.height=2,
                                      cols = c("analysis_factor_group"))
  
  name <- ifelse(center_scale, "z-scores","log2_CPM")
  res <- complex_hmap(mat=dends$TMMCPM, 
               scale=center_scale,
               name=name,
               hmap_anno_obj = hmap_anno_obj$annoColumn,
               dge_dendrograms.res = dends,
               samp_dend_order=order_ids,
               color_palette=heatmap_colors,
               split=split,
               show_sample_ids=TRUE)
  
  return(list("dends"=dends,"heatmap"=res))
  
}
```

```{r}
aa_motif_search <- function(motif, aa_to_search){
  
  outname <- str_split_fixed(names(aa_to_search), pattern = " | ", n=2)[,1] %>% 
    gsub("-",".", .)
  matches <- str_extract_all(string = as.character(aa_to_search),
                             pattern = motif) %>% 
    set_names(outname) 
  coords <- str_locate_all(string = as.character(aa_to_search), 
               pattern = motif) %>% 
    set_names(outname) 
          

  idx <- sapply(matches,length) > 0
  matches[!idx] <- ""
  toFill <- max(sapply(matches,length))
  matches_clean <- lapply(matches, function(x){
    n <- length(x)
    dummy_n <- toFill - n
    x <- c(x[1:n], rep("",dummy_n))
  }) %>%
  as.data.frame() %>%
  pivot_longer(cols = everything(),
               names_to = "aa_name",
               values_to = "match") %>% 
  arrange(aa_name) %>% 
  group_by(aa_name) %>% 
  filter(!duplicated(match)) %>% 
  mutate(keep=case_when(
    n() > 1 ~ !grepl("^$", match),
    TRUE ~ TRUE
  )) %>%
  filter(keep) %>% 
  summarise(n_matches=n(),
            match=paste(match[match != ""], collapse = "; ")) %>%
  ungroup()  %>% 
  mutate(n_matches=ifelse(grepl("^$", match), 0, n_matches))
  
  
  coords_clean <- purrr::map_dfr(names(coords), function(x){
    locs <- as.data.frame(coords[[x]])
    n <- nrow(locs)
    reps <- max(1,n)
    if(n < 1){
      df <- data.frame(aa_name=rep(x, reps)) %>% 
        mutate(start=NA_integer_, end=NA_integer_)
    }else{
      df <- data.frame(aa_name=rep(x, reps)) %>% 
       bind_cols(locs)
    }
  }) %>% 
    mutate(position=ifelse(!is.na(start), paste(start,end,sep="-"), "")) %>%
    group_by(aa_name) %>% 
    mutate(position = paste(position, collapse = "; ")) %>%
    ungroup() %>% 
    select(aa_name, position) %>% 
    distinct()
  
  results <- matches_clean %>% 
    left_join(coords_clean, by="aa_name")

  return(results)
}
```

# Sample Manifest and Expression Data

```{r}
pf_samples <- read.csv("samples/Kappe_s_2022.04_PF_LiverStage_ParasitesPaper_sample_manifest_7.25.2022.csv") %>% 
  set_rownames(.$sample_id_cat)

dim(pf_samples)
table(pf_samples$analysis_group)
```

```{r}
cb_se <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_SummarizedExperiment.RDS")

cb_se
```

```{r}
cb_eset <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_expressionSet.RDS")

# cb_eset
```

```{r}
counts <- read.csv("raw_counts/kappe_s_PfHsMmu_GRCh38_GRCm39_Pf3D7v58_raw_counts.csv")

dim(counts)
# head(counts)
```

```{r}
TPM <- read.csv("expression_data/kappe_s_PfHsMmu_PF3D7_TPM.csv")

head(TPM)
```

# Genome References 

```{r}
PF3D7 <- AnnotationDbi::loadDb("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7.sqlite")
# PF3D7
```

```{r}
genome <- Biostrings::readDNAStringSet("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_Genome.fasta")
names(genome) <- str_split_fixed(names(genome), pattern = " \\|", n=3)[,1]
```

```{r}
CDS <- cdsBy(PF3D7,by = "tx", use.names=TRUE)

minus_strand <- sapply(CDS, function(gr) (strand(gr) == "-")@values)
length(CDS)
```

```{r}
proteins_aa <- Biostrings::readAAStringSet("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_AnnotatedProteins.fasta")

head(proteins_aa)
length(proteins_aa) #5389
```

```{r}
cds_fasta <- Biostrings::readDNAStringSet("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_AnnotatedCDSs.fasta")

head(cds_fasta)
length(cds_fasta) #5389
```

```{r}
test_cds_gr <- CDS[minus_strand][["PF3D7_0100200.1"]]
cds_seq <- BSgenome::getSeq(genome, test_cds_gr)

# negative strand does not appear to affect the translation of the aa sequence
t1 <- xscat(translate(cds_seq[1]), translate(cds_seq[2])) %>% 
  narrow(.,end = 331) %>% 
  unname()

t2 <- proteins_aa[grep("PF3D7_0100200.1", names(proteins_aa))] %>% 
  unname()

t3 <- Biostrings::translate(cds_fasta[grep("PF3D7_0100200.1", names(cds_fasta))]) %>% 
    narrow(.,end = 331) %>% 
    unname()

identical(class(t1),class(t2)) # TRUE
identical(as.character(t1),as.character(t2)) # TRUE
identical(as.character(t1), as.character(t3))
```


# Reference Data

```{r}
conflicted::conflict_prefer("filter","dplyr")
conflicted::conflict_prefer("select","dplyr")
conflicted::conflict_prefer("arrange","dplyr")
conflicted::conflict_prefer("mutate","dplyr")
```

```{r}
apicoplast_genes <- read.delim("expression_data/gene_annots/PF3D7_Apicoplast_GenesBySubcellularLocalization_Summary.txt", sep="\t") %>% 
  janitor::clean_names()

head(apicoplast_genes)
dim(apicoplast_genes) #506
```

```{r warning=FALSE, eval=FALSE}
apicoTP <- tabulizer::extract_tables("references/PEXEL/pone.0036598.s009.pdf") %>% 
  lapply(., as.data.frame) %>% 
  bind_rows() %>% 
  set_colnames(., value=.[1,]) %>%
  janitor::clean_names() %>% 
  filter(!grepl("Gene", gene_id1)) %>% 
  pivot_longer(cols = everything(),
               names_to = "type",
               values_to = "value") %>%
  filter(!grepl("description", type)) %>% 
  mutate(type=gsub("[12]$","", type)) %>% 
  pivot_wider(names_from = type,
              values_from = value) %>% 
  unnest(cols=c("gene_id"))

dim(apicoTP)
# write.csv(apicoTP, "apicoTP_old_ids.csv", row.names = FALSE)
```

```{r}
apicoTP_update <- read.csv("expression_data/gene_annots/apicoTP_GeneByLocusTag_Summary.csv") %>% 
  janitor::clean_names() %>% 
  select(-source_id) %>% 
  distinct()

# head(apicoTP_update)
dim(apicoTP_update) #535   8
# table(duplicated(apicoTP_update$gene_id))
# unique(c(apicoplast_genes$gene_id, apicoTP_update$gene_id)) %>% length() #757
```

```{r}
mito_genes <- rowData(cb_se) %>% 
  as.data.frame() %>% 
  filter(seqnames == "Pf3D7_MIT_v3" | grepl("mitochondrial", description, ignore.case = TRUE))

dim(mito_genes)
```

# VAR Genes Expression

```{r}
sheet_names <- paste0("PfCVG_",c("var","rifin","stevor","Pfmc-2TM","PfACS","CLAG"))
```

```{r}
conflicted::conflict_prefer("select","dplyr")
cvgs <- purrr::map_dfr(sheet_names, function(sheet){
  openxlsx::readWorkbook("references/CVGs.xlsx",sheet) %>% 
    mutate(type=sheet) %>% 
    select(type, everything())
}) %>% 
  janitor::clean_names()

head(cvgs)
dim(cvgs) #316
table(cvgs$type)
```

```{r}
cvg_expn <- log2(assay(cb_se,"cpm")+1)
include <- intersect(cvgs$gene_id, rownames(cvg_expn))
cvg_expn <- cvg_expn[include,]

head(cvg_expn[,1:5])
dim(cvg_expn) #139  17
# any(duplicated(rownames(cvg_expn)))
range(cvg_expn)
```

```{r}
# cvgs %>% 
#   group_by(type) %>% 
#   dplyr::count() %>% 
#   write.csv(.,"cvgs_pexel_domains/cvgs_counts_table.csv", row.names = F)

cvgs %>%
  filter(gene_id %in% rownames(cvg_expn)) %>%
  group_by(type) %>%
  dplyr::count() 
#   write.csv(., "cvgs_pexel_domains/cvgs_counts_table_filtered_by_expression.csv", row.names = F)
```

```{r}
missing <- setdiff(cvgs$gene_id, rownames(cvg_expn))
# length(missing)

missing_raw_counts <- counts %>% 
  filter(gene_name %in% missing) %>% 
  pivot_longer(cols = -gene_name,
               names_to = "sample_id_cat",
               values_to = "counts") %>% 
  inner_join(., pf_samples, by="sample_id_cat") %>% 
  group_by(time_point) %>% 
  summarise(mean=mean(counts),
            median=median(counts), 
            sd=sd(counts))

# write.csv(missing_raw_counts,"cvgs_pexel_domains/missing_raw_counts_summary.csv", row.names = FALSE)
```

```{r}
cvg_expn_long <- cvg_expn %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols = -gene_id,
               names_to = "sample_id_cat",
               values_to = "log2CPM") %>% 
  inner_join(., pf_samples, by="sample_id_cat") %>%
  left_join(., cvgs, by=c("gene_id")) %>%
  select(gene_id:log2CPM, type, product, everything())

cvg_expn_long
# table(cvg_expn_long$type)
```

```{r eval=FALSE}
# Day 6 boxplots
purrr::map(sheet_names, function(cvg_type){
  input <- cvg_expn_long %>% 
    filter(time_point == "Day6") %>% 
    filter(type==cvg_type) %>% 
    group_by(gene_id) %>% 
    mutate(mean_expn=mean(log2CPM)) %>% 
    ungroup() %>% 
    arrange(desc(mean_expn)) %>% 
    mutate(gene_id=factor(gene_id, levels = unique(gene_id))) %>% 
    select(gene_id:log2CPM, mean_expn, everything())
  
  n <- length(unique(input$gene_id))
  fig_width <- ifelse(n<=11, 4, 15)
  
  # svglite(paste0("figures/combat_seq/cvgs_pexel/cvg_", cvg_type,"_expression_boxplot.svg"),
  #         fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"), height = 5, width = fig_width)
  print(ggplot(input, 
         aes(x=gene_id, y=log2CPM, fill=analysis_factor_group)) +
    geom_boxplot(color="grey50", alpha=0.2) +
    geom_point(aes(color=analysis_factor_group), 
               position = position_jitter(width = 0.3),
               size=2) +
    scale_fill_manual(values=ggpubr::get_palette("npg",4)[4]) +
    scale_color_manual(values=ggpubr::get_palette("npg",4)[4]) +
    facet_wrap(~type, scale="free_x") +
    labs(x="") +
    theme_classic() +
    theme(axis.text.x = element_text(angle=25, vjust=1,hjust=1),
          plot.margin = margin(l=6,unit="mm"),
          legend.position = "top"))
  # dev.off()
})
```


```{r fig.height=10, fig.width=10}
colanno <- pf_samples %>% 
  select(time_point=analysis_factor_group) %>% 
  HeatmapAnnotation(df = ., which = "column",
                    col = list(time_point=ggpubr::get_palette("jco",5) %>%
                                          set_names(unique(pf_samples$analysis_factor_group))))
# colanno

rowanno <- cvgs %>% 
  filter(gene_id %in% rownames(cvg_expn)) %>% 
  select(-product, CVG_type=type) %>% 
  column_to_rownames("gene_id") 

gap_index <- rowanno %>% 
  group_by(CVG_type) %>%
  mutate(n=n()) %>% 
  ungroup() %>% 
  arrange(n) %>% 
  mutate(CVG_type=factor(CVG_type, levels=unique(CVG_type))) %>% 
  select(-n)
  # mutate(nrow=1:nrow(.)) %>% 
  # group_by(CVG_type) %>%
  # summarise(index=max(nrow)) %>% 
  # pull(index, name = CVG_type)


rowanno <- rowanno %>%
  mutate(CVG_type=factor(CVG_type, levels=unique(gap_index$CVG_type))) %>%
  arrange(CVG_type)
gene_order <- rownames(rowanno)
rowanno <- rowanno %>% 
  HeatmapAnnotation(df = . , which = "row",
                    col = list(CVG_type=ggpubr::get_palette(palette = "npg", 6) %>%
                                        set_names(unique(rowanno$CVG_type))))

heat_colors <- colorRampPalette(viridis::turbo(10))(299)

# pdf("figures/combat_seq/cvgs_pexel/CGVs_timepoints_log2CPM_heatmap.pdf", height = 12, width = 10)
Heatmap(matrix = cvg_expn[gene_order, ], 
        col = heat_colors,
        name = "log2 CPM",
        split = gap_index,
        top_annotation = colanno, 
        left_annotation = rowanno, 
        show_column_names = FALSE,
        row_title_gp = gpar(fontsize = 6),
        # row_title_rot = 270,
        row_names_gp = gpar(fontsize = 6))
# dev.off()
```

# Search for PEXEL 

From PlasmoDB
Pexel genes: 
 R at position 1, any charge-neutral amino acid at position 2, L at position 3, any charge-neutral amino acid at position 4, and E, D, or Q at position 5 (R.L.E/D/Q),
 
PlasmoDB search: 
 Pattern	RXLX[EQD]
Organism	Plasmodium falciparum 3D7

```{r}
#For comparison from PlasmoDB
pexel_genes <- read.delim("expression_data/gene_annots/PEXEL_canonical_motif_TranscriptRecordClass_Summary.txt") %>% 
  janitor::clean_names()

head(pexel_genes)
dim(pexel_genes) #1,437   11
```

```{r}
# For comparison 
nc_pexel_motifs <- tabulizer::extract_tables("references/PEXEL/mmi13024_supplemental_table2.pdf")


nc_pexel_motifs_clean <- purrr::map(nc_pexel_motifs, function(x){
  df <- as.data.frame(x) 
  df %>%
    rename_all(~c("gene_id","description","motif")) %>%
    mutate(type="non-canonical")
}) %>% 
  bind_rows()

nc_pexel_motifs_clean %>% 
  filter(gene_id != "") %>% 
  pull(gene_id) %>% 
  unique() %>%
  length() #388 
```

```{r}
pexel_motifs <- data.frame(type=c("Canonical", rep("non-canonical",5)),
                           motif=c("R.L.E/D/Q","R.I.E/D/Q","K.L.E/D/Q",
                                   "H.L.E/D/Q","H.I.E/D/Q","A.L.E/D/Q"),
                           motif_regex=c("R.L.[EDQ]","R.I.[EDQ]","K.L.[EDQ]",
                                   "H.L.[EDQ]","H.I.[EDQ]","A.L.[EDQ]")) %>% 
  mutate(motif_name=paste0(tolower(type),"_",motif) %>% 
           gsub("[-\\.\\/]","", .))

pexel_motifs
# write.csv(pexel_motifs, "cvgs_pexel_domains/pexel_motifs.csv", row.names = FALSE)
```

```{r}
aa_lengths <- width(proteins_aa) 
ok_len <- aa_lengths >= 150

nterm_150aa <- proteins_aa[ok_len] %>% 
  narrow(., start = 1, width = 150) 

short_aa <- proteins_aa[!ok_len] %>% 
  narrow(., start = 1, width = aa_lengths[!ok_len]) 

nterm_150aa <- c(nterm_150aa, short_aa)


trimmed_aa_lens <- width(nterm_150aa)
ok_100_len <- trimmed_aa_lens >= 100

nterm_100aa <- nterm_150aa[ok_100_len] %>% 
  narrow(., start = 1, width = 100)

short_100aa <- nterm_150aa[!ok_100_len] %>% 
  narrow(., start = 1, width = trimmed_aa_lens[!ok_100_len]) 

nterm_100aa <- c(nterm_100aa, short_100aa)

# head(nterm_150aa)
# head(nterm_100aa)
# 
# tail(nterm_150aa)
# tail(nterm_100aa)
# Biostrings::writeXStringSet(nterm_100aa, filepath = "cvgs_pexel_domains/PF3D7_100aa_nterm_protein.fasta")
```

```{r eval=FALSE}
pexel_motif_patterns <- pull(pexel_motifs, motif_regex, name = motif_name)

pexel_hits <- purrr::map(.x = pexel_motif_patterns, 
                         .f = aa_motif_search, 
                         aa_to_search = nterm_150aa)

lapply(pexel_hits, head)
lapply(pexel_hits, dim)
# saveRDS(pexel_hits,"cvgs_pexel_domains/pexel_motif_search_150aa_results_list.RDS")
```

```{r}
pexel_hits_all <- read.csv("cvgs_pexel_domains/pexel_motif_search_150aa_results.csv", row.names = 1)

# new_col_names <- lapply(pexel_motifs$motif_name, function(x) paste(c("n_matches","match","position"), x, sep="_")) %>% unlist()
# pexel_hits_all <- purrr::reduce(pexel_hits, left_join, by="aa_name") %>%
#   rename_at(vars(matches("match|position")), ~new_col_names)
# 
head(pexel_hits_all)
# dim(pexel_hits_all) #5389   19
# 
# pexel_hits_all[,grep("n_matches", colnames(pexel_hits_all))] %>% 
#   sapply(., table)
# write.csv(pexel_hits_all, "cvgs_pexel_domains/pexel_motif_search_150aa_results.csv")
```

# Integrate TM, Api, Pexel Annotations 

```{r}
pexel_hits_bool <- pexel_hits_all %>% 
  select(aa_name, matches("n_matches")) %>%
  mutate(transcript_id=gsub("\\.p[0-9]","", aa_name),
         gene_id=gsub("\\.[0-9]+","",transcript_id)) %>% 
  select(gene_id,
         matches("n_matches")) %>%
  group_by(gene_id) %>% 
  mutate(across(matches("n_matches"), .fns = sum)) %>% 
  ungroup() %>% 
  distinct() %>% 
  pivot_longer(cols=matches("n_matches"),
               names_to = "type",
               values_to = "n_matches") %>%
  mutate(type=gsub("n_matches_","",type) %>%
           gsub("ical", "",.),
         has_pexel_domain=n_matches>0) %>% 
  group_by(gene_id) %>%
  mutate(any_pexel_domain=any(has_pexel_domain)) %>% 
  ungroup() %>% 
  pivot_wider(id_cols = c(gene_id,any_pexel_domain),
              names_from = type,
              values_from = has_pexel_domain,
              names_glue = "{.value}_{type}")

# pexel_hits_bool #5,318
# table(duplicated(pexel_hits_bool$gene_id)) # FALSE
# table(pexel_hits_bool$any_pexel_domain) #2346
# table(pexel_hits_bool$has_pexel_domain_canon_RLEDQ) #669 genes
```

```{r}
apicoplast_ids <- unique(c(apicoplast_genes$gene_id, apicoTP_update$gene_id))

Pf_domain_annots <- read.delim("expression_data/gene_annots/P33D7_transmembrane_signal_peptides_Summary.txt", sep = "\t") %>% 
  janitor::clean_names()  %>% 
  filter(!gene_id %in% mito_genes$gene_id) %>% 
  mutate_at(vars(signal_p_peptide,x_tm_domains), ~gsub("N/A",NA, .)) %>%
  
  group_by(gene_id) %>% 
  mutate(has_signal_peptide=any(!is.na(signal_p_peptide)),
         has_tm_domain=case_when(
           any(x_tm_domains > 0) ~ TRUE,
           TRUE ~ FALSE
         ),
         apicoplast_protein=gene_id %in% apicoplast_ids) %>% 
  ungroup() %>% 
  #de-duplicate gene_ids 
  select(-source_id, -signal_p_peptide, 
         -x_tm_domains,-matches("p_fam"), 
         -product_description) %>% 
  distinct() %>% 
  
  left_join(., pexel_hits_bool, by="gene_id") %>%
  select(gene_id,
         any_pexel_domain,
         matches("peptide|tm.domain"),
         apicoplast_protein,
         matches("has_pexel_domain"),
         everything())

head(Pf_domain_annots)
dim(Pf_domain_annots)  # 5598    8
# any(duplicated(Pf_domain_annots$gene_id)) #FALSE
# write.csv(Pf_domain_annots, "cvgs_pexel_domains/pexel_tm_apicoplast_gene_annots.csv", row.names = F)
```

```{r}
table(Pf_domain_annots$has_signal_peptide)
table(Pf_domain_annots$has_tm_domain)
table(Pf_domain_annots$apicoplast_protein)
table(Pf_domain_annots$any_pexel_domain)
```

```{r}
Pf_domain_annots[,grep("pexel", colnames(Pf_domain_annots))] %>% 
  sapply(., function(x) sum(x, na.rm = T)) %>% 
  as.data.frame() 

# Pf_domain_annots[,grep("pexel", colnames(Pf_domain_annots))] %>% 
#   sapply(., function(x) sum(is.na(x))) 
```


# Venn Diagram

```{r}
library(ggVennDiagram)
```

```{r}
cols <- c("has_signal_peptide","has_tm_domain","apicoplast_protein")
gene_lists <- lapply(cols, function(col){
  idx <- Pf_domain_annots[[col]]
  Pf_domain_annots$gene_id[idx]
})
names(gene_lists) <- cols

# gene_lists
sapply(gene_lists, length)
```

```{r fig.height=7, fig.width=7, eval=FALSE}
# svglite::svglite("figures/tm_signal_apicoplast_venn.svg", height = 10, width = 10, fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"))
ggVennDiagram(gene_lists,
              label_alpha=0.2,
              label_geom = "label",
              set_size=6, 
              label_size=6) + 
  scale_color_manual(values=c(viridis::viridis(n = 3))) +
  scale_fill_gradient(low="aliceblue", high = "deepskyblue3", 
                      name="Number of Genes")
  # theme(axis.text = element_text(size=4))
# dev.off()
```


# Pexel Genes Ranking 

```{r}
library(UpSetR)
```

Keep genes that have TM or signal peptides, but are NOT apicoplast genes

```{r}
input_venn <- Venn(gene_lists)
input_dat <- process_data(input_venn)
```

```{r}
input_dat@region[["count"]] %>% sum()
length(unique(unlist(gene_lists)))
```

```{r}
of_interest <- input_dat@region %>% 
  as_tibble() %>% 
  select(gene_id=item,count, name) %>% 
  filter(grepl("^has_signal_peptide$|^has_tm_domain$|^has_signal_peptide..has_tm_domain$",name)) %>% 
  unnest(cols=gene_id)  %>% 
  left_join(.,Pf_domain_annots, by="gene_id") %>%
  select(gene_id, matches("pexel"),
         everything(),
         -count, -name)

of_interest %>% dim()

table(of_interest$any_pexel_domain)
table(of_interest$apicoplast_protein)
# write.csv(of_interest,"cvgs_pexel_domains/pexel_tm_signal_peptide_genes_of_interest.csv", row.names = F)
```

```{r}
cpm <- assay(cb_se,"cpm") %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols = -gene_id, 
               names_to = "sample_id_cat",
               values_to = "CPM")

head(cpm)
# pdf("figures/combat_seq/cvgs_pexel/all_genes_expression_CPM_histogram.pdf", height = 5, width = 7)
hist <- ggplot(cpm, aes(x=log2(CPM+1))) +
  geom_histogram(binwidth  = 0.5) +
  scale_x_continuous(breaks = seq(0,20, by=2)) +
  geom_vline(xintercept = log2(10)) +
  geom_vline(xintercept = log2(5)) +
  theme_classic()
# hist
# dev.off()

```


```{r}
threshold <- 10
col_name <-glue::glue("samples_expressing_at_GE.{threshold}CPM")

pexel_tm_api_expn <- cpm %>% 
  filter(gene_id %in% of_interest$gene_id) %>%
  left_join(., pf_samples, by="sample_id_cat") %>% 
  group_by(analysis_factor_group,
           time_point,infection_status, gene_id) %>% 
  # mutate(!! col_name :=sum(CPM >= threshold)) %>% 
  select(gene_id:CPM,analysis_factor_group, 
         matches("samples_expressing"),
         everything()) %>% 
  summarise(!! col_name :=sum(CPM >= threshold),
            mean_CPM=mean(CPM),
            median_CPM=median(CPM),
            sd_CPM=sd(CPM),
            .groups = 'keep') %>% 
  ungroup() 

# pexel_tm_api_expn
length(unique(pexel_tm_api_expn$gene_id)) # 1,112
```

```{r}
pexel_annots <- of_interest %>%
  # ensure gene has expression 
  filter(gene_id %in% unique(pexel_tm_api_expn$gene_id)) %>% 
  select(gene_id, c(has_pexel_domain_canon_RLEDQ:has_tm_domain)) %>% 
  pivot_longer(cols=c(has_pexel_domain_canon_RLEDQ:has_tm_domain),
               values_to = "value",
               names_to = "stat") %>% 
  mutate(type=gsub("^.+_([cn][a-z]+)_[A-Z]+", "\\1", stat)) %>% 
  group_by(gene_id, type) %>% 
  # convert the number of non-canonical motifs to percent out of 5 total non-canonical motifs possible. 
  mutate(total=n(),
         value=value/n()) %>% 
  ungroup() %>% 
  select(gene_id:value)

# pexel_annots
pexel_tm_api_rank_data <- pexel_tm_api_expn %>% 
  filter(!grepl("Day0", time_point)) %>% # dont want SPZ expression to be part of the rank calc
  pivot_longer(cols=samples_expressing_at_GE.10CPM,
               values_to = "value",
               names_to = "stat") %>%
  mutate_at(vars(stat), ~ifelse(grepl("samples_expressing", .), paste0(time_point,"_", .), .)) %>%
  select(gene_id, stat, value) %>%
  distinct() %>%
  arrange(gene_id) %>% 
  group_by(gene_id) %>% 
  # convert to percent of expressors to be out of 12 total samples for all time points
  mutate(value=round(value/12,digits = 2)) %>%
  ungroup() %>%
  bind_rows(pexel_annots) %>%
  group_by(gene_id) %>%
  mutate(rank=sum(value)) %>%
  ungroup() %>%
  mutate(rank_scaled=round(scales::rescale(rank, to = c(1,10)), digits=2)) %>%
  arrange(desc(rank))

pexel_tm_api_ranks <- pexel_tm_api_rank_data %>% 
  select(gene_id, rank_scaled) %>% 
  distinct() %>% 
  as.data.frame() %>% 
  set_rownames(.$gene_id)

# pexel_tm_api_ranks
# range(pexel_tm_api_rank_data$rank_scaled)
# length(unique(pexel_tm_api_ranks$gene_id))
quantile(pexel_tm_api_ranks$rank_scaled, probs=seq(0,1,length.out=11))
```



```{r fig.height=10, fig.width=11}
top_hits <- pexel_tm_api_rank_data %>% 
  filter(rank_scaled >= 6.17) %>% 
  select(gene_id,value,stat)  %>% 
  mutate_at(vars(stat), ~gsub("samples_expressing_at_","expn_", .)) %>% 
  pivot_wider(id_cols=stat, 
              names_from = gene_id,
              values_from = value) %>% 
  as.data.frame() %>% 
  column_to_rownames("stat") %>% 
  as.matrix()
  

# top_hits
# pheatmap::pheatmap(top_hits,
#                    annotation_col = select(pexel_tm_api_ranks, rank_scaled))

col_fun = circlize::colorRamp2(seq(6,10, length.out=5),
                               RColorBrewer::brewer.pal(5,name = "Reds"))

input <- pexel_tm_api_ranks %>%
  filter(gene_id %in% colnames(top_hits)) %>%
  select(rank_scaled)
HA <- HeatmapAnnotation(df = input,
                        col =  list(rank_scaled = col_fun),
                        simple_anno_size = unit(4,"cm"))
# pdf("figures/combat_seq/cvgs_pexel/pexel_TM_signal_peptide_genes_ranked_heatmap.pdf", height = 10, width = 20)
ComplexHeatmap::Heatmap(matrix = top_hits,
                        col = colorRampPalette(RColorBrewer::brewer.pal(9,"Blues"))(50),
                        top_annotation = HA)
# dev.off()
```


```{r}
of_interest_rank_expn_stats <- pexel_tm_api_expn %>% 
  pivot_wider(id_cols = gene_id, 
              names_from = time_point, 
              values_from = c(samples_expressing_at_GE.10CPM:sd_CPM), 
              names_glue = "{time_point}_{.value}") %>% 
  select(colnames(.)[order(colnames(.))]) %>% 
  left_join(., pexel_tm_api_ranks, by="gene_id") %>%
  left_join(., of_interest, by="gene_id") %>%
  arrange(desc(rank_scaled)) %>% 
  select(gene_id, rank_scaled, 
         any_pexel_domain,has_pexel_domain_canon_RLEDQ, 
         has_signal_peptide,
         matches("Day[2456]"),
         matches("Day0"),
         everything())


of_interest_rank_expn_stats %>% dim()
# write.csv(of_interest_rank_expn_stats, "cvgs_pexel_domains/pexel_tm_signal_peptide_genes_of_interest_ranked.csv", row.names = FALSE)
```


```{r}
input <- of_interest %>% 
  filter(any_pexel_domain) %>% 
  select(gene_id, matches("has_pexel")) %>% 
  mutate_at(vars(matches("has_pexel")), ~as.numeric(.)) %>% 
  as.data.frame()

# pdf("figures/combat_seq/cvgs_pexel/pexel_TM_signal_peptide_genes_upset.pdf", height = 5, width = 8)
UpSetR::upset(data=input, 
              nintersects = NA,
              order.by=c("freq","degree"), 
              group.by = "degree")
# dev.off()
```

```{r}
signal_peps_upset <- of_interest %>% 
  filter(has_signal_peptide) %>% 
  select(gene_id, matches("has_pexel")) %>% 
  mutate_at(vars(matches("has_pexel")), ~as.numeric(.)) %>% 
  as.data.frame()

dim(signal_peps_upset)

# pdf("figures/combat_seq/cvgs_pexel/pexel_and_signal_peptide_ONLY_genes_upset.pdf", height = 5, width = 8)
UpSetR::upset(data=signal_peps_upset, 
              nintersects = NA)
# dev.off()
```

```{r fig.height=7, fig.width=7}
cols <- grep("has_pexel",colnames(of_interest), value=TRUE)
gene_lists_sub <- lapply(cols, function(col){
  idx <- of_interest[[col]]
  of_interest$gene_id[idx]
})
names(gene_lists_sub) <- cols
# sapply(gene_lists_sub, length)

# svglite::svglite("figures/pexel_tm_apicoplast_venn.svg", height = 10, width = 10, fix_text_size=FALSE, system_fonts = list(sans = "Open Sans"))
ggVennDiagram(gene_lists_sub,
              label_alpha=0.2,
              label_geom = "label",
              set_size=6, 
              label_size=6) 
# dev.off()
```


# Expression Plots 

### PTEX Complex

-	EXP2 (PF3D7_1471100)
-	PTEX 150 (PF3D7_1436300)
-	HSP101 (PF3D7_1116800)
-	TRX2 (PF3D7_1345100)
-	PTEX 88 (PF3D7_1105600)

```{r}
ptex_ids <- c("PF3D7_1471100","PF3D7_1436300","PF3D7_1116800","PF3D7_1345100","PF3D7_1105600")
```

```{r}
ptex_expn <- log2(assay(cb_se,"cpm")+1)
include <- intersect(ptex_ids, rownames(ptex_expn))
ptex_expn <- ptex_expn[include,pf_samples$sample_id_cat]
head(ptex_expn[,1:5])
dim(ptex_expn) #211 of 278 had expression 
```

```{r}
ptex_expn_long <- ptex_expn %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  pivot_longer(cols = -gene_id,
               names_to = "sample_id_cat",
               values_to = "log2CPM") %>% 
  inner_join(., pf_samples, by="sample_id_cat") %>%
  select(gene_id:log2CPM, everything())

head(ptex_expn_long)
```

```{r fig.width=10}
labels <- c("PF3D7_1471100"="EXP2","PF3D7_1436300"="PTEX150", "PF3D7_1116800"="HSP101", "PF3D7_1345100"="TRX2", "PF3D7_1105600"="PTEX88")

# svg("figures/combat_seq/cvgs_pexel/PTEX_complex_genes_expression_boxplot.svg", height = 5, width = 10)
ggplot(ptex_expn_long, aes(x=gene_id, y=log2CPM, fill=analysis_factor_group)) +
  geom_boxplot(alpha=0.2, color="grey60") +
  geom_point(aes(color=analysis_factor_group),
             position = position_jitterdodge(jitter.width = 0.1),
             size=2, alpha=0.75) +
  scale_x_discrete(labels=labels) +
  scale_color_viridis_d(end=0.925,direction = -1) +
  scale_fill_viridis_d(end=0.925,direction = -1) +
  labs(x="") +
  theme_classic()
# dev.off()
```


### Top Hits 

```{r}
pexel_expn <- log2(assay(cb_se,"cpm")+1)
pexel_expn <- pexel_expn[pexel_tm_api_ranks$gene_id,]
```

```{r}
highest_rank <- pexel_tm_api_ranks %>% 
  filter(rank_scaled > 6.17)

include <- intersect(highest_rank$gene_id, rownames(pexel_expn))
top_hits_expn <- pexel_expn[include,pf_samples$sample_id_cat]
colnames(top_hits_expn) <- pf_samples$analysis_group_id

head(top_hits_expn[,1:5])
dim(top_hits_expn) 
```


```{r fig.height=10, fig.width=10}
ann_colors <- list(analysis_factor_group=ggpubr::get_palette("jco",5) %>% 
                     set_names(., unique(pf_samples$analysis_factor_group)))

sample_groups <- select(pf_samples, analysis_factor_group) 
rownames(sample_groups) <- pf_samples$analysis_group_id

# range(scale(t(top_hits_expn)))
# pdf("figures/combat_seq/cvgs_pexel/pexel_top_ranked_genes.pdf", height = 15, width = 10)
pheatmap::pheatmap(top_hits_expn,
                   annotation_col = sample_groups,
                   annotation_colors = ann_colors,
                   fontsize = 5,
                   scale = "row")
# dev.off()
```



### Signal Peps

```{r}
signal_peps <- of_interest %>% 
  filter(has_signal_peptide) 

dim(signal_peps)
# signal_peps
```

```{r}
include <- intersect(signal_peps$gene_id, rownames(pexel_expn))
signal_peps_expn <- pexel_expn[include,pf_samples$sample_id_cat]
colnames(signal_peps_expn) <- pf_samples$analysis_group_id

head(signal_peps_expn[,1:5])
dim(signal_peps_expn) #211 of 278 had expression 
```

```{r}
ann_colors <- list(analysis_factor_group=ggpubr::get_palette("jco",5) %>% 
                     set_names(., unique(pf_samples$analysis_factor_group)))
```

```{r fig.height=10, fig.width=10}
sample_groups <- select(pf_samples, analysis_factor_group) 
rownames(sample_groups) <- pf_samples$analysis_group_id

# pdf("figures/combat_seq/cvgs_pexel/pexel_and_signal_peptide_ONLY_heatmap.pdf",
#     height = 15, width = 10)
pheatmap::pheatmap(signal_peps_expn,
                   annotation_col = sample_groups,
                   annotation_colors = ann_colors,
                   fontsize = 5,
                   scale = "row")
# dev.off()
```



# Share the Data 

```{bash}
destination="/active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/"
rsync -av --exclude .DS_Store cvgs_pexel_domains $destination
rsync -av --exclude .DS_Store figures $destination 
rsync -av --exclude .DS_Store presentations $destination 
```


# Session Information 

```{r}
sessionInfo()
```

