---
title: "PF Liver Stage Sample Clustering with Combat-seq Adjusted Counts"
author: "Jenny L Smith"
date: "`r Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(tidyr)
library(tibble)
library(magrittr)
library(dplyr)
library(stringr)
library(ggplot2)
library(gridExtra)
library(ggpubr)
library(patchwork)

library(limma)
library(edgeR)

getwd()
```

# Define Functions 

```{r}
#from https://github.com/mikelove/DESeq2/blob/master/R/plots.R
#Want to return the whole scores matrix so can examine 3d pca plots. 
plotPCA.DESeq.mod <- function(object, intgroup="condition", ntop=500, returnData=FALSE, PC3=FALSE)
{
  library(matrixStats)
  # calculate the variance for each gene
  rv <- rowVars(assay(object))
  
  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))
  
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )
  
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }
  
  # assembly the data for the plot - first 10 PCs
  # d <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], group=group, intgroup.df, name=colnames(object))
  n <- min(10, ncol(as.data.frame(pca$x)))
  d <- data.frame(as.data.frame(pca$x)[,1:n], group=group, intgroup.df, name=colnames(object))
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:10]
    rot <- pca$rotation[,1:10] #for first 10 PCs
    dat <- list("scores"=d,"rotation"=rot)
    return(dat)
  }
}

#Updated on 6/9/17 to use variance stabilized transformed data as input (not center scaled log2, like in princomp)
PCA <- function(expnData,phenovector,
                title="",round=TRUE,colorCodes=NULL,
                ntop=500, GOI=NULL){
  
  suppressPackageStartupMessages(library(DESeq2))
  library(ggplot2)
  #expnData is the raw counts (not normalized) has patient IDs as colnames and genes as rownames. 
  
  samples <- intersect(names(phenovector), colnames(expnData))
  countData <- expnData[,samples]
  phenovector <- phenovector[samples]
  colData <- as.data.frame(phenovector)
  
  if(round){
    countData <- round(countData, digits = 0)
  }
  
  #Create as DESeq data set object (dds)
  dds <- DESeqDataSetFromMatrix(countData = countData,
                                  colData = colData,
                                  design = ~ 1)
  
  #perform variance stabilized transformation 
  dds <- dds[ rowSums(counts(dds)) > 10, ]
  varianceStab <- vst(dds, blind = TRUE)
  
  #if given a list of genes of interest
  if (! is.null(GOI)){
    GOI <- intersect(GOI, rownames(assay(varianceStab)))
  }else{
    GOI <- 1:nrow(varianceStab)
  }
  
  #PCA data frame with the wieghts/loadings and eigen vectors
  pca.dat <- plotPCA.DESeq.mod(varianceStab[GOI,], 
                               intgroup = "phenovector", 
                               ntop = ntop,
                               returnData=TRUE)
  
  plots <- lapply(c(2:3), function(pc){
      percentVar <- attr(pca.dat$scores, which="percentVar")
      y_var <- paste0("PC",pc)
      
      p <- ggplot(data=pca.dat$scores, 
             aes_string(x="PC1", y=y_var, color="phenovector")) + 
        geom_point(size=3, alpha=0.75) +
        xlab(paste0("PC1: ",round(percentVar[1] * 100),"% variance")) +
        ylab(paste0(y_var,": ",round(percentVar[pc] * 100),"% variance")) +
        labs(title=title) +
        theme_classic() +
        theme(legend.position = "top")
      if(!is.null(colorCodes)){
        p <- p + 
          scale_color_manual(values=colorCodes)
      }
      return(p)
  })
  
  #Final Results object
  res <- list(dds, varianceStab,pca.dat, plots)
  names(res) <- c("dds", "vst","pca_data","pca_plots")
  
  if(is.character(GOI)){
    res[["GOI"]] <- GOI
  }
  
  return(res)
}
```

```{r}
#changed on 2/14/18, see bottom of Heatmaps_Function.r for the original one used.
dge_dendrograms <- function(expnData, pheno, method,
                            genelist=NULL,add.count=1, percent=0.05,
                            ntop=500,filterTopGenes=FALSE, createDGE=TRUE,log=FALSE){
  #df with count data, patient IDs are column names and rows are genes.
  #pheno is a character vector with patient IDs as names, and the status for each in each group (eg pos,neg)
  #genelist is a character vector with the genes of interest
  #percent is the % of samples in the input expn matrix that must express a gene at 1 CPM. Filter to remove low count genes.
  #set log=TRUE if providing a log2 expression dataframe.
  #filterTopGenes shuold be a logical. If TRUE filter top 1000 most varied genes.
  suppressPackageStartupMessages(require(edgeR))
  suppressPackageStartupMessages(library(dendextend))
  suppressPackageStartupMessages(library(matrixStats))

  expnData <- expnData[, intersect(names(pheno), colnames(expnData))] #ensure correct order, drop rows with nas just in case
  if(createDGE){
    dge <- DGEList(counts = expnData)
    #keep the rows (genes) with at least 1 CPM in a minimum of 2 samples or X% of samples. 
    keep.dge <- rowSums(cpm(dge) >= 1) >= max(2,(percent*ncol(expnData))) 
    # subset for those genes with cmp >= 1 per gene in samples
    dge <- dge[keep.dge,] #
    dge <- calcNormFactors(dge)
    TMMCPM <- cpm(dge, normalized.lib.sizes = TRUE,
                  log = TRUE, prior.count = add.count)

  }else{
    TMMCPM <- as.matrix(expnData)
    if(!log){
      TMMCPM <- log2(expnData+add.count) #log2 transform counts
    }
  }

  if(is.null(genelist) & filterTopGenes){
      # calculate the variance for each gene
      rv <- rowVars(as.matrix(TMMCPM))
      # select the ntop genes by variance
      select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
      #select the top 1000 most varied genes
      TMMCPM <- TMMCPM[select,]
  }else if(!is.null(genelist)){
    TMMCPM <- TMMCPM[which(rownames(TMMCPM) %in% genelist), ] #subset the matrix to genes of interest
  }


  d1 <- dist(t(TMMCPM), method = "euclidean", diag = FALSE,
             upper = FALSE) #sample distances WITHOUT SCALING
  d2 <- dist(TMMCPM, method = "euclidean", diag = FALSE,
             upper = TRUE) #gene distances WITHOUT SCaling
  samp.c1 <- hclust(d1, method = method, members = NULL) #sample clustering
  gene.c2 <- hclust(d2, method = method, members = NULL) #gene clustering
  list <- list(TMMCPM,samp.c1,gene.c2)
  names(list) <- c("TMMCPM","samp.c1", "gene.c2")

  return(list)
}

```

```{r}
create_annots_hmap <- function(expn, pheno_df, cols,
                               goi=NULL,cc=NULL, 
                               colorbar.height=5){
  #expn is the normalized expression values with genes as rownames
  #gene list a character vector
  #goi are genes of interest to highlight on the Heatmap. Character vector of gene symbols
  #cc are color codes in WHICH FORMAT?

  #cols is a character vector of column names

  #Example on how to use the reuslts

  #hmap <- ComplexHmap(XXXX, hmap_anno_obj=res$annoColumn, XXX)
  # draw(hmap + res$geneLabels, heatmap_legend_side="right", annotation_legend_side="right")

    library(dplyr)
    suppressPackageStartupMessages(library(ComplexHeatmap))

    #subset the expression matix and the phenotype dataframe
    anno <- pheno_df %>%
      filter(rownames(.) %in% colnames(expn)) %>% 
      rownames_to_column("id") %>% 
      dplyr::select(id, all_of(cols)) %>%
      mutate(id = factor(id, levels = colnames(expn))) %>%
      arrange(id) %>%  #ensure same order as the expn matrix
      column_to_rownames("id")
  
    #if no color codes provided, create one for each column
    if(is.null(cc)){
      pals <- c("npg", "aaas", "lancet", "jco", 
                    "ucscgb", "uchicago", "simpsons",
                    "rickandmorty")
      #if there are more columns than color palettes, just repeat the set of available color palettes
      if(length(ncol(anno)) > length(pals)){
        pals <- rep(pals,  ceiling(ncol(anno)/length(pals)))
      }
      #define unique colors for each column
      cc <- purrr::map(1:ncol(anno), function(i){
          grps <- unique(anno[[i]])
          ggpubr::get_palette(pals[i], k=length(grps)) %>% 
            magrittr::set_names(grps)
      })
      names(cc) <- colnames(anno)
    }
    
    #legend graphical parameters
    n <- max(sapply(cc, length))
    nrow <- ifelse(n <= 6, n, 6)
    ncol=ceiling(n/nrow)
    params <- list(show_legend=TRUE,
                  labels_gp= gpar(fontsize=12),
                  title_gp= gpar(fontsize=16),
                  nrow = nrow,
                  ncol=3,
                  by_row=TRUE)
    
    #Create the complex heatmap annotation column
    annoCol <- suppressWarnings(HeatmapAnnotation(df = dplyr::select(anno, all_of(cols)),
                                   name="Main Groups",
                                   col=cc,
                                   which="column",
                                   gap=unit(1,"mm"),
                                   border = T,
                                   show_annotation_name = TRUE,
                                   annotation_name_gp=gpar(fontsize=12),
                                   annotation_name_offset=unit(1,"mm"),
                                   annotation_name_side="left",
                                   annotation_height=unit(colorbar.height, "cm"),
                                   annotation_legend_param = params,
                                   simple_anno_size_adjust=TRUE))
    res <- list("annoColumn"=annoCol)
    if(!is.null(goi)){
      regex <- paste0("^",goi,"$", collapse = "|")
      goi.idx <- grep(regex, rownames(expn))
      labels <- rownames(expn)[goi.idx] #needs to be numeric indexes for complexheatmap
      #create the row (gene) labels object
      labs <- rowAnnotation(link = anno_mark(at=goi.idx,
                                             labels=labels,
                                             which="row",
                                             link_width=unit(1, "mm")),
                            width= unit(1, "mm") + max_text_width(labels),
                            gp=gpar(fontsize=4))

      res[["geneLabels"]] <-  labs
    }
    return(res)
}



complex_hmap <- function(mat,
                        hmap_anno_obj,
                        hmap_anno_obj_genes=NULL,
                        name="z-scores",
                        scale=TRUE,
                        space.type="sRGB",
                        color_palette=NULL,
                        split=NULL, 
                        cluster.method="ward.D2",
                        show_sample_ids=FALSE,
                        dge_dendrograms.res=NULL,
                        samp_dend_order=NULL){
  #mat is the normalized, log2 (usually) transformed counts
  #name is the title
  #scale is whether to scale by row
  #color palette is a colorRamp2() object.
  #threshold is whether to make all z-scores in a certain range.
  #hmap_anno_obj is from HeatmapAnnotation() function
  #space.type is for the color/shades on the heatmap. See ?Heatmap for all the choices.
  #dge_dendrograms.res is the list object output from the dge_dendrograms() function.
  #samp_dend_order is the numeric vector or character vector of column names from the matrix (mat) or the dge_dengrograms.res$TMMCPM matix, in the desired order.

  suppressPackageStartupMessages(library(ComplexHeatmap))
  suppressPackageStartupMessages(require(circlize))
  library(RColorBrewer)
  suppressPackageStartupMessages(library(dendextend))
  ht_opt$message = FALSE

  if(is.null(color_palette)){
    pal <- colorRamp2(c(-4,-2, 0, 2, 4),
                      c("deepskyblue3", "deepskyblue","white", "red", "red3"),
                      space=space.type)
  }else{
    pal <- color_palette
  }
  # col <- colorRampPalette(c("cyan1", "cyan2", "cyan3", "cyan4","azure4","magenta4", "magenta3", "magenta2", "magenta1"))(n=299)
  #colorRamp2(c(-2, 0, 4), c("deepskyblue","white", "red"), space="RGB") #use for breaks.
  # colorPal <- colorRampPalette(c("deepskyblue4", "deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=299)

  #legend graphical parameters
  params <-  list(color_bar="continuous",
       legend_direction="horizontal",
       title_position="leftcenter",
       legend_width=unit(5,"cm"),
       legend_height=unit(5,"cm"),
       title_gp=gpar(fontsize=10,
                     fontface="bold"))


    if(scale){
      mat <- t(scale(t(mat))) ##rowwise scaling
    }
    print(range(mat))
  
    if(!is.null(dge_dendrograms.res)){
        if(is.null(samp_dend_order)){
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                           order=c(ncol(mat):1))
        }else{
          clust <- dendextend::rotate(as.dendrogram(dge_dendrograms.res$samp.c1),
                                  order=samp_dend_order)
      }
    }else{
      clust=TRUE
    }

    #create the heatmap plot
    hmap <- Heatmap(mat,
                    name=name,
                    col=pal,

                    heatmap_legend_param=params,
                    row_title="Genes",
                    row_title_side="left",
                    row_title_gp=gpar(fontsize=15,
                                      fontface="bold"),
                    show_row_names=FALSE,
                    show_column_names=show_sample_ids,
                    row_names_gp=gpar(fontsize=3),

                    column_title="Samples",
                    column_title_side="bottom",
                    column_title_gp=gpar(fontsize=15,
                                         fontface="bold"),
                    column_title_rot=0,
                    row_dend_width=unit(8,"mm"),
                    column_dend_height=unit(22.5,"mm"),

                    top_annotation=hmap_anno_obj,
                    right_annotation = hmap_anno_obj_genes,
                    split=split,

                    clustering_distance_rows="euclidean",
                    clustering_method_rows=cluster.method,
                    clustering_method_columns = cluster.method,
                    cluster_columns = clust,
                    column_dend_reorder=FALSE)

  return(hmap)
}
```

```{r}
create_pf_heatmaps <- function(TMMCPM,sample_annots, gene_annots,
                               center_scale=TRUE,
                               ntop=1000,order_ids=NULL,
                               genelist=NULL,
                               heatmap_colors=NULL){
  pf_genes <- gene_annots %>% 
    filter(genome=="PF3D7") %>% 
    pull(gene_id)
  
  TMMCPM_all_pf <- TMMCPM[rownames(TMMCPM) %in% pf_genes, ]
  dends <- dge_dendrograms(expnData = TMMCPM_all_pf, 
                           pheno = pull(sample_annots, analysis_group, name = analysis_group_id), 
                           method = "ward.D2",
                           genelist = genelist, 
                           ntop=ntop,
                           log=FALSE,
                           filterTopGenes = TRUE,
                           createDGE = FALSE)
  
  # dim(dends$TMMCPM)
  hmap_anno_obj <- create_annots_hmap(expn=dends$TMMCPM,
                                      pheno_df = sample_annots,
                                      cc=NULL,
                                      colorbar.height=2,
                                      cols = c("analysis_group"))
  name <- ifelse(center_scale, "z-scores","log2_CPM")
  complex_hmap(mat=dends$TMMCPM, 
               scale=center_scale,
               name=name,
               hmap_anno_obj = hmap_anno_obj$annoColumn,
               dge_dendrograms.res = dends,
               samp_dend_order=order_ids,
               color_palette=heatmap_colors,
               show_sample_ids=TRUE)
  
}
```


# Project Files 

Main directory:
/active/kappe_s/kappe/Gigliola/BobM_DuffyNGS_PF_LiverStage

Raw Reads: 
/active/kappe_s/kappe/Gigliola/BobM_DuffyNGS_PF_LiverStage/FASTQ

Genome Refs:
/active/taylor_s/people/jsmi26/RSC/kappe_s_2022.04_rnaseq_quant/species_genomes/concat_genome_refs.sh
/active/taylor_s/people/jsmi26/RSC/kappe_s_2022.04_rnaseq_quant/species_genomes

GRCh38, ensembl v106 annotations
GRCm39, ensembl v106 annotations
Pf3D7, PlasmoDB v58 annotations


STAR aligner:
https://childrens-atlassian/bitbucket/projects/RSC/repos/kappe_s_2022.04_rnaseq_quant/browse?at=refs%2Fheads%2Fkappe_s_2022.06 


# Sample Info

```{r}
#For the concatenated samples 
sample_manifest_cated <- read.csv("samples/Kappe_s_2022.04_PF_LiverStage_sra_and_concat_sample_manifest.csv")  %>% 
  set_rownames(.$sample_id_cat)

dim(sample_manifest_cated)
table(sample_manifest_cated$analysis_group)
```

# Select Input Samples

```{r}
pf_samples <- read.csv("samples/Kappe_s_2022.04_PF_LiverStage_ParasitesPaper_sample_manifest_7.25.2022.csv") %>% 
  set_rownames(.$sample_id_cat)

dim(pf_samples)
table(pf_samples$analysis_group)
```

# Gene Counts 

```{r}
combat_matrix_cated <- readRDS("expression_data/kappe_s_PfHsMmu_cated_samples_Pf3D7v58_combat-seq_counts.RDS") %>% 
  as.data.frame()

dim(combat_matrix_cated) #5184   17
# head(combat_matrix_cated)
```



# Genome References

```{r}
#Gene Refs
species_genes <- read.csv("species_genomes/PfHsMmu_GRCh38_GRCm39_Pf3D7v58_genes_annots.csv") %>% 
  select(gene_name, gene_id, everything()) %>% 
  set_rownames(.$gene_id)

# head(species_genes)
```

# Counts and Normalization 

```{r}
pf_combat_dge <- DGEList(counts=combat_matrix_cated,
                         samples=sample_manifest_cated)

pf_combat_dge <- edgeR::calcNormFactors(pf_combat_dge, method="TMMwsp")

pf_TMMCPM <- cpm(pf_combat_dge, normalized.lib.sizes = TRUE,log=FALSE) %>% 
  as.data.frame()

dim(pf_TMMCPM)
```

```{r}
create_long_format_dfs <- function(species_genes,sample_manifest_cated,
                                   counts_matrix_cated, TMMCPM){
  CPM_matrix_anno <- species_genes %>% 
    inner_join(., rownames_to_column(TMMCPM,"gene_id"),
              by="gene_id") %>% 
    pivot_longer(cols = all_of(sample_manifest_cated$sample_id_cat),
                 names_to="sample_id_cat",
                 values_to="TMMCPM") %>% 
    inner_join(., sample_manifest_cated, by="sample_id_cat")

  counts_matrix_cated_anno <- species_genes %>% 
    inner_join(., rownames_to_column(counts_matrix_cated, "gene_id"),
              by="gene_id") %>% 
    pivot_longer(cols = all_of(sample_manifest_cated$sample_id_cat),
                 names_to="sample_id_cat",
                 values_to="counts") %>% 
    inner_join(., sample_manifest_cated, 
               by="sample_id_cat")
  
  return(list("CPM_anno"=CPM_matrix_anno,
              "counts_anno"=counts_matrix_cated_anno,
               "counts"=counts_matrix_cated,
              "TMMCPM"=TMMCPM))
}
```

```{r}
pf_expression <- create_long_format_dfs(species_genes=species_genes,
                                        sample_manifest_cated=sample_manifest_cated,
                                        counts_matrix_cated=combat_matrix_cated, 
                                        TMMCPM = pf_TMMCPM)
```


# Expression EDA 

## Plasmodium Genes 

```{r fig.height=7, fig.width=7}

pf_samples_ids <- unique(pf_expression$CPM_anno$sample_id_cat)

pf_density_plots <- purrr::map(pf_samples_ids,
                               function(samp, CPM_matrix_anno=pf_expression$CPM_anno){
  
  df <- filter(CPM_matrix_anno, sample_id_cat==samp)
  
  if( unique(df$analysis_group) != "control"){
    lab <- paste0("description: ", unique(df$analysis_group))
  }else{
    lab <- paste0("description: ", unique(df$tissue_source))
  }
  ggplot( df, 
          aes(y=genome, x=log2(TMMCPM+1), fill=genome)) +
      ggridges::geom_density_ridges(alpha=0.5, scale=0.75) +
      geom_boxplot(position = position_nudge(y=-0.2),
                   width=0.2,
                   outlier.size=0.5,
                   outlier.alpha=0.5) +
      labs(title=samp) +
      annotate(geom="text", x=10, y=3.5,
               label=lab, size=0.75) +
      theme_classic()
  
})


names(pf_density_plots) <- unique(pf_expression$CPM_anno$analysis_group_id)
length(pf_density_plots)
# pf_density_plots$`Day2_Heps_S1_Pf+`
```

```{r}
groups <- unique(sample_manifest_cated$analysis_group) 
grouped_plots <- purrr::map(groups, function(grp, density_plots=pf_density_plots){
  
  plots <- sample_manifest_cated %>% 
    filter(analysis_group==grp) %>% 
    pull(analysis_group_id)
  
  n <- length(plots)
  nrow <- ifelse(n<=2 | n==4, 2, 3)
  ncol=ceiling(n/nrow)
  p <- marrangeGrob(grobs=density_plots[plots], nrow=nrow, ncol=ncol)
  
  fname <- file.path("figures/combat_seq",paste0("kappe_s_cated_samples_pf_genes_only",grp,".pdf"))
  # print(fname)
  width <- ifelse(ncol>=2, 16, 8)
  ggsave(filename = fname,
         plot = p,
         device = "pdf",
         height = 12, width = width)
})
names(grouped_plots) <- unique(sample_manifest_cated$analysis_group)
```

## Boxplots of Expression 

```{r}
pos <- sample_manifest_cated %>% 
  filter(grepl("cont|spo|Pf\\+", analysis_group))

neg <- sample_manifest_cated %>% 
  filter(grepl("cont|spo|Pf\\-", analysis_group))
```


```{r fig.height=6, fig.width=10}
pos_gex <- pf_expression$CPM_anno %>% 
  filter(sample_id_cat %in% pos$sample_id_cat) %>% 
  mutate(log2_TMMCPM=log2(TMMCPM+1)) %>% 
  arrange(analysis_group) 

neg_gex <- pf_expression$CPM_anno %>% 
  filter(sample_id_cat %in% neg$sample_id_cat) %>% 
  mutate(log2_TMMCPM=log2(TMMCPM+1)) %>% 
  arrange(analysis_group) 


ggplot(pos_gex, aes(x=analysis_group_id, y=log2_TMMCPM)) +
  geom_boxplot(aes(fill=analysis_group)) +
  theme(axis.text.x = element_text(angle=25, vjust=1, hjust=1))

ggplot(neg_gex, aes(x=analysis_group_id, y=log2_TMMCPM)) +
  geom_boxplot(aes(fill=analysis_group)) +
  theme(axis.text.x = element_text(angle=25, vjust=1, hjust=1))
```


## PCA 

### Pf Genes 

```{r}
pf_pheno <- pull(sample_manifest_cated, analysis_group, name=sample_id_cat)
pf_PCA <- PCA(expnData = pf_expression$counts, phenovector = pf_pheno, ntop = 500)
```

```{r}
cc_analysis_group <- get_palette("jco", k=length(unique(sample_manifest_cated$analysis_group))) %>% 
  set_names(unique(sample_manifest_cated$analysis_group))

cc_analysis_group
```

```{r}
pf_pheno_gfp <- pull(sample_manifest_cated, infection_status, name=sample_id_cat)
pf_gfp_PCA <- PCA(expnData = pf_expression$counts, phenovector = pf_pheno_gfp, ntop = 500)
```

```{r fig.height=8}
# pdf("figures/exploratory/kappe_s_Pf_LS_cated_samples_labels_500_mostVarGenes_PF3D7genome.pdf", height = 6, width = 10)
pf_PCA$pca_plots[[1]] +
  scale_color_manual(values=cc_analysis_group) +
  xlim(-150,200) +
    ggrepel::geom_text_repel(data=pf_PCA$pca_data$scores,
                       mapping=aes(x=PC1, y=PC2, label=name),
                       max.overlaps=20, size=3,
                       point.padding = 0.01,
                       min.segment.length=0.01) 
# dev.off()
```

```{r fig.height=8, fig.width=16}
# pdf("figures/combat_seq//kappe_s_Pf_LS_cated_samples_byDay_500_mostVarGenes_PF3D7genome.pdf", height = 8, width = 16)
pf_PCA$pca_plots[[1]] + pf_PCA$pca_plots[[2]] 
# dev.off()

# pdf("figures/combat_seq/kappe_s_Pf_LS_cated_samples_byInfection_500_mostVarGenes_PF3D7genome.pdf", height = 8, width = 16)
pf_gfp_PCA$pca_plots[[1]] + pf_gfp_PCA$pca_plots[[2]]
# dev.off()
```

# Parasites Clustering 

## Most Varied Genes 

```{r}
suppressPackageStartupMessages(library(circlize))
```

```{r}
pf_samples <- pf_samples %>% 
  set_rownames(.$analysis_group_id)

TMMCPM_pf_pos <- pf_expression$TMMCPM[,pf_samples$sample_id_cat] %>% 
                     set_colnames(pf_samples$analysis_group_id)

dim(TMMCPM_pf_pos)
```


```{r fig.height=7, fig.width=14}
# col <- colorRampPalette(c("cyan1", "cyan2", "cyan3", "cyan4","azure4","magenta4", "magenta3", "magenta2", "magenta1"))(n=299)
color_map <- colorRamp2(c(-2.5,-1.5, 0,1.5, 2.5), c("dodgerblue4", "dodgerblue1","white","red1", "red4"), space="sRGB") #use for breaks.

# pdf("figures/combat_seq//kappe_s_pf_pos_top500MostVarGenes_allSamps_07.25.2022.pdf", height = 8, width = 14)
create_pf_heatmaps(TMMCPM=TMMCPM_pf_pos,
                   sample_annots=pf_samples, 
                   gene_annots=species_genes,
                   order_ids = pf_samples$analysis_group_id,
                   heatmap_colors = color_map,
                   ntop=500)
# dev.off()
```

```{r fig.height=7, fig.width=14}
# pdf("figures/combat_seq//kappe_s_pf_pos_top500MostVarGenes_D456_07.01.2022.pdf", height = 8, width = 14)
pf_samples_D456 <- pf_samples %>% 
  filter(!grepl("Day2", time_point)) %>% 
  set_rownames(.$analysis_group_id)

create_pf_heatmaps(TMMCPM=TMMCPM_pf_pos,
                   sample_annots=pf_samples_D456, 
                   gene_annots=species_genes,
                   order_ids = pf_samples_D456$analysis_group_id,
                   heatmap_colors = color_map,
                   ntop=500)
# dev.off()
```

```{r fig.height=7, fig.width=14}
# col <- colorRampPalette(c("cyan1", "cyan2", "cyan3", "cyan4","azure4","magenta4", "magenta3", "magenta2", "magenta1"))(n=299)
col <- colorRamp2(c(-3,-1.5, 0,1.5, 3), c("dodgerblue4", "dodgerblue1","white","red1", "red4"), space="sRGB") #use for breaks.

# pdf("figures/parasites_clustering/kappe_s_pf_pos_neg_controls_top500MostVarGenes_07.25.2022.pdf", height = 8, width = 14)
create_pf_heatmaps(TMMCPM=as.matrix(pf_expression$TMMCPM) %>% 
                     set_colnames(sample_manifest_cated$analysis_group_id),
                   sample_annots=sample_manifest_cated %>% set_rownames(.$analysis_group_id), 
                   gene_annots=species_genes,
                   order_ids = sample_manifest_cated$analysis_group_id,
                   heatmap_colors = col,
                   ntop=500)
# dev.off()
```


## Largest Fold-change Genes

```{r}
de_files <- dir("DEGs", pattern = "combat", full.names = TRUE)
DEGs <- purrr::map_dfr(de_files, function(file){
  condition <- basename(file) %>% 
    gsub("kappe_s_(D.+)_Heps.+","\\1", .)
  
  read.csv(file, comment.char = "#") %>% 
    mutate(direction=case_when(
              logFC > 0 ~ "up",
              logFC < 0 ~ "down"),
         time_point=condition)
})

dim(DEGs)
head(DEGs)
```

```{r}
sel_genes <- DEGs %>% 
  group_by(time_point,direction) %>%
  mutate(largest_effect_size=abs(logFC) >= quantile(abs(logFC), probs=0.90)) %>%
  ungroup() %>% 
  filter(largest_effect_size)


dim(sel_genes)
table(sel_genes$time_point,
      sel_genes$direction) 

```

```{r fig.height=7, fig.width=14}
SCALE=TRUE
if(SCALE){
col <- colorRamp2(c(-3,-1.5, 0,1.5, 3), 
                  c("dodgerblue4", "dodgerblue1","white","red1", "red4"), space="sRGB") #use for breaks.
type <- "zscores"
}else{
col <- colorRamp2(seq(0,20,length.out = 5),
                  c("white",paste0("firebrick",1:4)),
                  space="sRGB") #use for breaks
type="log2CPM"
}

day456_sel <-  sel_genes %>% 
  filter(time_point != "Day2")

# pdf(paste0("figures/combat_seq//kappe_s_pf_cated_samples_Pos_largest_absFC_",type,"_07.25.2022.pdf"), height = 8, width = 14)
no_day2 <- filter(pf_samples,!grepl("Day2", analysis_group)) 
create_pf_heatmaps(TMMCPM=TMMCPM_pf_pos,
                   sample_annots=no_day2, 
                   gene_annots=species_genes,
                   center_scale=SCALE,
                   order_ids = rev(no_day2$analysis_group_id),
                   genelist=unique(day456_sel$gene_id),
                   heatmap_colors = col)
# dev.off()
```

```{r fig.height=7, fig.width=14}
purrr::map(paste0("Day",c(2,4:6)), function(day){
  
  samps <- pf_samples %>% 
    filter(time_point=="Day0" | time_point==day)
  
  genes <- sel_genes %>% 
    filter(time_point==day) 
  
  col <- colorRamp2(c(-2.0,-1, 0, 1, 2.0), 
                    c("dodgerblue4", "dodgerblue1","white","red1", "red4"), 
                    space="sRGB") #use for breaks.
  
  # pdf(paste0("figures/combat_seq//kappe_s_pf_cated_samples_pos_largestFC_",day,"_07.25.2022.pdf"),
  #     height = 8, width = 14)
  print(create_pf_heatmaps(TMMCPM=TMMCPM_pf_pos,
                   sample_annots=samps,
                   gene_annots=species_genes,
                   order_ids = rev(samps$analysis_group_id),
                   genelist=unique(genes$gene_id),
                   heatmap_colors = col))
  # dev.off()
})
```


# Share the Data 

I think I need to consider best practices here... not sure yet. 

```{bash}
rsync -av figures /active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/
rsync -av expression_data /active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/
rsync -av presentations /active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/
```



# Session Information

```{r}
sessionInfo()
```



