---
title: "alternative splicing"
author: "Jenny L Smith"
date: "`r Sys.Date()`"
output: html_document
---

# Set-up 

```{r set-up}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE,
                      fig.align='center',
                      fig.width = 10, fig.height = 10)

options(stringsAsFactors = FALSE, max.print = 100)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(tidyr)
library(tibble)
library(magrittr)
library(dplyr)
library(stringr)
library(ggplot2)
library(gridExtra)
library(ggpubr)
library(patchwork)
library(limma)
library(edgeR)
library(SummarizedExperiment)
library(patchwork)
```

# Define Functions 

```{r}
#Function to create symlinks to the original data files on active to the current working directory of the R project. 
mk_symlinks <- function(linked_dirname,filepaths){
  
  if(!dir.exists(linked_dirname)){
    dir.create(linked_dirname, recursive = TRUE)
  }
  
  lapply(filepaths,function(file){
    target <- file.path(linked_dirname, basename(file))
    if(!file.exists(target)){
      command <- paste0("ln -svf '",file,"' '",target, "'")
      system(command)
    }
  })
}
```


# Sample Manifest and Expression Data

```{r}
pf_samples <- read.csv("samples/Kappe_s_2022.04_PF_LiverStage_ParasitesPaper_sample_manifest_7.25.2022.csv") %>% 
  set_rownames(.$sample_id_cat)

dim(pf_samples)
table(pf_samples$analysis_group)
```

```{r}
cb_se <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_SummarizedExperiment.RDS")

cb_se
```

```{r}
cb_eset <- readRDS("expression_data/kappe_s_PfHsMmu_cated_Pf_ParasitesPaper_samples_Pf3D7v58_combat-seq_expressionSet.RDS")

# cb_eset
```


# Genomic References 

```{r}
edb <- ensembldb::EnsDb("species_genomes/GRCh38_GRCm39_Pf3D7v58.sqlite")

edb
```

```{r}
genome_fasta <- Biostrings::readDNAStringSet("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_Genome.fasta")
names(genome_fasta) <- names(genome_fasta) %>%  str_split_fixed(" | ", n=2) %>% .[,1]
```

STAR can also utilize annotations formatted as a list of splice junctions coordinates in a text file:
--sjdbFileChrStartEnd /path/to/sjdbFile.txt. 

This file should contains 4 columns separated by tabs: Chr \tab Start \tab End \tab Strand=+/-/

Here Start and End are first and last bases of the introns (1-based chromosome coordinates)

```{r}
chrom_sizes <- GRanges(seqnames = S4Vectors::Rle(names(genome_fasta)),
                       ranges = IRanges(start=1,end = width(genome_fasta)))

# chrom_sizes
# export is making the BED file 0 based??
rtracklayer::export.bed(chrom_sizes,
                    con="species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_Genome.chrom_sizes.bed")
```


# Dasper 

Required Inputs:
- STAR splice junctions
- BigWig coverage files

https://bioconductor.org/packages/release/bioc/vignettes/dasper/inst/doc/dasper.html
https://github.com/LieberInstitute/megadepth

```{r}
# BiocManager::install("dasper")
# BiocManager::install("megadepth")

library(dasper)
library(megadepth)

# default installation path: ~/bin
# install_megadepth() # only run 1x
```

## Data Preparation 

### STAR output

```{r}
#symlink the STAR splice junctions 
star_output <- "/active/taylor_s/people/jsmi26/RSC/kappe_s_2022.04_rnaseq_quant/results/star"
indexes <- "/active/taylor_s/people/jsmi26/RSC/kappe_s_2022.04_rnaseq_quant/results/samtools"

if(length(dir("star/")) == 0 ){
 links <-  mk_symlinks(linked_dirname = here::here("star"),
                      filepaths = dir(star_output,
                                      pattern = "SJ.out.tab|.bam",
                                      full.names = TRUE))
 bai_links <- mk_symlinks(linked_dirname = here::here("star"), 
                         filepaths = dir(indexes,
                                         pattern = ".bai",
                                         full.names = TRUE))
}
```


```{r}
bams <- dir("star", 
            pattern = ".bam$", 
            full.names = TRUE) %>% 
  here::here(.)

head(bams)

dir.create("star/pf_aligned",recursive = TRUE)
outbams <- basename(bams) %>% 
  gsub(".Aligned.",".Pf_Aligned.", .) %>%  
  file.path(here::here(),"star/pf_aligned", .)

head(outbams)
```

```{r}
regions_bed <- here::here("species_genomes/plasmoDB/PlasmoDB-58_Pfalciparum3D7_Genome.chrom_sizes.bed")
save_scripts <- purrr::map2(.x = bams, .y = outbams, function(input,output){
  
  samtools_cmd <- glue::glue("samtools view -h -M -L {regions_bed} --output {output} {input}")
  
  script_name <- basename(output) %>% 
    gsub(".sortedByCoord.out.bam", ".sh", .)

  cat(samtools_cmd, 
      file=file.path("star/pf_aligned/", script_name),
      sep="\n")
  
})
```

```{bash}
conda env create -f env/samtools.yml
conda activate samtools
cd star/pf_aligned/
for file in $(ls -1 *.sh)
do
  chmod +x $file 
  ./$file
done
```


```{r}

```



### Convert BAM to BigWigs

```{r}
dir.create("star/bigwigs", recursive = T)

infile <- outbams[1]
in_prefix <- basename(infile) %>% 
  gsub(".bam","",.) %>% 
  paste0("star/bigwigs/", .)

megadepth::bam_to_bigwig(bam_file = infile,
                         prefix = in_prefix,
                         overwrite = TRUE)
```
```{r}
system(paste("ls", grep("Day6.GFP_L1Pf6xTR2_L1Pf6_S5", outbams, value=TRUE))
```


# Share the Data 

```{bash}
destination="/active/kappe_s/kappe/Gigliola/2022.04_jsmi26_PF_LiverStage/"

```



# Session Info

```{r}
sessionInfo()
```

